{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Project Set-up <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"package 'ape' was built under R version 3.6.3\"\n",
      "Warning message:\n",
      "\"package 'cowplot' was built under R version 3.6.3\"\n",
      "\n",
      "********************************************************\n",
      "\n",
      "Note: As of version 1.0.0, cowplot does not change the\n",
      "\n",
      "  default ggplot2 theme anymore. To recover the previous\n",
      "\n",
      "  behavior, execute:\n",
      "  theme_set(theme_cowplot())\n",
      "\n",
      "********************************************************\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'cluster' was built under R version 3.6.3\"\n",
      "Warning message:\n",
      "\"package 'data.table' was built under R version 3.6.3\"\n",
      "Warning message:\n",
      "\"package 'dplyr' was built under R version 3.6.3\"\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'expss' was built under R version 3.6.3\"\n",
      "\n",
      "Attaching package: 'expss'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    between, compute, contains, first, last, na_if, recode, vars\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:data.table':\n",
      "\n",
      "    copy, like\n",
      "\n",
      "\n",
      "The following object is masked from 'package:ape':\n",
      "\n",
      "    where\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'ggplot2' was built under R version 3.6.3\"\n",
      "\n",
      "Attaching package: 'ggplot2'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:expss':\n",
      "\n",
      "    vars\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'ggpubr' was built under R version 3.6.3\"\n",
      "Loading required package: magrittr\n",
      "\n",
      "\n",
      "Attaching package: 'magrittr'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:expss':\n",
      "\n",
      "    and, equals, or\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'ggpubr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:expss':\n",
      "\n",
      "    compare_means\n",
      "\n",
      "\n",
      "The following object is masked from 'package:cowplot':\n",
      "\n",
      "    get_legend\n",
      "\n",
      "\n",
      "The following object is masked from 'package:ape':\n",
      "\n",
      "    rotate\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'Hmisc' was built under R version 3.6.3\"\n",
      "Loading required package: lattice\n",
      "\n",
      "Warning message:\n",
      "\"package 'lattice' was built under R version 3.6.3\"\n",
      "Loading required package: survival\n",
      "\n",
      "Warning message:\n",
      "\"package 'survival' was built under R version 3.6.3\"\n",
      "Loading required package: Formula\n",
      "\n",
      "Registered S3 methods overwritten by 'Hmisc':\n",
      "  method                 from \n",
      "  [.labelled             expss\n",
      "  print.labelled         expss\n",
      "  as.data.frame.labelled expss\n",
      "\n",
      "\n",
      "Attaching package: 'Hmisc'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    src, summarize\n",
      "\n",
      "\n",
      "The following object is masked from 'package:ape':\n",
      "\n",
      "    zoom\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    format.pval, units\n",
      "\n",
      "\n",
      "Warning message:\n",
      "\"package 'otuSummary' was built under R version 3.6.3\"\n",
      "Loading required package: vegan\n",
      "\n",
      "Warning message:\n",
      "\"package 'vegan' was built under R version 3.6.3\"\n",
      "Loading required package: permute\n",
      "\n",
      "Warning message:\n",
      "\"package 'permute' was built under R version 3.6.3\"\n",
      "This is vegan 2.5-6\n",
      "\n",
      "Warning message:\n",
      "\"package 'tibble' was built under R version 3.6.3\"\n",
      "Warning message:\n",
      "\"package 'yaml' was built under R version 3.6.2\"\n"
     ]
    }
   ],
   "source": [
    "source(\"sources/ggrare.R\") #github library: https://rdrr.io/github/gauravsk/ranacapa/\n",
    "source(\"sources/miseqR.R\")\n",
    "source(\"sources/qiime2R.R\") #github https://github.com/jbisanz/qiime2R\n",
    "source('sources/abundances.R') #github https://github.com/microbiome/microbiome\n",
    "\n",
    "library(biomformat)\n",
    "library(ape)\n",
    "library(cowplot)\n",
    "library(cluster)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "library(\"expss\")\n",
    "library(ggplot2)\n",
    "library(\"ggpubr\")\n",
    "library(Hmisc)\n",
    "library(otuSummary)\n",
    "library(phyloseq)\n",
    "library(scales)\n",
    "library(tibble)\n",
    "library(vegan)\n",
    "library(yaml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_dir='/Users/slsevilla/Google Drive/MyDocuments_Current/Education/George Mason University/Dissertation/Data/Aim1/'\n",
    "out_dir=paste(proj_dir,'output/',sep=\"\")\n",
    "img_dir=paste(proj_dir,out_dir,'img/',sep=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Read in taxonomy file\n",
    "CONTdf = read.csv('../taxonomy/Taxonomy_Complete.csv',stringsAsFactors = FALSE)\n",
    "\n",
    "#Convert the tax file to SILVA naming with \"D_5__\"\n",
    "CONTdf_s = CONTdf\n",
    "for (i in 1:nrow(CONTdf)){\n",
    "    old = as.character(CONTdf[i,\"Genus\"])\n",
    "    new = paste(\"D_5__\",old,sep=\"\")\n",
    "    new = gsub(\" \", \".\", new, fixed = TRUE)\n",
    "    CONTdf_s[i,\"Genus\"] = new\n",
    "    \n",
    "    old = as.character(CONTdf[i,\"Family\"])\n",
    "    new = paste(\"D_4__\",old,sep=\"\")\n",
    "    new = gsub(\" \", \".\", new, fixed = TRUE)\n",
    "    CONTdf_s[i,\"Family\"] = new\n",
    "}\n",
    "\n",
    "#Convert the tax file to gg naming with \"f\"\n",
    "CONTdf_g = CONTdf\n",
    "for (i in 1:nrow(CONTdf)){\n",
    "    old = as.character(CONTdf[i,\"Genus\"])\n",
    "    new = paste(\"g__\",old,sep=\"\")\n",
    "    new = gsub(\" \", \".\", new, fixed = TRUE)\n",
    "    CONTdf_g[i,\"Genus\"] = new\n",
    "    \n",
    "    old = as.character(CONTdf[i,\"Family\"])\n",
    "    new = paste(\"f__\",old,sep=\"\")\n",
    "    new = gsub(\" \", \".\", new, fixed = TRUE)\n",
    "    CONTdf_g[i,\"Family\"] = new\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Create PhyloSeq Object<h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_create <- function(tax_in,tab_in){\n",
    "    #Read OTUS\n",
    "    otus<-read_qza(tab_in)\n",
    "\n",
    "    #Read Greengenes taxonomy file\n",
    "    taxonomy<-read_qza(tax_in)\n",
    "\n",
    "    #Create taxonomy table\n",
    "    tax_table = data.frame()\n",
    "    col = 1\n",
    "    row = 1\n",
    "\n",
    "    for (i in taxonomy$data$Taxon){\n",
    "        tax_list = as.list(strsplit(as.character(i),\";\"))\n",
    "        for (i in 1:lengths(tax_list)){\n",
    "            tax_table[row,col] = tax_list[[1]][i]\n",
    "            col = col + 1\n",
    "        }\n",
    "        col = 1\n",
    "        row = row + 1\n",
    "    }\n",
    "    rownames(tax_table)<-taxonomy$data$Feature.ID\n",
    "    colnames(tax_table)<-c(\"Domain\", \"Phylum\", \"Class\", \"Order\", \"Family\", \"Genus\", \"Species\")\n",
    "    tax_convert<-as.matrix(tax_table)  \n",
    "\n",
    "    #read metadata\n",
    "    metadata<-read.table('../manifest/m_cleaned.txt',sep='\\t', header=T, row.names=1, comment=\"\")\n",
    "\n",
    "    #Create phylo object\n",
    "    phy_obj<-phyloseq(otu_table(otus$data, taxa_are_rows = T), tax_table(tax_convert), sample_data(metadata))\n",
    "    print(phy_obj) \n",
    "    return(phy_obj)\n",
    "}   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "phy_ext <- function(phy_in){\n",
    "    #glom to tax level, subset for only ext controls, std processing receipts\n",
    "    phy_obj = tax_glom(phy_in,taxrank=\"Genus\") #only merges genera that have the same upper taxonomy\n",
    "    phy_obj = subset_samples(phy_obj,Sample.Type==\"Ext_Control\") #include only controls, no blanks\n",
    "    return(phy_obj)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "otu_extract <- function(phy_in){\n",
    "    # Extract abundance matrix from the phyloseq object\n",
    "    OTU1 = as(otu_table(phy_in), \"matrix\")\n",
    "    if(taxa_are_rows(phy_in)){OTU1 <- t(OTU1)} # transpose\n",
    "    OTUdf = as.data.frame(OTU1)# Coerce to data.frame\n",
    "    return(OTUdf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_extract <- function(phy_in){\n",
    "    #extract taxonomy information into dataframe\n",
    "    TAXdf = tax_table(phy_in)\n",
    "    return(TAXdf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "otu_tax_clean <- function(OTU_in,TAX_in){\n",
    "    #Remove zero cols\n",
    "    OTUdf_clean = OTU_in[, which(colSums(OTU_in) != 0)]\n",
    "\n",
    "    #rename col from features to genus\n",
    "    for (i in colnames(OTU_in)){\n",
    "        newname = paste(as.character(TAX_in[i,\"Family\"]),as.character(TAX_in[i,\"Genus\"]),sep=\"\")\n",
    "        \n",
    "        #standardize gg naming\n",
    "        newname = trimws(newname) #remove first space in gg naming\n",
    "        newname = sub(\" g__\",\"g__\",newname)\n",
    "        newname = sub(\"\\\\[\",\"\",newname)\n",
    "        newname = sub(\"\\\\]\",\"\",newname)\n",
    "        \n",
    "        if(is.na(newname)){\n",
    "            next;\n",
    "        } else{\n",
    "            names(OTUdf_clean)[names(OTUdf_clean) == i] <- newname\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    #add metadata col\n",
    "    metadata<-read.table('../manifest/m_cleaned.txt',sep='\\t', header=T, row.names=1, comment=\"\")\n",
    "    for (a in rownames(OTUdf_clean)){\n",
    "        OTUdf_clean[a,\"Subject.ID\"] = metadata[a,\"Subject.ID\"]\n",
    "        OTUdf_clean[a,\"Ext.Kit\"] = metadata[a,\"Ext.Kit\"]\n",
    "        OTUdf_clean[a,\"Homo.Status\"] = metadata[a,\"Homo.Status\"]\n",
    "    }\n",
    "    OTUdf_clean = data.frame(OTUdf_clean,stringsAsFactors=FALSE)\n",
    "    \n",
    "    return(OTUdf_clean)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create TP,FP,FN dataframe\n",
    "bench_compare <- function(OTUdf_in,CONTdf_in){\n",
    "    \n",
    "    F1df = OTUdf_in\n",
    "    F1df$Sample_names = rownames(F1df) #create col of sample_names\n",
    "    \n",
    "    #For each of the controls in the cleaned df\n",
    "    for (cont in unique(OTUdf_in$Subject.ID)){\n",
    "        temp=filter(CONTdf_in, get(cont) !=0) #create df of taxa related to control\n",
    "        \n",
    "        #Create taxa list, merging family and genus - this will help avoid id errors from taxonomic differences\n",
    "        #IE one taxa labeled at the genus level may be from family1 and the same taxa may also be from family2\n",
    "        taxa_list=list()\n",
    "        for (x in rownames(temp)){taxa_list = c(taxa_list,paste(temp[x,\"Family\"],temp[x,\"Genus\"],sep=\"\"))}  \n",
    "        \n",
    "        #Create list of samples that match the control\n",
    "        sample_list = filter(F1df, Subject.ID == cont)$Sample_names\n",
    "        \n",
    "        #If the data is not in the df (FN), then add the taxa as a new col\n",
    "        for (taxa in taxa_list){ \n",
    "            if(!(taxa %in% unique(colnames(F1df)))){\n",
    "                F1df[taxa[!(taxa %in% colnames(F1df))]] = 0\n",
    "            } \n",
    "        }\n",
    "\n",
    "        for (sample in sample_list){\n",
    "            for (col in unique(colnames(F1df))[unique(colnames(F1df)) != \"Subject.ID\" & unique(colnames(F1df)) != \"Sample_names\"\n",
    "                & unique(colnames(F1df)) != \"Ext.Kit\" & unique(colnames(F1df)) != \"Homo.Status\"]){ \n",
    "                #removes subjectID, sample_names cols from loop\n",
    "                value_check = F1df[sample,col]\n",
    "\n",
    "                if(col %in% taxa_list && value_check > 0){ #column matches expected taxa, has features\n",
    "                    F1df[sample,col] = \"TP\"\n",
    "                } else if (col %in% taxa_list && value_check == 0){ #column matches expected taxa, no features\n",
    "                    F1df[sample,col] = \"FN\"\n",
    "                } else if ((!col %in% taxa_list) && value_check > 0){ #column doesnt match expected taxa, has features\n",
    "                    F1df[sample,col] = \"FP\"\n",
    "                } else{ #column doesnt match expected taxa, has no features\n",
    "                    next\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    return(F1df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create counts of FP, FN, TP\n",
    "bench_counts <- function(OTUdf_in,F1_in,bmark_in,var_in){\n",
    "    F1df_counts = subset(OTUdf_in, select=c(\"Subject.ID\"))\n",
    "\n",
    "    for (sample in rownames(F1df)){\n",
    "        FN = rowSums(F1_in==\"FN\")[sample][[1]]\n",
    "        FP = rowSums(F1_in==\"FP\")[sample][[1]]\n",
    "        TP = rowSums(F1_in==\"TP\")[sample][[1]]\n",
    "\n",
    "        F1df_counts[sample,\"check\"]= FN + TP\n",
    "        F1df_counts[sample,\"FN\"] = FN\n",
    "        F1df_counts[sample,\"FP\"] = FP\n",
    "        F1df_counts[sample,\"TP\"] = TP\n",
    "        F1df_counts[sample,\"TAR\"] = TP / (TP + FP)\n",
    "        F1df_counts[sample,\"TDR\"] = TP / (TP + FN)\n",
    "\n",
    "        num = F1df_counts[sample,\"TAR\"]*F1df_counts[sample,\"TDR\"]\n",
    "        den = F1df_counts[sample,\"TAR\"]+F1df_counts[sample,\"TDR\"] \n",
    "        F1df_counts[sample,\"F1\"] = 2*(num/den)\n",
    "        F1df_counts[sample,\"Subject.ID\"] = OTUdf_in[sample,\"Subject.ID\"]\n",
    "        \n",
    "        F1df_counts[sample,\"bmark\"] = bmark_in\n",
    "        F1df_counts[sample,\"group.type\"] = var_in\n",
    "    }\n",
    "    return(F1df_counts)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Not used\n",
    "bench_means <- function(F1_in){\n",
    "    df = with(F1_in, tapply(X = FN,       \n",
    "                    INDEX = Subject.ID,   \n",
    "                    FUN = mean,     \n",
    "                    na.rm = TRUE))\n",
    "    return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Plotting <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plot_violin <- function(df_in,x_lab,y_ax,y_lab,col_in){\n",
    "    p = ggplot(df_in, aes(x=get(x_lab), y=get(y_ax), color=get(col_in))) + \n",
    "        geom_violin(trim=FALSE) + \n",
    "        labs(color = col_in) +\n",
    "        theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "              axis.title.x = element_blank(), \n",
    "             legend.position = \"none\") + \n",
    "        ylab(y_lab) +\n",
    "        stat_summary(fun.data=mean_sdl, color=\"red\")\n",
    "    return (p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_f1_violin <- function(F1_in,x_list,y_list,id_in,title,ftype){\n",
    "    \n",
    "    p_1 = plot_violin(F1_in,id_in,x_list[1],y_list[1],id_in)\n",
    "    p_2 = plot_violin(F1_in,id_in,x_list[2],y_list[2],id_in)\n",
    "    p_3 = plot_violin(F1_in,id_in,x_list[3],y_list[3],id_in)\n",
    "\n",
    "    #Create count violin plots\n",
    "    fig = ggarrange(p_1, p_2, p_3, \n",
    "              labels = c(\"A\", \"B\", \"C\"),\n",
    "              title = title,\n",
    "              #common.legend = TRUE,\n",
    "              ncol = 2, nrow = 2)\n",
    "    annotate_figure(fig, #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html\n",
    "                   top = text_grob(title, color = \"blue\", face = \"bold\", size = 14)\n",
    "    )\n",
    "    \n",
    "    ggsave(paste(out_dir,\"/img/violin_f1_\",ftype,\"_\",title,\".jpg\",sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Relative Abunandance <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "cal_relabun <- function(df_in){\n",
    "    temp = df_in\n",
    "    temp$Total = rowSums(df_in[,1:(ncol(df_in)-3)]) #increased from -1 to -3\n",
    "    col_list = colnames(temp[,1:(ncol(temp)-4)])\n",
    "    \n",
    "    for (row in rownames(temp)){\n",
    "        for (col in col_list){\n",
    "            val = temp[row,col]\n",
    "            if(val==0){\n",
    "                next\n",
    "            } else{\n",
    "                temp[row,col] = val/temp[row,\"Total\"]\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    return(temp)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Expected to Observed Correlation <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "cal_EO <- function(df_in,contdf_in,bench_in,title_in){\n",
    "    count=1\n",
    "    plot_list = list()\n",
    "    EOdf_merged = data.frame()\n",
    "    eo_out = data.frame()\n",
    "    \n",
    "    #For each of the controls in the cleaned df\n",
    "    for (cont in unique(df_in$Subject.ID)){\n",
    "\n",
    "        taxdf = create_taxadf(cont,contdf_in)\n",
    "        taxa_list=taxdf$taxa.id\n",
    "        \n",
    "        #Create list of samples that match the control\n",
    "        sample_list = filter(RAdf, Subject.ID == cont)$Sample_name\n",
    "        EOdf= data.frame()\n",
    "        \n",
    "        for (taxa in taxa_list){\n",
    "            for(sample in sample_list){\n",
    "                EOdf[nrow(EOdf)+1,\"Sample_name\"] = sample\n",
    "                EOdf[nrow(EOdf),\"taxa.id\"] = taxa\n",
    "                EOdf[nrow(EOdf),\"Subject.ID\"] = cont\n",
    "                EOdf[nrow(EOdf),\"exp\"] = taxdf[taxa,cont]\n",
    "                EOdf[nrow(EOdf),\"bench\"] = bench_in\n",
    "\n",
    "                if(is.null(df_in[sample,taxa])){\n",
    "                    next\n",
    "                } else{\n",
    "                    EOdf[nrow(EOdf),\"obs\"] = df_in[sample,taxa]\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "        EOdf_merged = rbind(EOdf_merged,EOdf) #Merge dfs\n",
    "        \n",
    "        #Create scatter plots\n",
    "        plot_list[[count]] = plot_EO(EOdf,cont)\n",
    "        count=count+1\n",
    "        \n",
    "        #Calc stats\n",
    "        EOdf[is.na(EOdf)] = 0\n",
    "        temp = cal_EO_stattest(EOdf,cont,bench_in)\n",
    "        eo_out = rbind(eo_out,temp) #merge df       \n",
    "    }\n",
    "    \n",
    "    #print plots\n",
    "    if(bench_in==\"MagAttract PowerSoil DNA Kit\"){ #four controls\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], plot_list[[3]], plot_list[[4]],\n",
    "                        labels = c(\"A\", \"B\", \"C\", \"D\"), #edit number of controls\n",
    "                        ncol = 2, nrow = 3)        \n",
    "    } else if(bench_in==\"QIAamp with Modifications\" | bench_in==\"DSP Virus\" | bench_in == \"96 MagBead DNA Extraction Kit\"){ \n",
    "        #two controls\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], \n",
    "                        labels = c(\"A\", \"B\"), #edit number of controls\n",
    "                        ncol = 2, nrow = 3)\n",
    "    } else{\n",
    "        fig = ggarrange(plot_list[[1]], plot_list[[2]], plot_list[[3]],\n",
    "                        labels = c(\"A\", \"B\", \"C\"), #edit number of controls\n",
    "                        ncol = 2, nrow = 3)\n",
    "    }\n",
    "\n",
    "    annotate_figure(fig, top = text_grob(title_in, color = \"blue\", face = \"bold\", size = 14))\n",
    "    ggsave(paste(out_dir,\"/img/scatter_extstd_\",bench_in,\".jpg\",sep=\"\"))\n",
    "\n",
    "    return(eo_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_taxadf <- function(cont,contdf_in){\n",
    "    taxdf=filter(contdf_in, get(cont) !=0) #create df of taxa related to control\n",
    "    taxdf$taxa.id=paste(taxdf$Family,taxdf$Genus,sep=\"\") #create fam/genus col\n",
    "    taxdf = taxdf[,c(\"taxa.id\",cont),drop=FALSE] #remove all cols but the merged and control col\n",
    "    taxdf = aggregate(get(cont) ~ taxa.id, data = taxdf, sum) #collapse df for any dups    \n",
    "    names(taxdf)[2] = cont\n",
    "    rownames(taxdf) = taxdf$taxa.id\n",
    "    return(taxdf)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_EO <- function(df_in,cont_in){\n",
    "    my_data=subset(df_in,Subject.ID==cont_in)\n",
    "\n",
    "    p = ggscatter(my_data, x = \"obs\", y = \"exp\", \n",
    "              add = \"reg.line\", conf.int = TRUE, \n",
    "              cor.coef = TRUE, cor.method = \"pearson\",\n",
    "              xlab = \"Expected Rel Abunance (%)\", ylab = \"Observed Rel Abun (%)\") +\n",
    "        ggtitle(paste(\"Scatter plot of Rel Abun:\",cont_in,sep=\"\"))\n",
    "    return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "cal_EO_stattest <- function(df_in,cont_in,bench_in){\n",
    "    my_data=subset(df_in,Subject.ID==cont_in)\n",
    "    df_out=data.frame()\n",
    "    \n",
    "    df_out[nrow(df_out)+1,\"bmark\"] = bench_in\n",
    "    df_out[nrow(df_out),\"cont\"] = cont_in\n",
    "    df_out[nrow(df_out),\"shapiro\"] = shapiro.test(my_data$obs)$p.value\n",
    "    df_out[nrow(df_out),\"spear.p\"] = cor.test(my_data$obs,my_data$exp,method = \"spearman\")$p.value\n",
    "    df_out[nrow(df_out),\"spear.rho\"] = cor.test(my_data$obs,my_data$exp,method = \"spearman\")$estimate[[1]]\n",
    "\n",
    "    #-1 indicates a strong negative correlation : this means that every time x increases, y decreases (left panel figure)\n",
    "    #0 means that there is no association between the two variables (x and y) (middle panel figure)\n",
    "    #1 indicates a strong positive correlation : this means that y increases with x (right panel figure)\n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Dissimilarity <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create averaged relative abundance df\n",
    "create_av_RAdf <- function(RAdf_in,bench_in,var_in){\n",
    "    tmp = aggregate(RAdf_in[,1:(ncol(RAdf_in)-5)], list(RAdf_in$Subject.ID), mean) #increase from 3 to 5\n",
    "    tmp$bmark=bench_in\n",
    "    tmp$group.type = var_in\n",
    "    return(tmp)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Merged metrics <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Add expected values to RA average dfs\n",
    "RAdf_addcontrols<-function(cont_in,df_in){\n",
    "    \n",
    "    tmp=get(df_in)\n",
    "    cont_list = unique(tmp$Group.1)\n",
    "    cont_leng = length(cont_list) #chnaged to Group.1\n",
    "    \n",
    "    for(i in 1:cont_leng){\n",
    "        cont = as.character(cont_list[i])\n",
    "        taxadf = create_taxadf(cont,get(cont_in))\n",
    "        tmp_count = nrow(tmp)+1\n",
    "        \n",
    "        for(tx_row in 1:nrow(taxadf)){\n",
    "            tmp[tmp_count,\"Group.1\"]=cont\n",
    "            tmp[tmp_count,taxadf[tx_row,\"taxa.id\"]] = taxadf[tx_row,cont]\n",
    "            tmp[tmp_count,\"bmark\"]=\"expected\"\n",
    "            tmp[tmp_count,\"group.type\"]=tmp[i,\"group.type\"]\n",
    "        }\n",
    "        \n",
    "    }\n",
    "    tmp[is.na(tmp)] = 0 \n",
    "    return(tmp)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "calc_dist <- function(RA_in,CONTdf_in){\n",
    "    rownames(RA_in)=paste(RA_in$Group.1,RA_in$bmark,sep=\"-\")\n",
    "    RA_in = subset(RA_in, select = -c(group.type,Group.1,bmark))\n",
    "\n",
    "    taxa_list<-unique(paste(CONTdf_in$Family,CONTdf_in$Genus,sep=\"\"))\n",
    "    df_clean = RA_in[,colnames(RA_in) %in% taxa_list] #remove cols not expected\n",
    "\n",
    "    df_out = vegdist(df_clean,method=\"bray\")   \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_dist_cluster <- function(df_in,ref_in){\n",
    "    fname = paste(out_dir,\"/img/hclust_summary_bc_allgroups_\",ref_in,\".png\",sep=\"\")\n",
    "    png(fname,width = 900, height = 600)\n",
    "    p = plot(hclust(df_in), cex = .9, main = \"All Groups - Bray-Curtis Dissimilarity\",\n",
    "             xlab=\"Bioinformatic Variables\", sub=\"\")\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_dist_violin <- function(df_in,ref_in){\n",
    "    df_in = as.data.frame(as.matrix(df_in))\n",
    "    \n",
    "    df_plot=data.frame()\n",
    "    for (row in rownames(df_in)){\n",
    "        splits = strsplit(row,\"-\")\n",
    "        if(splits[[1]][2]==\"expected\"){\n",
    "            next\n",
    "        }else{\n",
    "            df_plot[nrow(df_plot)+1,\"dist\"]=df_in[row,paste(splits[[1]][1],\"expected\",sep=\"-\")]\n",
    "            df_plot[nrow(df_plot),\"name\"]=splits[[1]][2]\n",
    "        }\n",
    "    }\n",
    "\n",
    "    p = ggplot(df_plot, aes(x=name, y=dist, color=name)) + \n",
    "            geom_violin(trim=FALSE) + \n",
    "            labs(color = df_plot$name) +\n",
    "            theme(axis.text.x = element_text(angle = 90, hjust = 1),\n",
    "                  axis.title.x = element_blank(), \n",
    "                 legend.position = \"none\") + \n",
    "            xlab(\"Benchmark ID\") +\n",
    "            ylab(\"Bray-Curtis Dissimilarity\") +\n",
    "            stat_summary(fun.data=mean_sdl, color=\"red\")\n",
    "    \n",
    "    \n",
    "    fig = ggarrange(p, \n",
    "              #labels = c(\"A\"),\n",
    "              #top = as.character(title_in),\n",
    "              #common.legend = TRUE#,\n",
    "              ncol = 1, nrow = 1)\n",
    "    annotate_figure(fig, top = text_grob(\"Bray-Curtis Dissimilarity\", color = \"blue\", face = \"bold\", size = 14)\n",
    "    )\n",
    "    ggsave(paste(out_dir,\"/img/violin_bc_summary_allgroups_\",ref_in,\".jpg\",sep=\"\"))\n",
    "    return(df_plot)\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Statistics <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "tstat_paired_f1 <- function(f1_in,subgroup_in){\n",
    "    \n",
    "    df_out = data.frame()\n",
    "    \n",
    "    cal_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "    \n",
    "    for (cal in cal_list){\n",
    "    \n",
    "        comb_list = combn(unique(f1_in[,subgroup_in]),2)\n",
    "        count=1\n",
    "        \n",
    "        for (i in 1:(length(comb_list)/2)){\n",
    "            \n",
    "            tstatdf=data.frame()\n",
    "            var1 = as.character(comb_list[count])\n",
    "            var2 = as.character(comb_list[count+1])\n",
    "            count=count+2\n",
    "\n",
    "            df = f1_in %>%\n",
    "                filter(f1_in[,subgroup_in] == var1 | f1_in[,subgroup_in] == var2) %>%\n",
    "                select(subgroup_in, cal)\n",
    "            \n",
    "            tmp = t.test(get(cal) ~ get(subgroup_in), data = df)\n",
    "            tstatdf[nrow(tstatdf)+1,\"level\"] = cal\n",
    "            tstatdf[nrow(tstatdf),\"comparison\"] = paste(var1,var2,sep=\"_\")\n",
    "            tstatdf[nrow(tstatdf),\"pval\"] = tmp$p.value\n",
    "            if(tmp$p.value<0.05){\n",
    "                tstatdf[nrow(tstatdf),\"sig\"] = \"sig\"\n",
    "            } else{\n",
    "                tstatdf[nrow(tstatdf),\"sig\"] = \"ns\"\n",
    "            }\n",
    "            \n",
    "\n",
    "            df_out = bind_rows(df_out,tstatdf)\n",
    "        }\n",
    "    }\n",
    "        \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "adonis_bc <- function(df_in){\n",
    "\n",
    "    df_out=data.frame()\n",
    "    \n",
    "    st = adonis(df_in[,1] ~ name,\n",
    "                data=df_in, permutations = 99, method=\"bray\")\n",
    "    df_out[nrow(df_out)+1,\"comparison\"] = \"bray-curtis\"\n",
    "    df_out[nrow(df_out),\"pval\"] = as.numeric(st$aov.tab$\"Pr(>F)\"[1])\n",
    "        \n",
    "    if(df_out[nrow(df_out),\"pval\"]<0.05){\n",
    "        df_out[nrow(df_out),\"sig\"] = \"sig\"\n",
    "    } else{\n",
    "        df_out[nrow(df_out),\"sig\"] = \"ns\"\n",
    "    }   \n",
    "    \n",
    "    dist_s = vegdist(subset(dissdf_s,select = -c(bmark,Group.1,group.type)),na.rm=TRUE)\n",
    "\n",
    "        \n",
    "    df_out[nrow(df_out)+1,\"homo_s\"] = anova(betadisper(dist_s,dissdf_s$bmark))$\"Pr(>F)\"[1]\n",
    "    \n",
    "    return(df_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Run Production Code <h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phyloseq-class experiment-level object\n",
      "otu_table()   OTU Table:         [ 1818 taxa and 351 samples ]\n",
      "sample_data() Sample Data:       [ 351 samples by 24 sample variables ]\n",
      "tax_table()   Taxonomy Table:    [ 1818 taxa by 7 taxonomic ranks ]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`set.seed(1)` was used to initialize repeatable random subsampling.\n",
      "\n",
      "Please record this for your records so others can reproduce.\n",
      "\n",
      "Try `set.seed(1); .Random.seed` for the full vector\n",
      "\n",
      "...\n",
      "\n",
      "40 samples removedbecause they contained fewer reads than `sample.size`.\n",
      "\n",
      "Up to first five removed samples are: \n",
      "\n",
      "\n",
      "SC249366-PC04925-D-01SC249406SC249422-PC04925-D-03SC249431SC253191\n",
      "\n",
      "...\n",
      "\n",
      "56OTUs were removed because they are no longer \n",
      "present in any sample after random subsampling\n",
      "\n",
      "\n",
      "...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Create the phyloseq object\n",
    "tax_file = paste(out_dir,'taxonomic_classification/classify-sklearn_silva-132-99-nb-classifier.qza',sep=\"\")\n",
    "table_file = paste(out_dir,'denoising/feature_tables/merged_filtered.qza',sep=\"\")\n",
    "phy_obj = phylo_create(tax_file,table_file)  \n",
    "\n",
    "#Create glomed, subsetted obj\n",
    "std_obj = phy_ext(phy_obj)\n",
    "\n",
    "#Rarefy samples\n",
    "phy_r = rarefy_even_depth(std_obj, rngseed=1, sample.size=20000, replace=F) \n",
    "\n",
    "#Create OTUdf\n",
    "OTUdf = otu_extract(phy_r)\n",
    "\n",
    "#Create TAXdf\n",
    "TAXdf = tax_extract(phy_r)\n",
    "\n",
    "#Clean df's\n",
    "OTUdf_clean = otu_tax_clean(OTUdf,TAXdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Standard Processing <h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>DSP Virus</li><li>96 MagBead DNA Extraction Kit</li><li>MagAttract PowerSoil DNA Kit</li><li>MagAttract PowerMicrobiome Kit</li><li>QIAamp with Modifications</li><li>DNeasy PowerSoil Pro kit</li><li>MagMax Microbiome Ultra Kit</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'96 MagBead DNA Extraction Kit'</li><li>'DNeasy PowerSoil Pro kit'</li><li>'DSP Virus'</li><li>'MagAttract PowerMicrobiome Kit'</li><li>'MagAttract PowerSoil DNA Kit'</li><li>'MagMax Microbiome Ultra Kit'</li><li>'QIAamp with Modifications'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item DSP Virus\n",
       "\\item 96 MagBead DNA Extraction Kit\n",
       "\\item MagAttract PowerSoil DNA Kit\n",
       "\\item MagAttract PowerMicrobiome Kit\n",
       "\\item QIAamp with Modifications\n",
       "\\item DNeasy PowerSoil Pro kit\n",
       "\\item MagMax Microbiome Ultra Kit\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item '96 MagBead DNA Extraction Kit'\n",
       "\\item 'DNeasy PowerSoil Pro kit'\n",
       "\\item 'DSP Virus'\n",
       "\\item 'MagAttract PowerMicrobiome Kit'\n",
       "\\item 'MagAttract PowerSoil DNA Kit'\n",
       "\\item 'MagMax Microbiome Ultra Kit'\n",
       "\\item 'QIAamp with Modifications'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. DSP Virus\n",
       "2. 96 MagBead DNA Extraction Kit\n",
       "3. MagAttract PowerSoil DNA Kit\n",
       "4. MagAttract PowerMicrobiome Kit\n",
       "5. QIAamp with Modifications\n",
       "6. DNeasy PowerSoil Pro kit\n",
       "7. MagMax Microbiome Ultra Kit\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. '96 MagBead DNA Extraction Kit'\n",
       "2. 'DNeasy PowerSoil Pro kit'\n",
       "3. 'DSP Virus'\n",
       "4. 'MagAttract PowerMicrobiome Kit'\n",
       "5. 'MagAttract PowerSoil DNA Kit'\n",
       "6. 'MagMax Microbiome Ultra Kit'\n",
       "7. 'QIAamp with Modifications'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] DSP Virus                      96 MagBead DNA Extraction Kit \n",
       "[3] MagAttract PowerSoil DNA Kit   MagAttract PowerMicrobiome Kit\n",
       "[5] QIAamp with Modifications      DNeasy PowerSoil Pro kit      \n",
       "[7] MagMax Microbiome Ultra Kit   \n",
       "7 Levels: 96 MagBead DNA Extraction Kit DNeasy PowerSoil Pro kit ... QIAamp with Modifications"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#List of all variables to test\n",
    "test_list = unique(sample_data(std_obj)$Ext.Kit)\n",
    "test_list\n",
    "#test_list = test_list[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a merged dfs for all variables\n",
    "mergedf_f1stats = data.frame()\n",
    "EOdf_stats_merged = data.frame()\n",
    "dissdf_s = data.frame()\n",
    "dissdf_g = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Sample_name                                          taxa.id\n",
      "1  SC249362-PC07578-H-08                    D_4__BacillaceaeD_5__Bacillus\n",
      "2  SC249368-PC07578-C-10                    D_4__BacillaceaeD_5__Bacillus\n",
      "3  SC249362-PC04924-E-01                    D_4__BacillaceaeD_5__Bacillus\n",
      "4  SC249368-PC04924-B-02                    D_4__BacillaceaeD_5__Bacillus\n",
      "5               SC253844                    D_4__BacillaceaeD_5__Bacillus\n",
      "6               SC253854                    D_4__BacillaceaeD_5__Bacillus\n",
      "7  SC249362-PC07578-H-08 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "8  SC249368-PC07578-C-10 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "9  SC249362-PC04924-E-01 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "10 SC249368-PC04924-B-02 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "11              SC253844 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "12              SC253854 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "13 SC249362-PC07578-H-08           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "14 SC249368-PC07578-C-10           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "15 SC249362-PC04924-E-01           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "16 SC249368-PC04924-B-02           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "17              SC253844           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "18              SC253854           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "19 SC249362-PC07578-H-08            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "20 SC249368-PC07578-C-10            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "21 SC249362-PC04924-E-01            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "22 SC249368-PC04924-B-02            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "23              SC253844            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "24              SC253854            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "25 SC249362-PC07578-H-08          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "26 SC249368-PC07578-C-10          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "27 SC249362-PC04924-E-01          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "28 SC249368-PC04924-B-02          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "29              SC253844          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "30              SC253854          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "31 SC249362-PC07578-H-08                   D_4__ListeriaceaeD_5__Listeria\n",
      "32 SC249368-PC07578-C-10                   D_4__ListeriaceaeD_5__Listeria\n",
      "33 SC249362-PC04924-E-01                   D_4__ListeriaceaeD_5__Listeria\n",
      "34 SC249368-PC04924-B-02                   D_4__ListeriaceaeD_5__Listeria\n",
      "35              SC253844                   D_4__ListeriaceaeD_5__Listeria\n",
      "36              SC253854                   D_4__ListeriaceaeD_5__Listeria\n",
      "37 SC249362-PC07578-H-08            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "38 SC249368-PC07578-C-10            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "39 SC249362-PC04924-E-01            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "40 SC249368-PC04924-B-02            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "41              SC253844            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "42              SC253854            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "43 SC249362-PC07578-H-08        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "44 SC249368-PC07578-C-10        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "45 SC249362-PC04924-E-01        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "46 SC249368-PC04924-B-02        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "47              SC253844        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "48              SC253854        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "   Subject.ID   exp     bench     obs\n",
      "1       D6300 0.174 DSP Virus 0.22350\n",
      "2       D6300 0.174 DSP Virus 0.13625\n",
      "3       D6300 0.174 DSP Virus 0.20140\n",
      "4       D6300 0.174 DSP Virus 0.12050\n",
      "5       D6300 0.174 DSP Virus 0.11235\n",
      "6       D6300 0.174 DSP Virus 0.15170\n",
      "7       D6300 0.101 DSP Virus 0.22525\n",
      "8       D6300 0.101 DSP Virus 0.28800\n",
      "9       D6300 0.101 DSP Virus 0.28110\n",
      "10      D6300 0.101 DSP Virus 0.33650\n",
      "11      D6300 0.101 DSP Virus 0.35670\n",
      "12      D6300 0.101 DSP Virus 0.27375\n",
      "13      D6300 0.104 DSP Virus 0.00000\n",
      "14      D6300 0.104 DSP Virus 0.00000\n",
      "15      D6300 0.104 DSP Virus 0.00000\n",
      "16      D6300 0.104 DSP Virus 0.00000\n",
      "17      D6300 0.104 DSP Virus 0.00000\n",
      "18      D6300 0.104 DSP Virus 0.00000\n",
      "19      D6300 0.099 DSP Virus 0.13165\n",
      "20      D6300 0.099 DSP Virus 0.09055\n",
      "21      D6300 0.099 DSP Virus 0.12905\n",
      "22      D6300 0.099 DSP Virus 0.09210\n",
      "23      D6300 0.099 DSP Virus 0.05685\n",
      "24      D6300 0.099 DSP Virus 0.07735\n",
      "25      D6300 0.184 DSP Virus 0.19185\n",
      "26      D6300 0.184 DSP Virus 0.24735\n",
      "27      D6300 0.184 DSP Virus 0.17180\n",
      "28      D6300 0.184 DSP Virus 0.25375\n",
      "29      D6300 0.184 DSP Virus 0.23365\n",
      "30      D6300 0.184 DSP Virus 0.19650\n",
      "31      D6300 0.141 DSP Virus 0.10675\n",
      "32      D6300 0.141 DSP Virus 0.10235\n",
      "33      D6300 0.141 DSP Virus 0.10310\n",
      "34      D6300 0.141 DSP Virus 0.08770\n",
      "35      D6300 0.141 DSP Virus 0.08860\n",
      "36      D6300 0.141 DSP Virus 0.10225\n",
      "37      D6300 0.042 DSP Virus 0.02690\n",
      "38      D6300 0.042 DSP Virus 0.04455\n",
      "39      D6300 0.042 DSP Virus 0.03360\n",
      "40      D6300 0.042 DSP Virus 0.04860\n",
      "41      D6300 0.042 DSP Virus 0.06465\n",
      "42      D6300 0.042 DSP Virus 0.05985\n",
      "43      D6300 0.155 DSP Virus 0.09305\n",
      "44      D6300 0.155 DSP Virus 0.08935\n",
      "45      D6300 0.155 DSP Virus 0.07885\n",
      "46      D6300 0.155 DSP Virus 0.06075\n",
      "47      D6300 0.155 DSP Virus 0.08600\n",
      "48      D6300 0.155 DSP Virus 0.13795\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Sample_name                                          taxa.id Subject.ID\n",
      "1     SC249363              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "2     SC253848              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "3     SC253851              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "4     SC249363      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "5     SC253848      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "6     SC253851      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "7     SC249363             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "8     SC253848             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "9     SC253851             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "10    SC249363           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "11    SC253848           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "12    SC253851           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "13    SC249363           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "14    SC253848           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "15    SC253851           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "16    SC249363 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "17    SC253848 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "18    SC253851 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "19    SC249363            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "20    SC253848            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "21    SC253851            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "22    SC249363        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "23    SC253848        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "24    SC253851        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "25    SC249363          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "26    SC253848          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "27    SC253851          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "28    SC249363            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "29    SC253848            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "30    SC253851            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "31    SC249363       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "32    SC253848       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "33    SC253851       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "34    SC249363          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "35    SC253848          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "36    SC253851          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "37    SC249363            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "38    SC253848            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "39    SC253851            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "40    SC249363          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "41    SC253848          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "42    SC253851          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "43    SC249363      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "44    SC253848      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "45    SC253851      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "46    SC249363  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "47    SC253848  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "48    SC253851  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "49    SC249363                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "50    SC253848                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "51    SC253851                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "52    SC249363         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "53    SC253848         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "54    SC253851         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "55    SC249363           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "56    SC253848           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "57    SC253851           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "          exp     bench     obs\n",
      "1  0.06263920 DSP Virus 0.10155\n",
      "2  0.06263920 DSP Virus 0.06640\n",
      "3  0.06263920 DSP Virus 0.09765\n",
      "4  0.06263920 DSP Virus 0.01415\n",
      "5  0.06263920 DSP Virus 0.01865\n",
      "6  0.06263920 DSP Virus 0.01365\n",
      "7  0.06263920 DSP Virus 0.00000\n",
      "8  0.06263920 DSP Virus 0.00000\n",
      "9  0.06263920 DSP Virus 0.00000\n",
      "10 0.06263920 DSP Virus 0.00020\n",
      "11 0.06263920 DSP Virus 0.00000\n",
      "12 0.06263920 DSP Virus 0.00000\n",
      "13 0.02352450 DSP Virus 0.02165\n",
      "14 0.02352450 DSP Virus 0.02860\n",
      "15 0.02352450 DSP Virus 0.02970\n",
      "16 0.06263920 DSP Virus 0.11175\n",
      "17 0.06263920 DSP Virus 0.08575\n",
      "18 0.06263920 DSP Virus 0.12330\n",
      "19 0.06263920 DSP Virus 0.01605\n",
      "20 0.06263920 DSP Virus 0.01145\n",
      "21 0.06263920 DSP Virus 0.02235\n",
      "22 0.03660913 DSP Virus 0.00000\n",
      "23 0.03660913 DSP Virus 0.00000\n",
      "24 0.03660913 DSP Virus 0.00000\n",
      "25 0.12527840 DSP Virus 0.25280\n",
      "26 0.12527840 DSP Virus 0.28030\n",
      "27 0.12527840 DSP Virus 0.22710\n",
      "28 0.06263920 DSP Virus 0.02235\n",
      "29 0.06263920 DSP Virus 0.01950\n",
      "30 0.06263920 DSP Virus 0.02120\n",
      "31 0.06263920 DSP Virus 0.09285\n",
      "32 0.06263920 DSP Virus 0.08445\n",
      "33 0.06263920 DSP Virus 0.12760\n",
      "34 0.03660913 DSP Virus 0.04075\n",
      "35 0.03660913 DSP Virus 0.14960\n",
      "36 0.03660913 DSP Virus 0.03470\n",
      "37 0.04175947 DSP Virus 0.02180\n",
      "38 0.04175947 DSP Virus 0.01165\n",
      "39 0.04175947 DSP Virus 0.02730\n",
      "40 0.03660913 DSP Virus 0.03295\n",
      "41 0.03660913 DSP Virus 0.02470\n",
      "42 0.03660913 DSP Virus 0.03230\n",
      "43 0.06263920 DSP Virus 0.00000\n",
      "44 0.06263920 DSP Virus 0.00000\n",
      "45 0.06263920 DSP Virus 0.00000\n",
      "46 0.06263920 DSP Virus 0.00000\n",
      "47 0.06263920 DSP Virus 0.00000\n",
      "48 0.06263920 DSP Virus 0.00000\n",
      "49 0.03660913 DSP Virus 0.00000\n",
      "50 0.03660913 DSP Virus 0.00000\n",
      "51 0.03660913 DSP Virus 0.00000\n",
      "52 0.01572940 DSP Virus 0.00725\n",
      "53 0.01572940 DSP Virus 0.00515\n",
      "54 0.01572940 DSP Virus 0.00635\n",
      "55 0.02087973 DSP Virus 0.00705\n",
      "56 0.02087973 DSP Virus 0.02560\n",
      "57 0.02087973 DSP Virus 0.00580\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 6 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 6 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 6 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 15 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 15 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 15 rows containing missing values (geom_point).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Sample_name                                          taxa.id\n",
      "1  SC249391-PC04925-H-01                    D_4__BacillaceaeD_5__Bacillus\n",
      "2               SC304092                    D_4__BacillaceaeD_5__Bacillus\n",
      "3               SC304097                    D_4__BacillaceaeD_5__Bacillus\n",
      "4  SC304924-PC07578-A-11                    D_4__BacillaceaeD_5__Bacillus\n",
      "5  SC304924-PC07578-B-07                    D_4__BacillaceaeD_5__Bacillus\n",
      "6               SC304927                    D_4__BacillaceaeD_5__Bacillus\n",
      "7               SC249390                    D_4__BacillaceaeD_5__Bacillus\n",
      "8  SC249391-PC04924-D-03                    D_4__BacillaceaeD_5__Bacillus\n",
      "9  SC249391-PC04925-H-01 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "10              SC304092 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "11              SC304097 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "12 SC304924-PC07578-A-11 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "13 SC304924-PC07578-B-07 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "14              SC304927 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "15              SC249390 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "16 SC249391-PC04924-D-03 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "17 SC249391-PC04925-H-01           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "18              SC304092           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "19              SC304097           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "20 SC304924-PC07578-A-11           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "21 SC304924-PC07578-B-07           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "22              SC304927           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "23              SC249390           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "24 SC249391-PC04924-D-03           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "25 SC249391-PC04925-H-01            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "26              SC304092            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "27              SC304097            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "28 SC304924-PC07578-A-11            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "29 SC304924-PC07578-B-07            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "30              SC304927            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "31              SC249390            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "32 SC249391-PC04924-D-03            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "33 SC249391-PC04925-H-01          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "34              SC304092          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "35              SC304097          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "36 SC304924-PC07578-A-11          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "37 SC304924-PC07578-B-07          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "38              SC304927          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "39              SC249390          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "40 SC249391-PC04924-D-03          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "41 SC249391-PC04925-H-01                   D_4__ListeriaceaeD_5__Listeria\n",
      "42              SC304092                   D_4__ListeriaceaeD_5__Listeria\n",
      "43              SC304097                   D_4__ListeriaceaeD_5__Listeria\n",
      "44 SC304924-PC07578-A-11                   D_4__ListeriaceaeD_5__Listeria\n",
      "45 SC304924-PC07578-B-07                   D_4__ListeriaceaeD_5__Listeria\n",
      "46              SC304927                   D_4__ListeriaceaeD_5__Listeria\n",
      "47              SC249390                   D_4__ListeriaceaeD_5__Listeria\n",
      "48 SC249391-PC04924-D-03                   D_4__ListeriaceaeD_5__Listeria\n",
      "49 SC249391-PC04925-H-01            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "50              SC304092            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "51              SC304097            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "52 SC304924-PC07578-A-11            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "53 SC304924-PC07578-B-07            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "54              SC304927            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "55              SC249390            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "56 SC249391-PC04924-D-03            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "57 SC249391-PC04925-H-01        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "58              SC304092        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "59              SC304097        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "60 SC304924-PC07578-A-11        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "61 SC304924-PC07578-B-07        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "62              SC304927        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "63              SC249390        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "64 SC249391-PC04924-D-03        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "   Subject.ID   exp                         bench     obs\n",
      "1       D6300 0.174 96 MagBead DNA Extraction Kit 0.08265\n",
      "2       D6300 0.174 96 MagBead DNA Extraction Kit 0.13085\n",
      "3       D6300 0.174 96 MagBead DNA Extraction Kit 0.15215\n",
      "4       D6300 0.174 96 MagBead DNA Extraction Kit 0.19030\n",
      "5       D6300 0.174 96 MagBead DNA Extraction Kit 0.20720\n",
      "6       D6300 0.174 96 MagBead DNA Extraction Kit 0.23260\n",
      "7       D6300 0.174 96 MagBead DNA Extraction Kit 0.09965\n",
      "8       D6300 0.174 96 MagBead DNA Extraction Kit 0.07795\n",
      "9       D6300 0.101 96 MagBead DNA Extraction Kit 0.33180\n",
      "10      D6300 0.101 96 MagBead DNA Extraction Kit 0.27315\n",
      "11      D6300 0.101 96 MagBead DNA Extraction Kit 0.28510\n",
      "12      D6300 0.101 96 MagBead DNA Extraction Kit 0.23675\n",
      "13      D6300 0.101 96 MagBead DNA Extraction Kit 0.16975\n",
      "14      D6300 0.101 96 MagBead DNA Extraction Kit 0.21785\n",
      "15      D6300 0.101 96 MagBead DNA Extraction Kit 0.32415\n",
      "16      D6300 0.101 96 MagBead DNA Extraction Kit 0.31570\n",
      "17      D6300 0.104 96 MagBead DNA Extraction Kit 0.00000\n",
      "18      D6300 0.104 96 MagBead DNA Extraction Kit 0.00000\n",
      "19      D6300 0.104 96 MagBead DNA Extraction Kit 0.00000\n",
      "20      D6300 0.104 96 MagBead DNA Extraction Kit 0.00000\n",
      "21      D6300 0.104 96 MagBead DNA Extraction Kit 0.00000\n",
      "22      D6300 0.104 96 MagBead DNA Extraction Kit 0.00000\n",
      "23      D6300 0.104 96 MagBead DNA Extraction Kit 0.00000\n",
      "24      D6300 0.104 96 MagBead DNA Extraction Kit 0.00000\n",
      "25      D6300 0.099 96 MagBead DNA Extraction Kit 0.05905\n",
      "26      D6300 0.099 96 MagBead DNA Extraction Kit 0.07080\n",
      "27      D6300 0.099 96 MagBead DNA Extraction Kit 0.08025\n",
      "28      D6300 0.099 96 MagBead DNA Extraction Kit 0.05505\n",
      "29      D6300 0.099 96 MagBead DNA Extraction Kit 0.08105\n",
      "30      D6300 0.099 96 MagBead DNA Extraction Kit 0.09000\n",
      "31      D6300 0.099 96 MagBead DNA Extraction Kit 0.07245\n",
      "32      D6300 0.099 96 MagBead DNA Extraction Kit 0.06765\n",
      "33      D6300 0.184 96 MagBead DNA Extraction Kit 0.27560\n",
      "34      D6300 0.184 96 MagBead DNA Extraction Kit 0.22240\n",
      "35      D6300 0.184 96 MagBead DNA Extraction Kit 0.18430\n",
      "36      D6300 0.184 96 MagBead DNA Extraction Kit 0.20345\n",
      "37      D6300 0.184 96 MagBead DNA Extraction Kit 0.18060\n",
      "38      D6300 0.184 96 MagBead DNA Extraction Kit 0.16275\n",
      "39      D6300 0.184 96 MagBead DNA Extraction Kit 0.23325\n",
      "40      D6300 0.184 96 MagBead DNA Extraction Kit 0.30760\n",
      "41      D6300 0.141 96 MagBead DNA Extraction Kit 0.08795\n",
      "42      D6300 0.141 96 MagBead DNA Extraction Kit 0.10750\n",
      "43      D6300 0.141 96 MagBead DNA Extraction Kit 0.09625\n",
      "44      D6300 0.141 96 MagBead DNA Extraction Kit 0.11650\n",
      "45      D6300 0.141 96 MagBead DNA Extraction Kit 0.10220\n",
      "46      D6300 0.141 96 MagBead DNA Extraction Kit 0.13965\n",
      "47      D6300 0.141 96 MagBead DNA Extraction Kit 0.10180\n",
      "48      D6300 0.141 96 MagBead DNA Extraction Kit 0.08200\n",
      "49      D6300 0.042 96 MagBead DNA Extraction Kit 0.03480\n",
      "50      D6300 0.042 96 MagBead DNA Extraction Kit 0.02870\n",
      "51      D6300 0.042 96 MagBead DNA Extraction Kit 0.04365\n",
      "52      D6300 0.042 96 MagBead DNA Extraction Kit 0.05665\n",
      "53      D6300 0.042 96 MagBead DNA Extraction Kit 0.14350\n",
      "54      D6300 0.042 96 MagBead DNA Extraction Kit 0.04455\n",
      "55      D6300 0.042 96 MagBead DNA Extraction Kit 0.03780\n",
      "56      D6300 0.042 96 MagBead DNA Extraction Kit 0.03420\n",
      "57      D6300 0.155 96 MagBead DNA Extraction Kit 0.11280\n",
      "58      D6300 0.155 96 MagBead DNA Extraction Kit 0.14050\n",
      "59      D6300 0.155 96 MagBead DNA Extraction Kit 0.14685\n",
      "60      D6300 0.155 96 MagBead DNA Extraction Kit 0.11525\n",
      "61      D6300 0.155 96 MagBead DNA Extraction Kit 0.09530\n",
      "62      D6300 0.155 96 MagBead DNA Extraction Kit 0.10950\n",
      "63      D6300 0.155 96 MagBead DNA Extraction Kit 0.12080\n",
      "64      D6300 0.155 96 MagBead DNA Extraction Kit 0.10120\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Sample_name                                          taxa.id Subject.ID\n",
      "1     SC304093              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "2     SC304926              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "3     SC249383              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "4     SC249388              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "5     SC304093      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "6     SC304926      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "7     SC249383      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "8     SC249388      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "9     SC304093             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "10    SC304926             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "11    SC249383             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "12    SC249388             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "13    SC304093           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "14    SC304926           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "15    SC249383           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "16    SC249388           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "17    SC304093           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "18    SC304926           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "19    SC249383           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "20    SC249388           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "21    SC304093 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "22    SC304926 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "23    SC249383 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "24    SC249388 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "25    SC304093            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "26    SC304926            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "27    SC249383            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "28    SC249388            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "29    SC304093        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "30    SC304926        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "31    SC249383        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "32    SC249388        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "33    SC304093          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "34    SC304926          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "35    SC249383          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "36    SC249388          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "37    SC304093            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "38    SC304926            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "39    SC249383            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "40    SC249388            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "41    SC304093       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "42    SC304926       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "43    SC249383       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "44    SC249388       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "45    SC304093          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "46    SC304926          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "47    SC249383          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "48    SC249388          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "49    SC304093            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "50    SC304926            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "51    SC249383            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "52    SC249388            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "53    SC304093          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "54    SC304926          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "55    SC249383          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "56    SC249388          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "57    SC304093      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "58    SC304926      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "59    SC249383      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "60    SC249388      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "61    SC304093  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "62    SC304926  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "63    SC249383  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "64    SC249388  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "65    SC304093                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "66    SC304926                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "67    SC249383                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "68    SC249388                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "69    SC304093         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "70    SC304926         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "71    SC249383         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "72    SC249388         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "73    SC304093           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "74    SC304926           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "75    SC249383           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "76    SC249388           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "          exp                         bench     obs\n",
      "1  0.06263920 96 MagBead DNA Extraction Kit 0.09170\n",
      "2  0.06263920 96 MagBead DNA Extraction Kit 0.08840\n",
      "3  0.06263920 96 MagBead DNA Extraction Kit 0.11700\n",
      "4  0.06263920 96 MagBead DNA Extraction Kit 0.10585\n",
      "5  0.06263920 96 MagBead DNA Extraction Kit 0.01890\n",
      "6  0.06263920 96 MagBead DNA Extraction Kit 0.01675\n",
      "7  0.06263920 96 MagBead DNA Extraction Kit 0.00395\n",
      "8  0.06263920 96 MagBead DNA Extraction Kit 0.00635\n",
      "9  0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "10 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "11 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "12 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "13 0.06263920 96 MagBead DNA Extraction Kit 0.00030\n",
      "14 0.06263920 96 MagBead DNA Extraction Kit 0.00115\n",
      "15 0.06263920 96 MagBead DNA Extraction Kit 0.00015\n",
      "16 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "17 0.02352450 96 MagBead DNA Extraction Kit 0.02505\n",
      "18 0.02352450 96 MagBead DNA Extraction Kit 0.03850\n",
      "19 0.02352450 96 MagBead DNA Extraction Kit 0.02510\n",
      "20 0.02352450 96 MagBead DNA Extraction Kit 0.02335\n",
      "21 0.06263920 96 MagBead DNA Extraction Kit 0.12500\n",
      "22 0.06263920 96 MagBead DNA Extraction Kit 0.11570\n",
      "23 0.06263920 96 MagBead DNA Extraction Kit 0.13235\n",
      "24 0.06263920 96 MagBead DNA Extraction Kit 0.16725\n",
      "25 0.06263920 96 MagBead DNA Extraction Kit 0.02585\n",
      "26 0.06263920 96 MagBead DNA Extraction Kit 0.03725\n",
      "27 0.06263920 96 MagBead DNA Extraction Kit 0.01270\n",
      "28 0.06263920 96 MagBead DNA Extraction Kit 0.01415\n",
      "29 0.03660913 96 MagBead DNA Extraction Kit 0.00000\n",
      "30 0.03660913 96 MagBead DNA Extraction Kit 0.00000\n",
      "31 0.03660913 96 MagBead DNA Extraction Kit 0.00000\n",
      "32 0.03660913 96 MagBead DNA Extraction Kit 0.00000\n",
      "33 0.12527840 96 MagBead DNA Extraction Kit 0.23090\n",
      "34 0.12527840 96 MagBead DNA Extraction Kit 0.18855\n",
      "35 0.12527840 96 MagBead DNA Extraction Kit 0.20130\n",
      "36 0.12527840 96 MagBead DNA Extraction Kit 0.19155\n",
      "37 0.06263920 96 MagBead DNA Extraction Kit 0.01815\n",
      "38 0.06263920 96 MagBead DNA Extraction Kit 0.01655\n",
      "39 0.06263920 96 MagBead DNA Extraction Kit 0.00960\n",
      "40 0.06263920 96 MagBead DNA Extraction Kit 0.00965\n",
      "41 0.06263920 96 MagBead DNA Extraction Kit 0.11490\n",
      "42 0.06263920 96 MagBead DNA Extraction Kit 0.10760\n",
      "43 0.06263920 96 MagBead DNA Extraction Kit 0.15045\n",
      "44 0.06263920 96 MagBead DNA Extraction Kit 0.13105\n",
      "45 0.03660913 96 MagBead DNA Extraction Kit 0.05020\n",
      "46 0.03660913 96 MagBead DNA Extraction Kit 0.04780\n",
      "47 0.03660913 96 MagBead DNA Extraction Kit 0.01830\n",
      "48 0.03660913 96 MagBead DNA Extraction Kit 0.01440\n",
      "49 0.04175947 96 MagBead DNA Extraction Kit 0.01610\n",
      "50 0.04175947 96 MagBead DNA Extraction Kit 0.04590\n",
      "51 0.04175947 96 MagBead DNA Extraction Kit 0.01750\n",
      "52 0.04175947 96 MagBead DNA Extraction Kit 0.01910\n",
      "53 0.03660913 96 MagBead DNA Extraction Kit 0.05730\n",
      "54 0.03660913 96 MagBead DNA Extraction Kit 0.03560\n",
      "55 0.03660913 96 MagBead DNA Extraction Kit 0.03760\n",
      "56 0.03660913 96 MagBead DNA Extraction Kit 0.03850\n",
      "57 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "58 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "59 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "60 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "61 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "62 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "63 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "64 0.06263920 96 MagBead DNA Extraction Kit 0.00000\n",
      "65 0.03660913 96 MagBead DNA Extraction Kit 0.00000\n",
      "66 0.03660913 96 MagBead DNA Extraction Kit 0.00000\n",
      "67 0.03660913 96 MagBead DNA Extraction Kit 0.00000\n",
      "68 0.03660913 96 MagBead DNA Extraction Kit 0.00000\n",
      "69 0.01572940 96 MagBead DNA Extraction Kit 0.00890\n",
      "70 0.01572940 96 MagBead DNA Extraction Kit 0.00875\n",
      "71 0.01572940 96 MagBead DNA Extraction Kit 0.01035\n",
      "72 0.01572940 96 MagBead DNA Extraction Kit 0.00780\n",
      "73 0.02087973 96 MagBead DNA Extraction Kit 0.01090\n",
      "74 0.02087973 96 MagBead DNA Extraction Kit 0.01240\n",
      "75 0.02087973 96 MagBead DNA Extraction Kit 0.00465\n",
      "76 0.02087973 96 MagBead DNA Extraction Kit 0.00400\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 8 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 8 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 8 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 20 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 20 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 20 rows containing missing values (geom_point).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 2 rows containing missing values (geom_segment).\"\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Sample_name                                          taxa.id\n",
      "1  SC249408-PC04925-E-02                    D_4__BacillaceaeD_5__Bacillus\n",
      "2               SC552959                    D_4__BacillaceaeD_5__Bacillus\n",
      "3               SC249408                    D_4__BacillaceaeD_5__Bacillus\n",
      "4  SC249420-PC07578-H-07                    D_4__BacillaceaeD_5__Bacillus\n",
      "5  SC249408-PC04924-C-05                    D_4__BacillaceaeD_5__Bacillus\n",
      "6               SC249412                    D_4__BacillaceaeD_5__Bacillus\n",
      "7  SC249420-PC04924-G-06                    D_4__BacillaceaeD_5__Bacillus\n",
      "8               SC249425                    D_4__BacillaceaeD_5__Bacillus\n",
      "9  SC249408-PC04925-E-02 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "10              SC552959 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "11              SC249408 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "12 SC249420-PC07578-H-07 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "13 SC249408-PC04924-C-05 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "14              SC249412 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "15 SC249420-PC04924-G-06 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "16              SC249425 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "17 SC249408-PC04925-E-02           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "18              SC552959           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "19              SC249408           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "20 SC249420-PC07578-H-07           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "21 SC249408-PC04924-C-05           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "22              SC249412           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "23 SC249420-PC04924-G-06           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "24              SC249425           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "25 SC249408-PC04925-E-02            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "26              SC552959            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "27              SC249408            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "28 SC249420-PC07578-H-07            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "29 SC249408-PC04924-C-05            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "30              SC249412            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "31 SC249420-PC04924-G-06            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "32              SC249425            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "33 SC249408-PC04925-E-02          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "34              SC552959          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "35              SC249408          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "36 SC249420-PC07578-H-07          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "37 SC249408-PC04924-C-05          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "38              SC249412          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "39 SC249420-PC04924-G-06          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "40              SC249425          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "41 SC249408-PC04925-E-02                   D_4__ListeriaceaeD_5__Listeria\n",
      "42              SC552959                   D_4__ListeriaceaeD_5__Listeria\n",
      "43              SC249408                   D_4__ListeriaceaeD_5__Listeria\n",
      "44 SC249420-PC07578-H-07                   D_4__ListeriaceaeD_5__Listeria\n",
      "45 SC249408-PC04924-C-05                   D_4__ListeriaceaeD_5__Listeria\n",
      "46              SC249412                   D_4__ListeriaceaeD_5__Listeria\n",
      "47 SC249420-PC04924-G-06                   D_4__ListeriaceaeD_5__Listeria\n",
      "48              SC249425                   D_4__ListeriaceaeD_5__Listeria\n",
      "49 SC249408-PC04925-E-02            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "50              SC552959            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "51              SC249408            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "52 SC249420-PC07578-H-07            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "53 SC249408-PC04924-C-05            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "54              SC249412            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "55 SC249420-PC04924-G-06            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "56              SC249425            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "57 SC249408-PC04925-E-02        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "58              SC552959        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "59              SC249408        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "60 SC249420-PC07578-H-07        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "61 SC249408-PC04924-C-05        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "62              SC249412        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "63 SC249420-PC04924-G-06        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "64              SC249425        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "   Subject.ID   exp                        bench     obs\n",
      "1       D6300 0.174 MagAttract PowerSoil DNA Kit 0.08035\n",
      "2       D6300 0.174 MagAttract PowerSoil DNA Kit 0.01090\n",
      "3       D6300 0.174 MagAttract PowerSoil DNA Kit 0.07865\n",
      "4       D6300 0.174 MagAttract PowerSoil DNA Kit 0.09355\n",
      "5       D6300 0.174 MagAttract PowerSoil DNA Kit 0.08695\n",
      "6       D6300 0.174 MagAttract PowerSoil DNA Kit 0.06950\n",
      "7       D6300 0.174 MagAttract PowerSoil DNA Kit 0.08845\n",
      "8       D6300 0.174 MagAttract PowerSoil DNA Kit 0.08380\n",
      "9       D6300 0.101 MagAttract PowerSoil DNA Kit 0.33105\n",
      "10      D6300 0.101 MagAttract PowerSoil DNA Kit 0.00075\n",
      "11      D6300 0.101 MagAttract PowerSoil DNA Kit 0.29995\n",
      "12      D6300 0.101 MagAttract PowerSoil DNA Kit 0.26780\n",
      "13      D6300 0.101 MagAttract PowerSoil DNA Kit 0.30785\n",
      "14      D6300 0.101 MagAttract PowerSoil DNA Kit 0.27755\n",
      "15      D6300 0.101 MagAttract PowerSoil DNA Kit 0.32880\n",
      "16      D6300 0.101 MagAttract PowerSoil DNA Kit 0.31235\n",
      "17      D6300 0.104 MagAttract PowerSoil DNA Kit 0.00000\n",
      "18      D6300 0.104 MagAttract PowerSoil DNA Kit 0.00000\n",
      "19      D6300 0.104 MagAttract PowerSoil DNA Kit 0.00000\n",
      "20      D6300 0.104 MagAttract PowerSoil DNA Kit 0.00000\n",
      "21      D6300 0.104 MagAttract PowerSoil DNA Kit 0.00000\n",
      "22      D6300 0.104 MagAttract PowerSoil DNA Kit 0.00000\n",
      "23      D6300 0.104 MagAttract PowerSoil DNA Kit 0.00000\n",
      "24      D6300 0.104 MagAttract PowerSoil DNA Kit 0.00000\n",
      "25      D6300 0.099 MagAttract PowerSoil DNA Kit 0.04635\n",
      "26      D6300 0.099 MagAttract PowerSoil DNA Kit 0.00000\n",
      "27      D6300 0.099 MagAttract PowerSoil DNA Kit 0.04620\n",
      "28      D6300 0.099 MagAttract PowerSoil DNA Kit 0.05710\n",
      "29      D6300 0.099 MagAttract PowerSoil DNA Kit 0.04930\n",
      "30      D6300 0.099 MagAttract PowerSoil DNA Kit 0.04985\n",
      "31      D6300 0.099 MagAttract PowerSoil DNA Kit 0.03860\n",
      "32      D6300 0.099 MagAttract PowerSoil DNA Kit 0.04840\n",
      "33      D6300 0.184 MagAttract PowerSoil DNA Kit 0.21555\n",
      "34      D6300 0.184 MagAttract PowerSoil DNA Kit 0.00000\n",
      "35      D6300 0.184 MagAttract PowerSoil DNA Kit 0.21620\n",
      "36      D6300 0.184 MagAttract PowerSoil DNA Kit 0.25395\n",
      "37      D6300 0.184 MagAttract PowerSoil DNA Kit 0.24690\n",
      "38      D6300 0.184 MagAttract PowerSoil DNA Kit 0.26460\n",
      "39      D6300 0.184 MagAttract PowerSoil DNA Kit 0.26235\n",
      "40      D6300 0.184 MagAttract PowerSoil DNA Kit 0.25095\n",
      "41      D6300 0.141 MagAttract PowerSoil DNA Kit 0.05735\n",
      "42      D6300 0.141 MagAttract PowerSoil DNA Kit 0.94555\n",
      "43      D6300 0.141 MagAttract PowerSoil DNA Kit 0.05855\n",
      "44      D6300 0.141 MagAttract PowerSoil DNA Kit 0.08435\n",
      "45      D6300 0.141 MagAttract PowerSoil DNA Kit 0.06990\n",
      "46      D6300 0.141 MagAttract PowerSoil DNA Kit 0.07245\n",
      "47      D6300 0.141 MagAttract PowerSoil DNA Kit 0.08105\n",
      "48      D6300 0.141 MagAttract PowerSoil DNA Kit 0.06650\n",
      "49      D6300 0.042 MagAttract PowerSoil DNA Kit 0.20775\n",
      "50      D6300 0.042 MagAttract PowerSoil DNA Kit 0.03075\n",
      "51      D6300 0.042 MagAttract PowerSoil DNA Kit 0.23675\n",
      "52      D6300 0.042 MagAttract PowerSoil DNA Kit 0.18575\n",
      "53      D6300 0.042 MagAttract PowerSoil DNA Kit 0.18810\n",
      "54      D6300 0.042 MagAttract PowerSoil DNA Kit 0.19265\n",
      "55      D6300 0.042 MagAttract PowerSoil DNA Kit 0.15015\n",
      "56      D6300 0.042 MagAttract PowerSoil DNA Kit 0.18000\n",
      "57      D6300 0.155 MagAttract PowerSoil DNA Kit 0.05675\n",
      "58      D6300 0.155 MagAttract PowerSoil DNA Kit 0.00000\n",
      "59      D6300 0.155 MagAttract PowerSoil DNA Kit 0.06230\n",
      "60      D6300 0.155 MagAttract PowerSoil DNA Kit 0.05030\n",
      "61      D6300 0.155 MagAttract PowerSoil DNA Kit 0.05055\n",
      "62      D6300 0.155 MagAttract PowerSoil DNA Kit 0.05645\n",
      "63      D6300 0.155 MagAttract PowerSoil DNA Kit 0.05045\n",
      "64      D6300 0.155 MagAttract PowerSoil DNA Kit 0.05730\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Sample_name                                        taxa.id Subject.ID\n",
      "1     SC552954                  D_4__BacillaceaeD_5__Bacillus    DZ35316\n",
      "2     SC552954    D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35316\n",
      "3     SC552954       D_4__BrevibacteriaceaeD_5__Stomatococcus    DZ35316\n",
      "4     SC552954     D_4__CampylobacteriaceaeD_5__Campylobacter    DZ35316\n",
      "5     SC552954         D_4__CoriobacteriaceaeD_5__Eggerthella    DZ35316\n",
      "6     SC552954             D_4__CoriobacteriaceaeD_5__Slackia    DZ35316\n",
      "7     SC552954         D_4__EnterobacteriaceaeD_5__Klebsiella    DZ35316\n",
      "8     SC552954          D_4__EubacteriaceaeD_5__Mogibacterium    DZ35316\n",
      "9     SC552954      D_4__FlavobacteriaceaeD_5__Capnocytophaga    DZ35316\n",
      "10    SC552954        D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35316\n",
      "11    SC552954         D_4__LeptotrichiaceaeD_5__Leptotrichia    DZ35316\n",
      "12    SC552954            D_4__LeuconostocaceaeD_5__Weissella    DZ35316\n",
      "13    SC552954               D_4__NeisseriaceaeD_5__Eikenella    DZ35316\n",
      "14    SC552954               D_4__NeisseriaceaeD_5__Neisseria    DZ35316\n",
      "15    SC552954           D_4__PeptoniphilaceaeD_5__Parvimonas    DZ35316\n",
      "16    SC552954        D_4__PorphyromonadaceaeD_5__Barnesiella    DZ35316\n",
      "17    SC552954             D_4__PrevotellaceaeD_5__Prevotella    DZ35316\n",
      "18    SC552954       D_4__StreptococcaceaeD_5__Granulicatella    DZ35316\n",
      "19    SC552954        D_4__StreptococcaceaeD_5__Streptococcus    DZ35316\n",
      "20    SC552954 D_4__Tenericutes.Incertae.Sedis.XID_5__Gemella    DZ35316\n",
      "21    SC552954             D_4__VeillonellaceaeD_5__Dialister    DZ35316\n",
      "22    SC552954           D_4__VeillonellaceaeD_5__Veillonella    DZ35316\n",
      "           exp                        bench     obs\n",
      "1  0.041791461 MagAttract PowerSoil DNA Kit 0.04160\n",
      "2  0.062687191 MagAttract PowerSoil DNA Kit 0.03125\n",
      "3  0.062687191 MagAttract PowerSoil DNA Kit 0.00000\n",
      "4  0.031343595 MagAttract PowerSoil DNA Kit 0.00000\n",
      "5  0.036637180 MagAttract PowerSoil DNA Kit 0.00000\n",
      "6  0.036637180 MagAttract PowerSoil DNA Kit 0.00000\n",
      "7  0.062687191 MagAttract PowerSoil DNA Kit 0.00000\n",
      "8  0.001044787 MagAttract PowerSoil DNA Kit 0.00000\n",
      "9  0.005223933 MagAttract PowerSoil DNA Kit 0.00270\n",
      "10 0.062687191 MagAttract PowerSoil DNA Kit 0.07955\n",
      "11 0.052239326 MagAttract PowerSoil DNA Kit 0.07180\n",
      "12 0.052239326 MagAttract PowerSoil DNA Kit 0.01675\n",
      "13 0.041791461 MagAttract PowerSoil DNA Kit 0.05995\n",
      "14 0.062687191 MagAttract PowerSoil DNA Kit 0.14720\n",
      "15 0.031343595 MagAttract PowerSoil DNA Kit 0.00000\n",
      "16 0.062687191 MagAttract PowerSoil DNA Kit 0.00000\n",
      "17 0.062687191 MagAttract PowerSoil DNA Kit 0.07450\n",
      "18 0.062687191 MagAttract PowerSoil DNA Kit 0.00000\n",
      "19 0.062687191 MagAttract PowerSoil DNA Kit 0.01720\n",
      "20 0.052239326 MagAttract PowerSoil DNA Kit 0.00000\n",
      "21 0.001044787 MagAttract PowerSoil DNA Kit 0.00000\n",
      "22 0.052239326 MagAttract PowerSoil DNA Kit 0.06785\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Sample_name                                          taxa.id Subject.ID\n",
      "1    SC552960                    D_4__BacillaceaeD_5__Bacillus      D6310\n",
      "2    SC552960 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella      D6310\n",
      "3    SC552960           D_4__EnterobacteriaceaeD_5__Salmonella      D6310\n",
      "4    SC552960            D_4__EnterococcaceaeD_5__Enterococcus      D6310\n",
      "5    SC552960          D_4__LactobacillaceaeD_5__Lactobacillus      D6310\n",
      "6    SC552960                   D_4__ListeriaceaeD_5__Listeria      D6310\n",
      "7    SC552960            D_4__PseudomonadaceaeD_5__Pseudomonas      D6310\n",
      "8    SC552960        D_4__StaphylococcaceaeD_5__Staphylococcus      D6310\n",
      "       exp                        bench     obs\n",
      "1 1.20e-02 MagAttract PowerSoil DNA Kit 0.10960\n",
      "2 6.90e-04 MagAttract PowerSoil DNA Kit 0.30915\n",
      "3 7.00e-04 MagAttract PowerSoil DNA Kit 0.00000\n",
      "4 6.70e-06 MagAttract PowerSoil DNA Kit 0.19780\n",
      "5 1.20e-04 MagAttract PowerSoil DNA Kit 0.05805\n",
      "6 9.59e-01 MagAttract PowerSoil DNA Kit 0.04395\n",
      "7 2.80e-02 MagAttract PowerSoil DNA Kit 0.14260\n",
      "8 1.00e-06 MagAttract PowerSoil DNA Kit 0.13865\n",
      "   Sample_name                                          taxa.id Subject.ID\n",
      "1     SC249409              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "2     SC249413              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "3     SC249427              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "4     SC249428              D_4__BacteroidaceaeD_5__Bacteroides    DZ35322\n",
      "5     SC249409      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "6     SC249413      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "7     SC249427      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "8     SC249428      D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35322\n",
      "9     SC249409             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "10    SC249413             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "11    SC249427             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "12    SC249428             D_4__BukerholderiaceaeD_5__Ralstonia    DZ35322\n",
      "13    SC249409           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "14    SC249413           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "15    SC249427           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "16    SC249428           D_4__CoriobacteriaceaeD_5__Collinsella    DZ35322\n",
      "17    SC249409           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "18    SC249413           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "19    SC249427           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "20    SC249428           D_4__DesulfovibrionaceaeD_5__Bilophila    DZ35322\n",
      "21    SC249409 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "22    SC249413 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "23    SC249427 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "24    SC249428 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella    DZ35322\n",
      "25    SC249409            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "26    SC249413            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "27    SC249427            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "28    SC249428            D_4__EnterococcaceaeD_5__Enterococcus    DZ35322\n",
      "29    SC249409        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "30    SC249413        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "31    SC249427        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "32    SC249428        D_4__ErysipelotrichidaeD_5__Coprobacillus    DZ35322\n",
      "33    SC249409          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "34    SC249413          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "35    SC249427          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "36    SC249428          D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35322\n",
      "37    SC249409            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "38    SC249413            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "39    SC249427            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "40    SC249428            D_4__LachnospiraceaeD_5__Anaerostipes    DZ35322\n",
      "41    SC249409       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "42    SC249413       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "43    SC249427       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "44    SC249428       D_4__LachnospiraceaeD_5__Lachnoclostridium    DZ35322\n",
      "45    SC249409          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "46    SC249413          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "47    SC249427          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "48    SC249428          D_4__LactobacillaceaeD_5__Lactobacillus    DZ35322\n",
      "49    SC249409            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "50    SC249413            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "51    SC249427            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "52    SC249428            D_4__LactobacillaceaeD_5__Pediococcus    DZ35322\n",
      "53    SC249409          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "54    SC249413          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "55    SC249427          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "56    SC249428          D_4__PaenibacillaceaeD_5__Paenibacillus    DZ35322\n",
      "57    SC249409      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "58    SC249413      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "59    SC249427      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "60    SC249428      D_4__PorphyromonadaceaeD_5__Parabacteroides    DZ35322\n",
      "61    SC249409  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "62    SC249413  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "63    SC249427  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "64    SC249428  D_4__PropionibacteriaceaeD_5__Propionibacterium    DZ35322\n",
      "65    SC249409                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "66    SC249413                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "67    SC249427                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "68    SC249428                D_4__RikenellaceaeD_5__Allistipes    DZ35322\n",
      "69    SC249409         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "70    SC249413         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "71    SC249427         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "72    SC249428         D_4__RuminococcaceaeD_5__Subdoligranulum    DZ35322\n",
      "73    SC249409           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "74    SC249413           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "75    SC249427           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "76    SC249428           D_4__SynergistaceaeD_5__Pyramidobacter    DZ35322\n",
      "          exp                        bench     obs\n",
      "1  0.06263920 MagAttract PowerSoil DNA Kit 0.10280\n",
      "2  0.06263920 MagAttract PowerSoil DNA Kit 0.09670\n",
      "3  0.06263920 MagAttract PowerSoil DNA Kit 0.10105\n",
      "4  0.06263920 MagAttract PowerSoil DNA Kit 0.10810\n",
      "5  0.06263920 MagAttract PowerSoil DNA Kit 0.00360\n",
      "6  0.06263920 MagAttract PowerSoil DNA Kit 0.00475\n",
      "7  0.06263920 MagAttract PowerSoil DNA Kit 0.00515\n",
      "8  0.06263920 MagAttract PowerSoil DNA Kit 0.00530\n",
      "9  0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "10 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "11 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "12 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "13 0.06263920 MagAttract PowerSoil DNA Kit 0.00065\n",
      "14 0.06263920 MagAttract PowerSoil DNA Kit 0.00070\n",
      "15 0.06263920 MagAttract PowerSoil DNA Kit 0.00015\n",
      "16 0.06263920 MagAttract PowerSoil DNA Kit 0.00045\n",
      "17 0.02352450 MagAttract PowerSoil DNA Kit 0.03710\n",
      "18 0.02352450 MagAttract PowerSoil DNA Kit 0.04275\n",
      "19 0.02352450 MagAttract PowerSoil DNA Kit 0.04485\n",
      "20 0.02352450 MagAttract PowerSoil DNA Kit 0.04220\n",
      "21 0.06263920 MagAttract PowerSoil DNA Kit 0.14355\n",
      "22 0.06263920 MagAttract PowerSoil DNA Kit 0.13610\n",
      "23 0.06263920 MagAttract PowerSoil DNA Kit 0.13630\n",
      "24 0.06263920 MagAttract PowerSoil DNA Kit 0.12330\n",
      "25 0.06263920 MagAttract PowerSoil DNA Kit 0.00885\n",
      "26 0.06263920 MagAttract PowerSoil DNA Kit 0.00590\n",
      "27 0.06263920 MagAttract PowerSoil DNA Kit 0.00655\n",
      "28 0.06263920 MagAttract PowerSoil DNA Kit 0.00550\n",
      "29 0.03660913 MagAttract PowerSoil DNA Kit 0.00000\n",
      "30 0.03660913 MagAttract PowerSoil DNA Kit 0.00000\n",
      "31 0.03660913 MagAttract PowerSoil DNA Kit 0.00000\n",
      "32 0.03660913 MagAttract PowerSoil DNA Kit 0.00000\n",
      "33 0.12527840 MagAttract PowerSoil DNA Kit 0.19595\n",
      "34 0.12527840 MagAttract PowerSoil DNA Kit 0.21170\n",
      "35 0.12527840 MagAttract PowerSoil DNA Kit 0.21430\n",
      "36 0.12527840 MagAttract PowerSoil DNA Kit 0.23515\n",
      "37 0.06263920 MagAttract PowerSoil DNA Kit 0.02380\n",
      "38 0.06263920 MagAttract PowerSoil DNA Kit 0.02615\n",
      "39 0.06263920 MagAttract PowerSoil DNA Kit 0.02215\n",
      "40 0.06263920 MagAttract PowerSoil DNA Kit 0.02320\n",
      "41 0.06263920 MagAttract PowerSoil DNA Kit 0.12130\n",
      "42 0.06263920 MagAttract PowerSoil DNA Kit 0.09065\n",
      "43 0.06263920 MagAttract PowerSoil DNA Kit 0.09510\n",
      "44 0.06263920 MagAttract PowerSoil DNA Kit 0.08680\n",
      "45 0.03660913 MagAttract PowerSoil DNA Kit 0.01405\n",
      "46 0.03660913 MagAttract PowerSoil DNA Kit 0.03600\n",
      "47 0.03660913 MagAttract PowerSoil DNA Kit 0.03650\n",
      "48 0.03660913 MagAttract PowerSoil DNA Kit 0.02380\n",
      "49 0.04175947 MagAttract PowerSoil DNA Kit 0.01020\n",
      "50 0.04175947 MagAttract PowerSoil DNA Kit 0.00750\n",
      "51 0.04175947 MagAttract PowerSoil DNA Kit 0.00605\n",
      "52 0.04175947 MagAttract PowerSoil DNA Kit 0.00950\n",
      "53 0.03660913 MagAttract PowerSoil DNA Kit 0.03370\n",
      "54 0.03660913 MagAttract PowerSoil DNA Kit 0.04475\n",
      "55 0.03660913 MagAttract PowerSoil DNA Kit 0.04145\n",
      "56 0.03660913 MagAttract PowerSoil DNA Kit 0.03220\n",
      "57 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "58 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "59 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "60 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "61 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "62 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "63 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "64 0.06263920 MagAttract PowerSoil DNA Kit 0.00000\n",
      "65 0.03660913 MagAttract PowerSoil DNA Kit 0.00000\n",
      "66 0.03660913 MagAttract PowerSoil DNA Kit 0.00000\n",
      "67 0.03660913 MagAttract PowerSoil DNA Kit 0.00000\n",
      "68 0.03660913 MagAttract PowerSoil DNA Kit 0.00000\n",
      "69 0.01572940 MagAttract PowerSoil DNA Kit 0.01415\n",
      "70 0.01572940 MagAttract PowerSoil DNA Kit 0.01275\n",
      "71 0.01572940 MagAttract PowerSoil DNA Kit 0.01205\n",
      "72 0.01572940 MagAttract PowerSoil DNA Kit 0.01180\n",
      "73 0.02087973 MagAttract PowerSoil DNA Kit 0.00390\n",
      "74 0.02087973 MagAttract PowerSoil DNA Kit 0.00785\n",
      "75 0.02087973 MagAttract PowerSoil DNA Kit 0.00775\n",
      "76 0.02087973 MagAttract PowerSoil DNA Kit 0.00610\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 8 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 8 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 8 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 20 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 20 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 20 rows containing missing values (geom_point).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Sample_name                                          taxa.id\n",
      "1  SC249432-PC04925-B-04                    D_4__BacillaceaeD_5__Bacillus\n",
      "2  SC249437-PC04925-F-04                    D_4__BacillaceaeD_5__Bacillus\n",
      "3               SC249437                    D_4__BacillaceaeD_5__Bacillus\n",
      "4  SC249451-PC07578-F-06                    D_4__BacillaceaeD_5__Bacillus\n",
      "5  SC249432-PC04924-A-08                    D_4__BacillaceaeD_5__Bacillus\n",
      "6  SC249437-PC04924-F-08                    D_4__BacillaceaeD_5__Bacillus\n",
      "7               SC249442                    D_4__BacillaceaeD_5__Bacillus\n",
      "8  SC249451-PC04924-C-10                    D_4__BacillaceaeD_5__Bacillus\n",
      "9  SC249432-PC04925-B-04 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "10 SC249437-PC04925-F-04 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "11              SC249437 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "12 SC249451-PC07578-F-06 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "13 SC249432-PC04924-A-08 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "14 SC249437-PC04924-F-08 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "15              SC249442 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "16 SC249451-PC04924-C-10 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "17 SC249432-PC04925-B-04           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "18 SC249437-PC04925-F-04           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "19              SC249437           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "20 SC249451-PC07578-F-06           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "21 SC249432-PC04924-A-08           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "22 SC249437-PC04924-F-08           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "23              SC249442           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "24 SC249451-PC04924-C-10           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "25 SC249432-PC04925-B-04            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "26 SC249437-PC04925-F-04            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "27              SC249437            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "28 SC249451-PC07578-F-06            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "29 SC249432-PC04924-A-08            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "30 SC249437-PC04924-F-08            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "31              SC249442            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "32 SC249451-PC04924-C-10            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "33 SC249432-PC04925-B-04          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "34 SC249437-PC04925-F-04          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "35              SC249437          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "36 SC249451-PC07578-F-06          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "37 SC249432-PC04924-A-08          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "38 SC249437-PC04924-F-08          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "39              SC249442          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "40 SC249451-PC04924-C-10          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "41 SC249432-PC04925-B-04                   D_4__ListeriaceaeD_5__Listeria\n",
      "42 SC249437-PC04925-F-04                   D_4__ListeriaceaeD_5__Listeria\n",
      "43              SC249437                   D_4__ListeriaceaeD_5__Listeria\n",
      "44 SC249451-PC07578-F-06                   D_4__ListeriaceaeD_5__Listeria\n",
      "45 SC249432-PC04924-A-08                   D_4__ListeriaceaeD_5__Listeria\n",
      "46 SC249437-PC04924-F-08                   D_4__ListeriaceaeD_5__Listeria\n",
      "47              SC249442                   D_4__ListeriaceaeD_5__Listeria\n",
      "48 SC249451-PC04924-C-10                   D_4__ListeriaceaeD_5__Listeria\n",
      "49 SC249432-PC04925-B-04            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "50 SC249437-PC04925-F-04            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "51              SC249437            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "52 SC249451-PC07578-F-06            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "53 SC249432-PC04924-A-08            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "54 SC249437-PC04924-F-08            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "55              SC249442            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "56 SC249451-PC04924-C-10            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "57 SC249432-PC04925-B-04        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "58 SC249437-PC04925-F-04        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "59              SC249437        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "60 SC249451-PC07578-F-06        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "61 SC249432-PC04924-A-08        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "62 SC249437-PC04924-F-08        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "63              SC249442        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "64 SC249451-PC04924-C-10        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "   Subject.ID   exp                          bench     obs\n",
      "1       D6300 0.174 MagAttract PowerMicrobiome Kit 0.21485\n",
      "2       D6300 0.174 MagAttract PowerMicrobiome Kit 0.18315\n",
      "3       D6300 0.174 MagAttract PowerMicrobiome Kit 0.15515\n",
      "4       D6300 0.174 MagAttract PowerMicrobiome Kit 0.24425\n",
      "5       D6300 0.174 MagAttract PowerMicrobiome Kit 0.14770\n",
      "6       D6300 0.174 MagAttract PowerMicrobiome Kit 0.20325\n",
      "7       D6300 0.174 MagAttract PowerMicrobiome Kit 0.22570\n",
      "8       D6300 0.174 MagAttract PowerMicrobiome Kit 0.21085\n",
      "9       D6300 0.101 MagAttract PowerMicrobiome Kit 0.12580\n",
      "10      D6300 0.101 MagAttract PowerMicrobiome Kit 0.13640\n",
      "11      D6300 0.101 MagAttract PowerMicrobiome Kit 0.12770\n",
      "12      D6300 0.101 MagAttract PowerMicrobiome Kit 0.11560\n",
      "13      D6300 0.101 MagAttract PowerMicrobiome Kit 0.12130\n",
      "14      D6300 0.101 MagAttract PowerMicrobiome Kit 0.10355\n",
      "15      D6300 0.101 MagAttract PowerMicrobiome Kit 0.14135\n",
      "16      D6300 0.101 MagAttract PowerMicrobiome Kit 0.13535\n",
      "17      D6300 0.104 MagAttract PowerMicrobiome Kit 0.00000\n",
      "18      D6300 0.104 MagAttract PowerMicrobiome Kit 0.00000\n",
      "19      D6300 0.104 MagAttract PowerMicrobiome Kit 0.00000\n",
      "20      D6300 0.104 MagAttract PowerMicrobiome Kit 0.00000\n",
      "21      D6300 0.104 MagAttract PowerMicrobiome Kit 0.00000\n",
      "22      D6300 0.104 MagAttract PowerMicrobiome Kit 0.00000\n",
      "23      D6300 0.104 MagAttract PowerMicrobiome Kit 0.00000\n",
      "24      D6300 0.104 MagAttract PowerMicrobiome Kit 0.00000\n",
      "25      D6300 0.099 MagAttract PowerMicrobiome Kit 0.09300\n",
      "26      D6300 0.099 MagAttract PowerMicrobiome Kit 0.10400\n",
      "27      D6300 0.099 MagAttract PowerMicrobiome Kit 0.10490\n",
      "28      D6300 0.099 MagAttract PowerMicrobiome Kit 0.13375\n",
      "29      D6300 0.099 MagAttract PowerMicrobiome Kit 0.11160\n",
      "30      D6300 0.099 MagAttract PowerMicrobiome Kit 0.11930\n",
      "31      D6300 0.099 MagAttract PowerMicrobiome Kit 0.12625\n",
      "32      D6300 0.099 MagAttract PowerMicrobiome Kit 0.12890\n",
      "33      D6300 0.184 MagAttract PowerMicrobiome Kit 0.07065\n",
      "34      D6300 0.184 MagAttract PowerMicrobiome Kit 0.06365\n",
      "35      D6300 0.184 MagAttract PowerMicrobiome Kit 0.07225\n",
      "36      D6300 0.184 MagAttract PowerMicrobiome Kit 0.06050\n",
      "37      D6300 0.184 MagAttract PowerMicrobiome Kit 0.06660\n",
      "38      D6300 0.184 MagAttract PowerMicrobiome Kit 0.05800\n",
      "39      D6300 0.184 MagAttract PowerMicrobiome Kit 0.05320\n",
      "40      D6300 0.184 MagAttract PowerMicrobiome Kit 0.05410\n",
      "41      D6300 0.141 MagAttract PowerMicrobiome Kit 0.21900\n",
      "42      D6300 0.141 MagAttract PowerMicrobiome Kit 0.16690\n",
      "43      D6300 0.141 MagAttract PowerMicrobiome Kit 0.17765\n",
      "44      D6300 0.141 MagAttract PowerMicrobiome Kit 0.16610\n",
      "45      D6300 0.141 MagAttract PowerMicrobiome Kit 0.23020\n",
      "46      D6300 0.141 MagAttract PowerMicrobiome Kit 0.21375\n",
      "47      D6300 0.141 MagAttract PowerMicrobiome Kit 0.17200\n",
      "48      D6300 0.141 MagAttract PowerMicrobiome Kit 0.18550\n",
      "49      D6300 0.042 MagAttract PowerMicrobiome Kit 0.06700\n",
      "50      D6300 0.042 MagAttract PowerMicrobiome Kit 0.07045\n",
      "51      D6300 0.042 MagAttract PowerMicrobiome Kit 0.07585\n",
      "52      D6300 0.042 MagAttract PowerMicrobiome Kit 0.09815\n",
      "53      D6300 0.042 MagAttract PowerMicrobiome Kit 0.08035\n",
      "54      D6300 0.042 MagAttract PowerMicrobiome Kit 0.06515\n",
      "55      D6300 0.042 MagAttract PowerMicrobiome Kit 0.09100\n",
      "56      D6300 0.042 MagAttract PowerMicrobiome Kit 0.09275\n",
      "57      D6300 0.155 MagAttract PowerMicrobiome Kit 0.20545\n",
      "58      D6300 0.155 MagAttract PowerMicrobiome Kit 0.27475\n",
      "59      D6300 0.155 MagAttract PowerMicrobiome Kit 0.28640\n",
      "60      D6300 0.155 MagAttract PowerMicrobiome Kit 0.17365\n",
      "61      D6300 0.155 MagAttract PowerMicrobiome Kit 0.23780\n",
      "62      D6300 0.155 MagAttract PowerMicrobiome Kit 0.23700\n",
      "63      D6300 0.155 MagAttract PowerMicrobiome Kit 0.18605\n",
      "64      D6300 0.155 MagAttract PowerMicrobiome Kit 0.18990\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Sample_name                                          taxa.id Subject.ID\n",
      "1    SC552974                    D_4__BacillaceaeD_5__Bacillus      D6310\n",
      "2    SC552974 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella      D6310\n",
      "3    SC552974           D_4__EnterobacteriaceaeD_5__Salmonella      D6310\n",
      "4    SC552974            D_4__EnterococcaceaeD_5__Enterococcus      D6310\n",
      "5    SC552974          D_4__LactobacillaceaeD_5__Lactobacillus      D6310\n",
      "6    SC552974                   D_4__ListeriaceaeD_5__Listeria      D6310\n",
      "7    SC552974            D_4__PseudomonadaceaeD_5__Pseudomonas      D6310\n",
      "8    SC552974        D_4__StaphylococcaceaeD_5__Staphylococcus      D6310\n",
      "       exp                          bench     obs\n",
      "1 1.20e-02 MagAttract PowerMicrobiome Kit 0.09395\n",
      "2 6.90e-04 MagAttract PowerMicrobiome Kit 0.35470\n",
      "3 7.00e-04 MagAttract PowerMicrobiome Kit 0.00000\n",
      "4 6.70e-06 MagAttract PowerMicrobiome Kit 0.19650\n",
      "5 1.20e-04 MagAttract PowerMicrobiome Kit 0.06145\n",
      "6 9.59e-01 MagAttract PowerMicrobiome Kit 0.05395\n",
      "7 2.80e-02 MagAttract PowerMicrobiome Kit 0.08780\n",
      "8 1.00e-06 MagAttract PowerMicrobiome Kit 0.15165\n",
      "              Sample_name                                          taxa.id\n",
      "1   SC249439-PC07578-G-07              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "2   SC249443-PC07578-E-05              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "3                SC249434              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "4   SC249439-PC04924-H-08              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "5   SC249443-PC04924-D-09              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "6                SC249446              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "7   SC249439-PC07578-G-07      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "8   SC249443-PC07578-E-05      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "9                SC249434      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "10  SC249439-PC04924-H-08      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "11  SC249443-PC04924-D-09      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "12               SC249446      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "13  SC249439-PC07578-G-07             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "14  SC249443-PC07578-E-05             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "15               SC249434             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "16  SC249439-PC04924-H-08             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "17  SC249443-PC04924-D-09             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "18               SC249446             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "19  SC249439-PC07578-G-07           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "20  SC249443-PC07578-E-05           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "21               SC249434           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "22  SC249439-PC04924-H-08           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "23  SC249443-PC04924-D-09           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "24               SC249446           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "25  SC249439-PC07578-G-07           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "26  SC249443-PC07578-E-05           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "27               SC249434           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "28  SC249439-PC04924-H-08           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "29  SC249443-PC04924-D-09           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "30               SC249446           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "31  SC249439-PC07578-G-07 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "32  SC249443-PC07578-E-05 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "33               SC249434 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "34  SC249439-PC04924-H-08 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "35  SC249443-PC04924-D-09 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "36               SC249446 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "37  SC249439-PC07578-G-07            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "38  SC249443-PC07578-E-05            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "39               SC249434            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "40  SC249439-PC04924-H-08            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "41  SC249443-PC04924-D-09            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "42               SC249446            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "43  SC249439-PC07578-G-07        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "44  SC249443-PC07578-E-05        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "45               SC249434        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "46  SC249439-PC04924-H-08        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "47  SC249443-PC04924-D-09        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "48               SC249446        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "49  SC249439-PC07578-G-07          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "50  SC249443-PC07578-E-05          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "51               SC249434          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "52  SC249439-PC04924-H-08          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "53  SC249443-PC04924-D-09          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "54               SC249446          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "55  SC249439-PC07578-G-07            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "56  SC249443-PC07578-E-05            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "57               SC249434            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "58  SC249439-PC04924-H-08            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "59  SC249443-PC04924-D-09            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "60               SC249446            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "61  SC249439-PC07578-G-07       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "62  SC249443-PC07578-E-05       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "63               SC249434       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "64  SC249439-PC04924-H-08       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "65  SC249443-PC04924-D-09       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "66               SC249446       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "67  SC249439-PC07578-G-07          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "68  SC249443-PC07578-E-05          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "69               SC249434          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "70  SC249439-PC04924-H-08          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "71  SC249443-PC04924-D-09          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "72               SC249446          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "73  SC249439-PC07578-G-07            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "74  SC249443-PC07578-E-05            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "75               SC249434            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "76  SC249439-PC04924-H-08            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "77  SC249443-PC04924-D-09            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "78               SC249446            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "79  SC249439-PC07578-G-07          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "80  SC249443-PC07578-E-05          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "81               SC249434          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "82  SC249439-PC04924-H-08          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "83  SC249443-PC04924-D-09          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "84               SC249446          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "85  SC249439-PC07578-G-07      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "86  SC249443-PC07578-E-05      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "87               SC249434      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "88  SC249439-PC04924-H-08      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "89  SC249443-PC04924-D-09      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "90               SC249446      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "91  SC249439-PC07578-G-07  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "92  SC249443-PC07578-E-05  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "93               SC249434  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "94  SC249439-PC04924-H-08  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "95  SC249443-PC04924-D-09  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "96               SC249446  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "97  SC249439-PC07578-G-07                D_4__RikenellaceaeD_5__Allistipes\n",
      "98  SC249443-PC07578-E-05                D_4__RikenellaceaeD_5__Allistipes\n",
      "99               SC249434                D_4__RikenellaceaeD_5__Allistipes\n",
      "100 SC249439-PC04924-H-08                D_4__RikenellaceaeD_5__Allistipes\n",
      "101 SC249443-PC04924-D-09                D_4__RikenellaceaeD_5__Allistipes\n",
      "102              SC249446                D_4__RikenellaceaeD_5__Allistipes\n",
      "103 SC249439-PC07578-G-07         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "104 SC249443-PC07578-E-05         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "105              SC249434         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "106 SC249439-PC04924-H-08         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "107 SC249443-PC04924-D-09         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "108              SC249446         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "109 SC249439-PC07578-G-07           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "110 SC249443-PC07578-E-05           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "111              SC249434           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "112 SC249439-PC04924-H-08           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "113 SC249443-PC04924-D-09           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "114              SC249446           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "    Subject.ID        exp                          bench     obs\n",
      "1      DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.05600\n",
      "2      DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.05635\n",
      "3      DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.07655\n",
      "4      DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.06585\n",
      "5      DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.05120\n",
      "6      DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.05280\n",
      "7      DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.01900\n",
      "8      DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.02700\n",
      "9      DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.02040\n",
      "10     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.01650\n",
      "11     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.02470\n",
      "12     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00715\n",
      "13     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "14     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "15     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "16     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "17     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "18     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "19     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00860\n",
      "20     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.01330\n",
      "21     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00485\n",
      "22     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00515\n",
      "23     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.01075\n",
      "24     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00795\n",
      "25     DZ35322 0.02352450 MagAttract PowerMicrobiome Kit 0.03980\n",
      "26     DZ35322 0.02352450 MagAttract PowerMicrobiome Kit 0.03390\n",
      "27     DZ35322 0.02352450 MagAttract PowerMicrobiome Kit 0.05405\n",
      "28     DZ35322 0.02352450 MagAttract PowerMicrobiome Kit 0.03490\n",
      "29     DZ35322 0.02352450 MagAttract PowerMicrobiome Kit 0.03445\n",
      "30     DZ35322 0.02352450 MagAttract PowerMicrobiome Kit 0.02735\n",
      "31     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.06995\n",
      "32     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.08045\n",
      "33     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.09730\n",
      "34     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.07860\n",
      "35     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.09050\n",
      "36     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.07765\n",
      "37     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.04710\n",
      "38     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.10705\n",
      "39     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.05530\n",
      "40     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.05485\n",
      "41     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.09645\n",
      "42     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.07350\n",
      "43     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "44     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "45     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "46     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "47     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "48     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "49     DZ35322 0.12527840 MagAttract PowerMicrobiome Kit 0.15590\n",
      "50     DZ35322 0.12527840 MagAttract PowerMicrobiome Kit 0.13780\n",
      "51     DZ35322 0.12527840 MagAttract PowerMicrobiome Kit 0.18465\n",
      "52     DZ35322 0.12527840 MagAttract PowerMicrobiome Kit 0.13400\n",
      "53     DZ35322 0.12527840 MagAttract PowerMicrobiome Kit 0.11970\n",
      "54     DZ35322 0.12527840 MagAttract PowerMicrobiome Kit 0.08470\n",
      "55     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00790\n",
      "56     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.01645\n",
      "57     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.01190\n",
      "58     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00970\n",
      "59     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.01570\n",
      "60     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.01220\n",
      "61     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.05045\n",
      "62     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.05800\n",
      "63     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.08355\n",
      "64     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.05690\n",
      "65     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.06060\n",
      "66     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.03730\n",
      "67     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.02445\n",
      "68     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.04050\n",
      "69     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.07595\n",
      "70     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.02500\n",
      "71     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.04035\n",
      "72     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00835\n",
      "73     DZ35322 0.04175947 MagAttract PowerMicrobiome Kit 0.06260\n",
      "74     DZ35322 0.04175947 MagAttract PowerMicrobiome Kit 0.08660\n",
      "75     DZ35322 0.04175947 MagAttract PowerMicrobiome Kit 0.07040\n",
      "76     DZ35322 0.04175947 MagAttract PowerMicrobiome Kit 0.06145\n",
      "77     DZ35322 0.04175947 MagAttract PowerMicrobiome Kit 0.09365\n",
      "78     DZ35322 0.04175947 MagAttract PowerMicrobiome Kit 0.05335\n",
      "79     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.01945\n",
      "80     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.01460\n",
      "81     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.02130\n",
      "82     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.01800\n",
      "83     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.01405\n",
      "84     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.01175\n",
      "85     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "86     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "87     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "88     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "89     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "90     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "91     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "92     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "93     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "94     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "95     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "96     DZ35322 0.06263920 MagAttract PowerMicrobiome Kit 0.00000\n",
      "97     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "98     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "99     DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "100    DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "101    DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "102    DZ35322 0.03660913 MagAttract PowerMicrobiome Kit 0.00000\n",
      "103    DZ35322 0.01572940 MagAttract PowerMicrobiome Kit 0.02220\n",
      "104    DZ35322 0.01572940 MagAttract PowerMicrobiome Kit 0.02965\n",
      "105    DZ35322 0.01572940 MagAttract PowerMicrobiome Kit 0.01070\n",
      "106    DZ35322 0.01572940 MagAttract PowerMicrobiome Kit 0.02005\n",
      "107    DZ35322 0.01572940 MagAttract PowerMicrobiome Kit 0.02995\n",
      "108    DZ35322 0.01572940 MagAttract PowerMicrobiome Kit 0.06490\n",
      "109    DZ35322 0.02087973 MagAttract PowerMicrobiome Kit 0.00715\n",
      "110    DZ35322 0.02087973 MagAttract PowerMicrobiome Kit 0.00955\n",
      "111    DZ35322 0.02087973 MagAttract PowerMicrobiome Kit 0.02370\n",
      "112    DZ35322 0.02087973 MagAttract PowerMicrobiome Kit 0.00565\n",
      "113    DZ35322 0.02087973 MagAttract PowerMicrobiome Kit 0.00815\n",
      "114    DZ35322 0.02087973 MagAttract PowerMicrobiome Kit 0.00260\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 8 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 8 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 8 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 30 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 30 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 30 rows containing missing values (geom_point).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Sample_name                                          taxa.id\n",
      "1  SC253181-PC04925-G-05                    D_4__BacillaceaeD_5__Bacillus\n",
      "2               SC253189                    D_4__BacillaceaeD_5__Bacillus\n",
      "3               SC253194                    D_4__BacillaceaeD_5__Bacillus\n",
      "4  SC253196-PC04925-F-07                    D_4__BacillaceaeD_5__Bacillus\n",
      "5  SC253181-PC07578-G-05                    D_4__BacillaceaeD_5__Bacillus\n",
      "6  SC253196-PC07578-F-07                    D_4__BacillaceaeD_5__Bacillus\n",
      "7  SC253181-PC04925-G-05 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "8               SC253189 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "9               SC253194 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "10 SC253196-PC04925-F-07 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "11 SC253181-PC07578-G-05 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "12 SC253196-PC07578-F-07 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "13 SC253181-PC04925-G-05           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "14              SC253189           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "15              SC253194           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "16 SC253196-PC04925-F-07           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "17 SC253181-PC07578-G-05           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "18 SC253196-PC07578-F-07           D_4__EnterobacteriaceaeD_5__Salmonella\n",
      "19 SC253181-PC04925-G-05            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "20              SC253189            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "21              SC253194            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "22 SC253196-PC04925-F-07            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "23 SC253181-PC07578-G-05            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "24 SC253196-PC07578-F-07            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "25 SC253181-PC04925-G-05          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "26              SC253189          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "27              SC253194          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "28 SC253196-PC04925-F-07          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "29 SC253181-PC07578-G-05          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "30 SC253196-PC07578-F-07          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "31 SC253181-PC04925-G-05                   D_4__ListeriaceaeD_5__Listeria\n",
      "32              SC253189                   D_4__ListeriaceaeD_5__Listeria\n",
      "33              SC253194                   D_4__ListeriaceaeD_5__Listeria\n",
      "34 SC253196-PC04925-F-07                   D_4__ListeriaceaeD_5__Listeria\n",
      "35 SC253181-PC07578-G-05                   D_4__ListeriaceaeD_5__Listeria\n",
      "36 SC253196-PC07578-F-07                   D_4__ListeriaceaeD_5__Listeria\n",
      "37 SC253181-PC04925-G-05            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "38              SC253189            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "39              SC253194            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "40 SC253196-PC04925-F-07            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "41 SC253181-PC07578-G-05            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "42 SC253196-PC07578-F-07            D_4__PseudomonadaceaeD_5__Pseudomonas\n",
      "43 SC253181-PC04925-G-05        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "44              SC253189        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "45              SC253194        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "46 SC253196-PC04925-F-07        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "47 SC253181-PC07578-G-05        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "48 SC253196-PC07578-F-07        D_4__StaphylococcaceaeD_5__Staphylococcus\n",
      "   Subject.ID   exp                     bench     obs\n",
      "1       D6300 0.174 QIAamp with Modifications 0.14820\n",
      "2       D6300 0.174 QIAamp with Modifications 0.17310\n",
      "3       D6300 0.174 QIAamp with Modifications 0.13215\n",
      "4       D6300 0.174 QIAamp with Modifications 0.19795\n",
      "5       D6300 0.174 QIAamp with Modifications 0.12655\n",
      "6       D6300 0.174 QIAamp with Modifications 0.17295\n",
      "7       D6300 0.101 QIAamp with Modifications 0.21590\n",
      "8       D6300 0.101 QIAamp with Modifications 0.19300\n",
      "9       D6300 0.101 QIAamp with Modifications 0.16840\n",
      "10      D6300 0.101 QIAamp with Modifications 0.17415\n",
      "11      D6300 0.101 QIAamp with Modifications 0.18575\n",
      "12      D6300 0.101 QIAamp with Modifications 0.16445\n",
      "13      D6300 0.104 QIAamp with Modifications 0.00000\n",
      "14      D6300 0.104 QIAamp with Modifications 0.00000\n",
      "15      D6300 0.104 QIAamp with Modifications 0.00000\n",
      "16      D6300 0.104 QIAamp with Modifications 0.00000\n",
      "17      D6300 0.104 QIAamp with Modifications 0.00000\n",
      "18      D6300 0.104 QIAamp with Modifications 0.00000\n",
      "19      D6300 0.099 QIAamp with Modifications 0.08715\n",
      "20      D6300 0.099 QIAamp with Modifications 0.08380\n",
      "21      D6300 0.099 QIAamp with Modifications 0.12020\n",
      "22      D6300 0.099 QIAamp with Modifications 0.10450\n",
      "23      D6300 0.099 QIAamp with Modifications 0.09365\n",
      "24      D6300 0.099 QIAamp with Modifications 0.10245\n",
      "25      D6300 0.184 QIAamp with Modifications 0.21240\n",
      "26      D6300 0.184 QIAamp with Modifications 0.20745\n",
      "27      D6300 0.184 QIAamp with Modifications 0.13425\n",
      "28      D6300 0.184 QIAamp with Modifications 0.12640\n",
      "29      D6300 0.184 QIAamp with Modifications 0.24645\n",
      "30      D6300 0.184 QIAamp with Modifications 0.14250\n",
      "31      D6300 0.141 QIAamp with Modifications 0.14600\n",
      "32      D6300 0.141 QIAamp with Modifications 0.14690\n",
      "33      D6300 0.141 QIAamp with Modifications 0.19560\n",
      "34      D6300 0.141 QIAamp with Modifications 0.18030\n",
      "35      D6300 0.141 QIAamp with Modifications 0.14615\n",
      "36      D6300 0.141 QIAamp with Modifications 0.19370\n",
      "37      D6300 0.042 QIAamp with Modifications 0.11245\n",
      "38      D6300 0.042 QIAamp with Modifications 0.10700\n",
      "39      D6300 0.042 QIAamp with Modifications 0.07625\n",
      "40      D6300 0.042 QIAamp with Modifications 0.08470\n",
      "41      D6300 0.042 QIAamp with Modifications 0.11570\n",
      "42      D6300 0.042 QIAamp with Modifications 0.09670\n",
      "43      D6300 0.155 QIAamp with Modifications 0.07770\n",
      "44      D6300 0.155 QIAamp with Modifications 0.08825\n",
      "45      D6300 0.155 QIAamp with Modifications 0.16985\n",
      "46      D6300 0.155 QIAamp with Modifications 0.12000\n",
      "47      D6300 0.155 QIAamp with Modifications 0.08035\n",
      "48      D6300 0.155 QIAamp with Modifications 0.12195\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Sample_name                                          taxa.id\n",
      "1               SC253185              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "2               SC253192              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "3  SC253195-PC04925-E-07              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "4               SC253204              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "5  SC253195-PC07578-E-07              D_4__BacteroidaceaeD_5__Bacteroides\n",
      "6               SC253185      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "7               SC253192      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "8  SC253195-PC04925-E-07      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "9               SC253204      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "10 SC253195-PC07578-E-07      D_4__BifidobacteriaceaeD_5__Bifidobacterium\n",
      "11              SC253185             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "12              SC253192             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "13 SC253195-PC04925-E-07             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "14              SC253204             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "15 SC253195-PC07578-E-07             D_4__BukerholderiaceaeD_5__Ralstonia\n",
      "16              SC253185           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "17              SC253192           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "18 SC253195-PC04925-E-07           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "19              SC253204           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "20 SC253195-PC07578-E-07           D_4__CoriobacteriaceaeD_5__Collinsella\n",
      "21              SC253185           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "22              SC253192           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "23 SC253195-PC04925-E-07           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "24              SC253204           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "25 SC253195-PC07578-E-07           D_4__DesulfovibrionaceaeD_5__Bilophila\n",
      "26              SC253185 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "27              SC253192 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "28 SC253195-PC04925-E-07 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "29              SC253204 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "30 SC253195-PC07578-E-07 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella\n",
      "31              SC253185            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "32              SC253192            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "33 SC253195-PC04925-E-07            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "34              SC253204            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "35 SC253195-PC07578-E-07            D_4__EnterococcaceaeD_5__Enterococcus\n",
      "36              SC253185        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "37              SC253192        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "38 SC253195-PC04925-E-07        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "39              SC253204        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "40 SC253195-PC07578-E-07        D_4__ErysipelotrichidaeD_5__Coprobacillus\n",
      "41              SC253185          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "42              SC253192          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "43 SC253195-PC04925-E-07          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "44              SC253204          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "45 SC253195-PC07578-E-07          D_4__FusobacteriaceaeD_5__Fusobacterium\n",
      "46              SC253185            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "47              SC253192            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "48 SC253195-PC04925-E-07            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "49              SC253204            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "50 SC253195-PC07578-E-07            D_4__LachnospiraceaeD_5__Anaerostipes\n",
      "51              SC253185       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "52              SC253192       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "53 SC253195-PC04925-E-07       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "54              SC253204       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "55 SC253195-PC07578-E-07       D_4__LachnospiraceaeD_5__Lachnoclostridium\n",
      "56              SC253185          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "57              SC253192          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "58 SC253195-PC04925-E-07          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "59              SC253204          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "60 SC253195-PC07578-E-07          D_4__LactobacillaceaeD_5__Lactobacillus\n",
      "61              SC253185            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "62              SC253192            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "63 SC253195-PC04925-E-07            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "64              SC253204            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "65 SC253195-PC07578-E-07            D_4__LactobacillaceaeD_5__Pediococcus\n",
      "66              SC253185          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "67              SC253192          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "68 SC253195-PC04925-E-07          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "69              SC253204          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "70 SC253195-PC07578-E-07          D_4__PaenibacillaceaeD_5__Paenibacillus\n",
      "71              SC253185      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "72              SC253192      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "73 SC253195-PC04925-E-07      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "74              SC253204      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "75 SC253195-PC07578-E-07      D_4__PorphyromonadaceaeD_5__Parabacteroides\n",
      "76              SC253185  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "77              SC253192  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "78 SC253195-PC04925-E-07  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "79              SC253204  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "80 SC253195-PC07578-E-07  D_4__PropionibacteriaceaeD_5__Propionibacterium\n",
      "81              SC253185                D_4__RikenellaceaeD_5__Allistipes\n",
      "82              SC253192                D_4__RikenellaceaeD_5__Allistipes\n",
      "83 SC253195-PC04925-E-07                D_4__RikenellaceaeD_5__Allistipes\n",
      "84              SC253204                D_4__RikenellaceaeD_5__Allistipes\n",
      "85 SC253195-PC07578-E-07                D_4__RikenellaceaeD_5__Allistipes\n",
      "86              SC253185         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "87              SC253192         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "88 SC253195-PC04925-E-07         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "89              SC253204         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "90 SC253195-PC07578-E-07         D_4__RuminococcaceaeD_5__Subdoligranulum\n",
      "91              SC253185           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "92              SC253192           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "93 SC253195-PC04925-E-07           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "94              SC253204           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "95 SC253195-PC07578-E-07           D_4__SynergistaceaeD_5__Pyramidobacter\n",
      "   Subject.ID        exp                     bench     obs\n",
      "1     DZ35322 0.06263920 QIAamp with Modifications 0.09010\n",
      "2     DZ35322 0.06263920 QIAamp with Modifications 0.10395\n",
      "3     DZ35322 0.06263920 QIAamp with Modifications 0.07115\n",
      "4     DZ35322 0.06263920 QIAamp with Modifications 0.07795\n",
      "5     DZ35322 0.06263920 QIAamp with Modifications 0.14310\n",
      "6     DZ35322 0.06263920 QIAamp with Modifications 0.00570\n",
      "7     DZ35322 0.06263920 QIAamp with Modifications 0.00655\n",
      "8     DZ35322 0.06263920 QIAamp with Modifications 0.02130\n",
      "9     DZ35322 0.06263920 QIAamp with Modifications 0.01660\n",
      "10    DZ35322 0.06263920 QIAamp with Modifications 0.01915\n",
      "11    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "12    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "13    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "14    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "15    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "16    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "17    DZ35322 0.06263920 QIAamp with Modifications 0.00165\n",
      "18    DZ35322 0.06263920 QIAamp with Modifications 0.00405\n",
      "19    DZ35322 0.06263920 QIAamp with Modifications 0.00275\n",
      "20    DZ35322 0.06263920 QIAamp with Modifications 0.00245\n",
      "21    DZ35322 0.02352450 QIAamp with Modifications 0.03555\n",
      "22    DZ35322 0.02352450 QIAamp with Modifications 0.03680\n",
      "23    DZ35322 0.02352450 QIAamp with Modifications 0.02775\n",
      "24    DZ35322 0.02352450 QIAamp with Modifications 0.04340\n",
      "25    DZ35322 0.02352450 QIAamp with Modifications 0.02675\n",
      "26    DZ35322 0.06263920 QIAamp with Modifications 0.15825\n",
      "27    DZ35322 0.06263920 QIAamp with Modifications 0.12490\n",
      "28    DZ35322 0.06263920 QIAamp with Modifications 0.10395\n",
      "29    DZ35322 0.06263920 QIAamp with Modifications 0.10885\n",
      "30    DZ35322 0.06263920 QIAamp with Modifications 0.09125\n",
      "31    DZ35322 0.06263920 QIAamp with Modifications 0.01455\n",
      "32    DZ35322 0.06263920 QIAamp with Modifications 0.03575\n",
      "33    DZ35322 0.06263920 QIAamp with Modifications 0.03955\n",
      "34    DZ35322 0.06263920 QIAamp with Modifications 0.04365\n",
      "35    DZ35322 0.06263920 QIAamp with Modifications 0.03470\n",
      "36    DZ35322 0.03660913 QIAamp with Modifications 0.00000\n",
      "37    DZ35322 0.03660913 QIAamp with Modifications 0.00000\n",
      "38    DZ35322 0.03660913 QIAamp with Modifications 0.00000\n",
      "39    DZ35322 0.03660913 QIAamp with Modifications 0.00000\n",
      "40    DZ35322 0.03660913 QIAamp with Modifications 0.00000\n",
      "41    DZ35322 0.12527840 QIAamp with Modifications 0.18045\n",
      "42    DZ35322 0.12527840 QIAamp with Modifications 0.15875\n",
      "43    DZ35322 0.12527840 QIAamp with Modifications 0.22865\n",
      "44    DZ35322 0.12527840 QIAamp with Modifications 0.18275\n",
      "45    DZ35322 0.12527840 QIAamp with Modifications 0.15885\n",
      "46    DZ35322 0.06263920 QIAamp with Modifications 0.01385\n",
      "47    DZ35322 0.06263920 QIAamp with Modifications 0.01920\n",
      "48    DZ35322 0.06263920 QIAamp with Modifications 0.01855\n",
      "49    DZ35322 0.06263920 QIAamp with Modifications 0.01265\n",
      "50    DZ35322 0.06263920 QIAamp with Modifications 0.01840\n",
      "51    DZ35322 0.06263920 QIAamp with Modifications 0.11275\n",
      "52    DZ35322 0.06263920 QIAamp with Modifications 0.11545\n",
      "53    DZ35322 0.06263920 QIAamp with Modifications 0.10705\n",
      "54    DZ35322 0.06263920 QIAamp with Modifications 0.13485\n",
      "55    DZ35322 0.06263920 QIAamp with Modifications 0.09925\n",
      "56    DZ35322 0.03660913 QIAamp with Modifications 0.01485\n",
      "57    DZ35322 0.03660913 QIAamp with Modifications 0.01550\n",
      "58    DZ35322 0.03660913 QIAamp with Modifications 0.02305\n",
      "59    DZ35322 0.03660913 QIAamp with Modifications 0.01700\n",
      "60    DZ35322 0.03660913 QIAamp with Modifications 0.02155\n",
      "61    DZ35322 0.04175947 QIAamp with Modifications 0.01620\n",
      "62    DZ35322 0.04175947 QIAamp with Modifications 0.03730\n",
      "63    DZ35322 0.04175947 QIAamp with Modifications 0.03105\n",
      "64    DZ35322 0.04175947 QIAamp with Modifications 0.04250\n",
      "65    DZ35322 0.04175947 QIAamp with Modifications 0.02705\n",
      "66    DZ35322 0.03660913 QIAamp with Modifications 0.04810\n",
      "67    DZ35322 0.03660913 QIAamp with Modifications 0.04225\n",
      "68    DZ35322 0.03660913 QIAamp with Modifications 0.04405\n",
      "69    DZ35322 0.03660913 QIAamp with Modifications 0.04320\n",
      "70    DZ35322 0.03660913 QIAamp with Modifications 0.03295\n",
      "71    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "72    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "73    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "74    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "75    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "76    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "77    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "78    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "79    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "80    DZ35322 0.06263920 QIAamp with Modifications 0.00000\n",
      "81    DZ35322 0.03660913 QIAamp with Modifications 0.00000\n",
      "82    DZ35322 0.03660913 QIAamp with Modifications 0.00000\n",
      "83    DZ35322 0.03660913 QIAamp with Modifications 0.00000\n",
      "84    DZ35322 0.03660913 QIAamp with Modifications 0.00000\n",
      "85    DZ35322 0.03660913 QIAamp with Modifications 0.00000\n",
      "86    DZ35322 0.01572940 QIAamp with Modifications 0.01070\n",
      "87    DZ35322 0.01572940 QIAamp with Modifications 0.00950\n",
      "88    DZ35322 0.01572940 QIAamp with Modifications 0.00895\n",
      "89    DZ35322 0.01572940 QIAamp with Modifications 0.01170\n",
      "90    DZ35322 0.01572940 QIAamp with Modifications 0.00845\n",
      "91    DZ35322 0.02087973 QIAamp with Modifications 0.00430\n",
      "92    DZ35322 0.02087973 QIAamp with Modifications 0.00445\n",
      "93    DZ35322 0.02087973 QIAamp with Modifications 0.00505\n",
      "94    DZ35322 0.02087973 QIAamp with Modifications 0.00620\n",
      "95    DZ35322 0.02087973 QIAamp with Modifications 0.00360\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 6 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 6 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 6 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 25 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 25 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 25 rows containing missing values (geom_point).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Sample_name                                        taxa.id Subject.ID\n",
      "1     SC552982                  D_4__BacillaceaeD_5__Bacillus    DZ35316\n",
      "2     SC552982    D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35316\n",
      "3     SC552982       D_4__BrevibacteriaceaeD_5__Stomatococcus    DZ35316\n",
      "4     SC552982     D_4__CampylobacteriaceaeD_5__Campylobacter    DZ35316\n",
      "5     SC552982         D_4__CoriobacteriaceaeD_5__Eggerthella    DZ35316\n",
      "6     SC552982             D_4__CoriobacteriaceaeD_5__Slackia    DZ35316\n",
      "7     SC552982         D_4__EnterobacteriaceaeD_5__Klebsiella    DZ35316\n",
      "8     SC552982          D_4__EubacteriaceaeD_5__Mogibacterium    DZ35316\n",
      "9     SC552982      D_4__FlavobacteriaceaeD_5__Capnocytophaga    DZ35316\n",
      "10    SC552982        D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35316\n",
      "11    SC552982         D_4__LeptotrichiaceaeD_5__Leptotrichia    DZ35316\n",
      "12    SC552982            D_4__LeuconostocaceaeD_5__Weissella    DZ35316\n",
      "13    SC552982               D_4__NeisseriaceaeD_5__Eikenella    DZ35316\n",
      "14    SC552982               D_4__NeisseriaceaeD_5__Neisseria    DZ35316\n",
      "15    SC552982           D_4__PeptoniphilaceaeD_5__Parvimonas    DZ35316\n",
      "16    SC552982        D_4__PorphyromonadaceaeD_5__Barnesiella    DZ35316\n",
      "17    SC552982             D_4__PrevotellaceaeD_5__Prevotella    DZ35316\n",
      "18    SC552982       D_4__StreptococcaceaeD_5__Granulicatella    DZ35316\n",
      "19    SC552982        D_4__StreptococcaceaeD_5__Streptococcus    DZ35316\n",
      "20    SC552982 D_4__Tenericutes.Incertae.Sedis.XID_5__Gemella    DZ35316\n",
      "21    SC552982             D_4__VeillonellaceaeD_5__Dialister    DZ35316\n",
      "22    SC552982           D_4__VeillonellaceaeD_5__Veillonella    DZ35316\n",
      "           exp                    bench     obs\n",
      "1  0.041791461 DNeasy PowerSoil Pro kit 0.02715\n",
      "2  0.062687191 DNeasy PowerSoil Pro kit 0.03665\n",
      "3  0.062687191 DNeasy PowerSoil Pro kit 0.00000\n",
      "4  0.031343595 DNeasy PowerSoil Pro kit 0.00000\n",
      "5  0.036637180 DNeasy PowerSoil Pro kit 0.00000\n",
      "6  0.036637180 DNeasy PowerSoil Pro kit 0.00000\n",
      "7  0.062687191 DNeasy PowerSoil Pro kit 0.00000\n",
      "8  0.001044787 DNeasy PowerSoil Pro kit 0.00000\n",
      "9  0.005223933 DNeasy PowerSoil Pro kit 0.00165\n",
      "10 0.062687191 DNeasy PowerSoil Pro kit 0.05220\n",
      "11 0.052239326 DNeasy PowerSoil Pro kit 0.03990\n",
      "12 0.052239326 DNeasy PowerSoil Pro kit 0.02565\n",
      "13 0.041791461 DNeasy PowerSoil Pro kit 0.04080\n",
      "14 0.062687191 DNeasy PowerSoil Pro kit 0.19005\n",
      "15 0.031343595 DNeasy PowerSoil Pro kit 0.00000\n",
      "16 0.062687191 DNeasy PowerSoil Pro kit 0.00000\n",
      "17 0.062687191 DNeasy PowerSoil Pro kit 0.04070\n",
      "18 0.062687191 DNeasy PowerSoil Pro kit 0.00000\n",
      "19 0.062687191 DNeasy PowerSoil Pro kit 0.05040\n",
      "20 0.052239326 DNeasy PowerSoil Pro kit 0.00000\n",
      "21 0.001044787 DNeasy PowerSoil Pro kit 0.00000\n",
      "22 0.052239326 DNeasy PowerSoil Pro kit 0.05030\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Sample_name                                          taxa.id Subject.ID\n",
      "1    SC552988                    D_4__BacillaceaeD_5__Bacillus      D6310\n",
      "2    SC552988 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella      D6310\n",
      "3    SC552988           D_4__EnterobacteriaceaeD_5__Salmonella      D6310\n",
      "4    SC552988            D_4__EnterococcaceaeD_5__Enterococcus      D6310\n",
      "5    SC552988          D_4__LactobacillaceaeD_5__Lactobacillus      D6310\n",
      "6    SC552988                   D_4__ListeriaceaeD_5__Listeria      D6310\n",
      "7    SC552988            D_4__PseudomonadaceaeD_5__Pseudomonas      D6310\n",
      "8    SC552988        D_4__StaphylococcaceaeD_5__Staphylococcus      D6310\n",
      "       exp                    bench     obs\n",
      "1 1.20e-02 DNeasy PowerSoil Pro kit 0.00960\n",
      "2 6.90e-04 DNeasy PowerSoil Pro kit 0.00110\n",
      "3 7.00e-04 DNeasy PowerSoil Pro kit 0.00000\n",
      "4 6.70e-06 DNeasy PowerSoil Pro kit 0.00000\n",
      "5 1.20e-04 DNeasy PowerSoil Pro kit 0.00000\n",
      "6 9.59e-01 DNeasy PowerSoil Pro kit 0.96810\n",
      "7 2.80e-02 DNeasy PowerSoil Pro kit 0.02025\n",
      "8 1.00e-06 DNeasy PowerSoil Pro kit 0.00010\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Sample_name                                          taxa.id Subject.ID   exp\n",
      "1    SC552990                    D_4__BacillaceaeD_5__Bacillus      D6300 0.174\n",
      "2    SC552990 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella      D6300 0.101\n",
      "3    SC552990           D_4__EnterobacteriaceaeD_5__Salmonella      D6300 0.104\n",
      "4    SC552990            D_4__EnterococcaceaeD_5__Enterococcus      D6300 0.099\n",
      "5    SC552990          D_4__LactobacillaceaeD_5__Lactobacillus      D6300 0.184\n",
      "6    SC552990                   D_4__ListeriaceaeD_5__Listeria      D6300 0.141\n",
      "7    SC552990            D_4__PseudomonadaceaeD_5__Pseudomonas      D6300 0.042\n",
      "8    SC552990        D_4__StaphylococcaceaeD_5__Staphylococcus      D6300 0.155\n",
      "                     bench     obs\n",
      "1 DNeasy PowerSoil Pro kit 0.00735\n",
      "2 DNeasy PowerSoil Pro kit 0.00115\n",
      "3 DNeasy PowerSoil Pro kit 0.00000\n",
      "4 DNeasy PowerSoil Pro kit 0.00000\n",
      "5 DNeasy PowerSoil Pro kit 0.00000\n",
      "6 DNeasy PowerSoil Pro kit 0.96660\n",
      "7 DNeasy PowerSoil Pro kit 0.01810\n",
      "8 DNeasy PowerSoil Pro kit 0.00000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_point).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in max(data$density):\n",
      "\"no non-missing arguments to max; returning -Inf\"\n",
      "Warning message:\n",
      "\"Computation failed in `stat_ydensity()`:\n",
      "replacement has 1 row, data has 0\"\n",
      "Warning message:\n",
      "\"Removed 3 rows containing missing values (geom_segment).\"\n",
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Sample_name                                        taxa.id Subject.ID\n",
      "1     SC553025                  D_4__BacillaceaeD_5__Bacillus    DZ35316\n",
      "2     SC553025    D_4__BifidobacteriaceaeD_5__Bifidobacterium    DZ35316\n",
      "3     SC553025       D_4__BrevibacteriaceaeD_5__Stomatococcus    DZ35316\n",
      "4     SC553025     D_4__CampylobacteriaceaeD_5__Campylobacter    DZ35316\n",
      "5     SC553025         D_4__CoriobacteriaceaeD_5__Eggerthella    DZ35316\n",
      "6     SC553025             D_4__CoriobacteriaceaeD_5__Slackia    DZ35316\n",
      "7     SC553025         D_4__EnterobacteriaceaeD_5__Klebsiella    DZ35316\n",
      "8     SC553025          D_4__EubacteriaceaeD_5__Mogibacterium    DZ35316\n",
      "9     SC553025      D_4__FlavobacteriaceaeD_5__Capnocytophaga    DZ35316\n",
      "10    SC553025        D_4__FusobacteriaceaeD_5__Fusobacterium    DZ35316\n",
      "11    SC553025         D_4__LeptotrichiaceaeD_5__Leptotrichia    DZ35316\n",
      "12    SC553025            D_4__LeuconostocaceaeD_5__Weissella    DZ35316\n",
      "13    SC553025               D_4__NeisseriaceaeD_5__Eikenella    DZ35316\n",
      "14    SC553025               D_4__NeisseriaceaeD_5__Neisseria    DZ35316\n",
      "15    SC553025           D_4__PeptoniphilaceaeD_5__Parvimonas    DZ35316\n",
      "16    SC553025        D_4__PorphyromonadaceaeD_5__Barnesiella    DZ35316\n",
      "17    SC553025             D_4__PrevotellaceaeD_5__Prevotella    DZ35316\n",
      "18    SC553025       D_4__StreptococcaceaeD_5__Granulicatella    DZ35316\n",
      "19    SC553025        D_4__StreptococcaceaeD_5__Streptococcus    DZ35316\n",
      "20    SC553025 D_4__Tenericutes.Incertae.Sedis.XID_5__Gemella    DZ35316\n",
      "21    SC553025             D_4__VeillonellaceaeD_5__Dialister    DZ35316\n",
      "22    SC553025           D_4__VeillonellaceaeD_5__Veillonella    DZ35316\n",
      "           exp                       bench     obs\n",
      "1  0.041791461 MagMax Microbiome Ultra Kit 0.05345\n",
      "2  0.062687191 MagMax Microbiome Ultra Kit 0.00465\n",
      "3  0.062687191 MagMax Microbiome Ultra Kit 0.00000\n",
      "4  0.031343595 MagMax Microbiome Ultra Kit 0.00000\n",
      "5  0.036637180 MagMax Microbiome Ultra Kit 0.00000\n",
      "6  0.036637180 MagMax Microbiome Ultra Kit 0.00000\n",
      "7  0.062687191 MagMax Microbiome Ultra Kit 0.00000\n",
      "8  0.001044787 MagMax Microbiome Ultra Kit 0.00000\n",
      "9  0.005223933 MagMax Microbiome Ultra Kit 0.00210\n",
      "10 0.062687191 MagMax Microbiome Ultra Kit 0.11300\n",
      "11 0.052239326 MagMax Microbiome Ultra Kit 0.08725\n",
      "12 0.052239326 MagMax Microbiome Ultra Kit 0.00200\n",
      "13 0.041791461 MagMax Microbiome Ultra Kit 0.04940\n",
      "14 0.062687191 MagMax Microbiome Ultra Kit 0.18610\n",
      "15 0.031343595 MagMax Microbiome Ultra Kit 0.00000\n",
      "16 0.062687191 MagMax Microbiome Ultra Kit 0.00000\n",
      "17 0.062687191 MagMax Microbiome Ultra Kit 0.05845\n",
      "18 0.062687191 MagMax Microbiome Ultra Kit 0.00000\n",
      "19 0.062687191 MagMax Microbiome Ultra Kit 0.01515\n",
      "20 0.052239326 MagMax Microbiome Ultra Kit 0.00000\n",
      "21 0.001044787 MagMax Microbiome Ultra Kit 0.00125\n",
      "22 0.052239326 MagMax Microbiome Ultra Kit 0.07665\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n",
      "Warning message in cor.test.default(my_data$obs, my_data$exp, method = \"spearman\"):\n",
      "\"Cannot compute exact p-value with ties\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Sample_name                                          taxa.id Subject.ID   exp\n",
      "1    SC553030                    D_4__BacillaceaeD_5__Bacillus      D6300 0.174\n",
      "2    SC553030 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella      D6300 0.101\n",
      "3    SC553030           D_4__EnterobacteriaceaeD_5__Salmonella      D6300 0.104\n",
      "4    SC553030            D_4__EnterococcaceaeD_5__Enterococcus      D6300 0.099\n",
      "5    SC553030          D_4__LactobacillaceaeD_5__Lactobacillus      D6300 0.184\n",
      "6    SC553030                   D_4__ListeriaceaeD_5__Listeria      D6300 0.141\n",
      "7    SC553030            D_4__PseudomonadaceaeD_5__Pseudomonas      D6300 0.042\n",
      "8    SC553030        D_4__StaphylococcaceaeD_5__Staphylococcus      D6300 0.155\n",
      "                        bench     obs\n",
      "1 MagMax Microbiome Ultra Kit 0.11235\n",
      "2 MagMax Microbiome Ultra Kit 0.33090\n",
      "3 MagMax Microbiome Ultra Kit 0.00000\n",
      "4 MagMax Microbiome Ultra Kit 0.16395\n",
      "5 MagMax Microbiome Ultra Kit 0.07105\n",
      "6 MagMax Microbiome Ultra Kit 0.06160\n",
      "7 MagMax Microbiome Ultra Kit 0.03240\n",
      "8 MagMax Microbiome Ultra Kit 0.16670\n",
      "  Sample_name                                          taxa.id Subject.ID\n",
      "1    SC553031                    D_4__BacillaceaeD_5__Bacillus      D6310\n",
      "2    SC553031 D_4__EnterobacteriaceaeD_5__Escherichia.Shigella      D6310\n",
      "3    SC553031           D_4__EnterobacteriaceaeD_5__Salmonella      D6310\n",
      "4    SC553031            D_4__EnterococcaceaeD_5__Enterococcus      D6310\n",
      "5    SC553031          D_4__LactobacillaceaeD_5__Lactobacillus      D6310\n",
      "6    SC553031                   D_4__ListeriaceaeD_5__Listeria      D6310\n",
      "7    SC553031            D_4__PseudomonadaceaeD_5__Pseudomonas      D6310\n",
      "8    SC553031        D_4__StaphylococcaceaeD_5__Staphylococcus      D6310\n",
      "       exp                       bench     obs\n",
      "1 1.20e-02 MagMax Microbiome Ultra Kit 0.00780\n",
      "2 6.90e-04 MagMax Microbiome Ultra Kit 0.00650\n",
      "3 7.00e-04 MagMax Microbiome Ultra Kit 0.00000\n",
      "4 6.70e-06 MagMax Microbiome Ultra Kit 0.00355\n",
      "5 1.20e-04 MagMax Microbiome Ultra Kit 0.00150\n",
      "6 9.59e-01 MagMax Microbiome Ultra Kit 0.33795\n",
      "7 2.80e-02 MagMax Microbiome Ultra Kit 0.01045\n",
      "8 1.00e-06 MagMax Microbiome Ultra Kit 0.00170\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 10 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_point).\"\n",
      "`geom_smooth()` using formula 'y ~ x'\n",
      "\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_smooth).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing non-finite values (stat_cor).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_point).\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (variable in test_list){\n",
    "    \n",
    "    #Subset dataframe\n",
    "    OTUdf_clean2 = subset(OTUdf_clean,Ext.Kit==variable & Homo.Status==\"Standard\" & Subject.ID!=\"Water\")\n",
    "    \n",
    "    #Create TP,FP,FN df\n",
    "    CONTdf_use = CONTdf_s\n",
    "    F1df = bench_compare(OTUdf_clean2, CONTdf_use)\n",
    "    \n",
    "    #Create TP,FP,FN counts    \n",
    "    bench_key = \"standard\"\n",
    "    F1df_counts = bench_counts(OTUdf_clean2,F1df,variable, bench_key)#swapped variable and benchkey from original\n",
    "    temp = F1df_counts\n",
    "    temp$sample_name = rownames(temp)\n",
    "    rownames(temp) = c()\n",
    "    mergedf_f1stats = rbind(mergedf_f1stats,temp)\n",
    "\n",
    "    #Create violin plots for each control - counts\n",
    "    bench_key = variable\n",
    "    x_list = c(\"FN\",\"TP\",\"FP\")\n",
    "    y_labs = c(\"False negative\",\"True positive\", \"False positive\")\n",
    "    plot_f1_violin(F1df_counts,x_list,y_labs,\"Subject.ID\",bench_key,\"counts_extstd\")\n",
    "        \n",
    "    #Create violin plots for each control - summary\n",
    "    bench_key = variable\n",
    "    x_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "    y_labs = c(\"Taxonmic Accuracy Rate\",\"Taxonmic Detection Rate\", \"F-Measure\")\n",
    "    plot_f1_violin(F1df_counts,x_list,y_labs,\"Subject.ID\",bench_key,\"summary_extstd\")\n",
    "    \n",
    "    #calculate relative abundance\n",
    "    RAdf = cal_relabun(OTUdf_clean2)\n",
    "    RAdf$Sample_name = rownames(RAdf)\n",
    "    \n",
    "    #Create scatter plots\n",
    "    #Determine expected to observed scatter plots, normality, spearman coeff \n",
    "    title = paste(\"Scatterplots of controls: \", variable,sep=\"\")\n",
    "    EOdf_stats = cal_EO(RAdf,CONTdf_use,variable,title)    \n",
    "    #EOdf_stats_merged = rbind(EOdf_stats_merged,EOdf_stats) \n",
    "    \n",
    "    #Create averaged OTU df\n",
    "    bench_key = \"std\"\n",
    "    dissdf = create_av_RAdf(RAdf,variable,bench_key) #swap order of bench and variable\n",
    "    dissdf_s = bind_rows(dissdf_s,dissdf)\n",
    "    \n",
    "    #save files\n",
    "    #write.csv(OTUdf_clean,paste(out_dir,\"bench_files/OTUcounts_\",bench_key,\".csv\",sep=\"\"))\n",
    "    #write.csv(F1df,paste(out_dir,\"/bench_files/F1counts_\",bench_key,\".csv\",sep=\"\"))\n",
    "    write.csv(F1df_counts,paste(out_dir,\"bench_files/F1stats_extstd_\",variable,\".csv\",sep=\"\"))#swap bench for variable\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(EOdf_stats_merged,paste(out_dir,\"bench_files/EO_stats_summary_extstd.csv\",sep=\"\"))\n",
    "write.csv(mergedf_f1stats,paste(out_dir,\"bench_files/F1_stats_summary_extstd.csv\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4> Benchmarked Assessment<h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Distdf_merged=data.frame()\n",
    "#Silva\n",
    "if(length(test_list)>1){\n",
    "    RAdf_oe = RAdf_addcontrols(\"CONTdf_s\",\"dissdf_s\")\n",
    "    RAdf_clean = calc_dist(RAdf_oe,CONTdf_s)\n",
    "    plot_dist_cluster(RAdf_clean,\"extstd\")\n",
    "    Distdf = plot_dist_violin(RAdf_clean,\"extstd\")\n",
    "} \n",
    "Distdf_merged = bind_rows(Distdf_merged,Distdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in betadisper(dist_s, dissdf_s$bmark):\n",
      "\"some squared distances are negative and changed to zero\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 2  4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>comparison</th><th scope=col>pval</th><th scope=col>sig</th><th scope=col>homo_s</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>bray-curtis</td><td>0.78</td><td>ns</td><td>       NA</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>NA         </td><td>  NA</td><td>NA</td><td>0.9781328</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2  4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & comparison & pval & sig & homo\\_s\\\\\n",
       "  & <chr> & <dbl> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & bray-curtis & 0.78 & ns &        NA\\\\\n",
       "\t2 & NA          &   NA & NA & 0.9781328\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2  4\n",
       "\n",
       "| <!--/--> | comparison &lt;chr&gt; | pval &lt;dbl&gt; | sig &lt;chr&gt; | homo_s &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | bray-curtis | 0.78 | ns |        NA |\n",
       "| 2 | NA          |   NA | NA | 0.9781328 |\n",
       "\n"
      ],
      "text/plain": [
       "  comparison  pval sig homo_s   \n",
       "1 bray-curtis 0.78 ns         NA\n",
       "2 NA            NA NA  0.9781328"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "adonis_stats = adonis_bc(Distdf_merged)\n",
    "adonis_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in as_grob.default(plot):\n",
      "\"Cannot convert object of class character into a grob.\"\n",
      "Saving 6.67 x 6.67 in image\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Overall violin plots\n",
    "x_list = c(\"TAR\",\"TDR\",\"F1\")\n",
    "y_labs = c(\"Taxonmic Accuracy Rate\",\"Taxonmic Detection Rate\", \"F-Measure\")\n",
    "\n",
    "#Create violin plots collapsed to the group.type (merges all ref dbs, and tax class methods)\n",
    "plot_f1_violin(mergedf_f1stats,x_list,y_labs,\"bmark\",\"Ext Protocols Ext Std\",\"summary\")\n",
    "\n",
    "#Create violin plots collapsed to the tax.type (merges all ref dbs, and grouping methods)\n",
    "#plot_f1_violin(mergedf_f1stats,x_list,y_labs,\"tax.type\",\"Tax Types\",\"summary\")\n",
    "\n",
    "#Create violin plots collapsed to the tax.ref (merges all tax class methods and grouping methods)\n",
    "#plot_f1_violin(mergedf_f1stats,x_list,y_labs,\"tax.ref\",\"Tax Ref Types\",\"summary\")\n",
    "\n",
    "#Create violin plots for all binf variations (merges all variables)\n",
    "#plot_f1_violin(mergedf_f1stats,x_list,y_labs,\"bmark\",\"All Grouping and Assignment Variables\",\"summary\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Note: Using an external vector in selections is ambiguous.\n",
      "\u001b[34mi\u001b[39m Use `all_of(subgroup_in)` instead of `subgroup_in` to silence this message.\n",
      "\u001b[34mi\u001b[39m See <https://tidyselect.r-lib.org/reference/faq-external-vector.html>.\n",
      "\u001b[90mThis message is displayed once per session.\u001b[39m\n",
      "\n",
      "Note: Using an external vector in selections is ambiguous.\n",
      "\u001b[34mi\u001b[39m Use `all_of(cal)` instead of `cal` to silence this message.\n",
      "\u001b[34mi\u001b[39m See <https://tidyselect.r-lib.org/reference/faq-external-vector.html>.\n",
      "\u001b[90mThis message is displayed once per session.\u001b[39m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#T-tests\n",
    "tstatdf_merged=data.frame()\n",
    "variable_list = c(\"bmark\")\n",
    "\n",
    "for (variable in variable_list){\n",
    "    tstatdf = tstat_paired_f1(mergedf_f1stats,variable)\n",
    "    tstatdf_merged=bind_rows(tstatdf_merged,tstatdf)  \n",
    "}\n",
    "write.csv(tstatdf_merged,paste(out_dir,\"bench_files/ttest_stats_extstd.csv\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Blank assessment <h3>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
