---
title: "R Notebook"
output: word_document
editor_options: 
  chunk_output_type: console
---

#To Update
```{r}
project_name="Project_NP0084-MB4_5_pt6"
manifest_name="NP0084-MB4_08_29_19_metadata.txt"
failed_name = "samplelist_baseline_failed.txt"
sample_depth = 20000
```


#Load required libraries
############### 
```{r}
library("biom")
library(data.table)
library(ggplot2)
library(tm)
library(vegan)
library("biom")
library(tidyr)
library(ggplot2)
library(cowplot)
library(scales)
library(phyloseq)
library(qiime2R)
library(tibble)
library(gridExtra)
source("sources/miseqR.R")
```

#Analysis
############### 
#1.Create PhySeq Object
```{r}
project_location = paste("T:\\DCEG\\Projects\\Microbiome\\CGR_MB\\MicroBiome\\",project_name,sep="")

#Read OTUS
otus <- read_qza(paste(project_location,"\\Output\\qza_results\\table_dada2_qza_merged_parts_final\\table_dada2_merged_final_filt.qza",sep=""))

#Read rooted tree
tree <- read_qza(paste(project_location,"\\Output\\qza_results\\phylogeny_qza_results\\rooted_tree.qza",sep=""))

#Read Greengenes taxonomy file
taxonomy <- read_qza(paste(project_location,"\\Output\\qza_results\\taxonomy_qza_results\\taxonomy_greengenes.qza",sep=""))
##Edit dtable
tax_table<-do.call(rbind, strsplit(as.character(taxonomy$data$Taxon), "; "))
  colnames(tax_table)<-c("Kingdom","Phylum","Class","Order","Family","Genus","Species")
  rownames(tax_table)<-taxonomy$data$Feature.ID

#read metadata
metadata <-read.table(paste(project_location,manifest_name,sep="\\"),sep='\t', header=T, row.names=1, comment="")

#Create filtered list of samples
samplelist_passed_q2 <- as.list(sample_names(otu_table(otus$data,taxa_are_rows = T)))
samples_missing_q2 <- setdiff(row.names(metadata),samplelist_passed_q2)
write.csv(samples_missing_q2,paste(project_location,"samplelist_baseline_failed.txt",sep="\\"))

#Create phylo object
physeq_complete<-phyloseq(otu_table(otus$data, taxa_are_rows = T), phy_tree(tree$data), tax_table(tax_table), sample_data(metadata))

physeq_complete #verify results are as expected

s<-summary(sample_data(physeq_complete))
capture.output(s, file = "Data/Baseline/summary_prefilter.txt")
```

#2 Prune for bacteria only, Create prefilter summary files
```{r}
#Only bacteria
physeq_filt <- physeq_complete %>%
  subset_taxa(
    Kingdom == "k__Bacteria" &
    Family  != "k__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__mitochondria" &
    Class   != "k__Bacteria; p__Cyanobacteria; c__Chloroplast"
  )
physeq_filt

#Print summaries
a<-summary(sample_data(physeq_filt)$Sample.Type)
capture.output(a, file = "Data/Baseline/summary_sampletype_prefilter.txt")
```

#3. Determine read counts and plot
```{r}
# Make a data frame with a column for the read counts of each sample
sample_sum_df <- data.frame(sum = sample_sums(physeq_filt),sample_data(physeq_filt))
colnames(sample_sum_df)

# Histogram of sample read counts by sampletype
p1<-ggplot(sample_sum_df, aes(x = sum, color=Sample.Type)) + 
  geom_histogram(fill="white",binwidth = 2500) +
  xlab("Read counts") +
  ggtitle("Distribution of sample sequencing depth - Pre-Filtering")+
  theme(axis.title.y = element_blank())

# Histogram of sample read counts by subject
p2<-ggplot(sample_sum_df, aes(x = sum, color=SubjectID)) + 
  geom_histogram(fill="white",binwidth = 2500) +
  xlab("Read counts") +
  theme(axis.title.y = element_blank())

#Save file
file_name =("Graphs/Baseline/SeqDepth_prefilter.tiff")
tiff(file_name, width=800)
print(p1)
dev.off()

```

#4. Filter taxa >.001 and samples with less than 1000 reads
```{r}
physeq_filt #note taxa number
physeq_filt = filter_taxa(physeq_filt, function(x) mean(x) > 1e-2, TRUE)
physeq_filt #note new taxa number and sample number


physeq_filt = prune_samples(sample_sums(physeq_filt) > 10000, physeq_filt)
physeq_filt #note new sample number
```

#5. Review filtered read counts and plot
```{r}
# Make a data frame with a column for the read counts of each sample
sample_sum_df <- data.frame(sum = sample_sums(physeq_filt),sample_data(physeq_filt))
colnames(sample_sum_df)

# Histogram of sample read counts by sampletype
p1<-ggplot(sample_sum_df, aes(x = sum, color=Sample.Type)) + 
  geom_histogram(fill="white",binwidth = 2500) +
  xlab("Read counts") +
  ggtitle("Distribution of sample sequencing depth - Post-Filtering") +
  theme(axis.title.y = element_blank())

# Histogram of sample read counts by sample type
p2<-ggplot(sample_sum_df, aes(x = sum, color=SubjectID)) + 
  geom_histogram(fill="white",binwidth = 2500) +
  xlab("Read counts") +
  theme(axis.title.y = element_blank())

#Save file
file_name =("Graphs/Baseline/SeqDepth_filter.tiff")
tiff(file_name, width=800)
print(p1)
dev.off()

#Print summaries
s<-summary(sample_data(physeq_filt))
capture.output(s, file = "Data/Baseline/summary_filter.txt")
a<-summary(sample_data(physeq_filt)$Sample.Type)
capture.output(a, file = "Data/Baseline/summary_sampletype_postfilter.txt")
a<-summary(sample_data(physeq_filt)$Subject)
capture.output(a, file = "Data/Baseline/summary_subject_postfilter.txt")
```

#6. Subset filtered data, sample to specified depth
```{r}
#Only samples
physeq_filt_seqcontrols <- physeq_filt %>%
  subset_samples(
    Sample.Type == "SeqBlank" | Sample.Type=="SeqControl" 
  )
physeq_filt_seqcontrols #verify sample number

#Only controls
physeq_filt_extcontrols <- physeq_filt %>%
  subset_samples(
    Sample.Type == "ExtBlank" | Sample.Type=="ExtControl"
  )
physeq_filt_extcontrols #verify sample number

#Only replicates
physeq_filt_study <- physeq_filt %>%
  subset_samples(
    Sample.Type == "Study" 
  )
physeq_filt_study #verify sample number

# Scale reads to even depth 
physeq_scale <- physeq_filt %>% scale_reads(n=sample_depth) 
physeq_filt_seqcontrols <- physeq_filt_seqcontrols %>% scale_reads(n=sample_depth) 
physeq_filt_extcontrols <- physeq_filt_extcontrols %>% scale_reads(n=sample_depth) 
physeq_filt_study <- physeq_filt_study %>% scale_reads(n=sample_depth) 
```

#7 Alpha Diversity - All samples
```{r}
# Scale reads to even depth 
physeq_scale <- physeq_filt %>% scale_reads(n=1000) 

#Sample Type
p<- plot_richness(physeq_scale, x="Sample.Type") + geom_boxplot()
file_name =("Graphs/Baseline/sampletype_alphadiv.tiff")
tiff(file_name, width=800)
print(p)
dev.off()
```

#8 Alpha Metrics
```{r}
# Scale reads to even depth 
physeq_scale <- physeq_filt %>%
  scale_reads(n=1000) 

#Sequencing Batch
p<- plot_richness(physeq_scale, x="Run.ID") + geom_boxplot()
file_name =("Graphs/Baseline/alpha_sequencingbatch.tiff")
tiff(file_name, width=800)
print(p)
dev.off()

```

#9 Save data tables for downstream use
```{r}
phylos_obs <- list(physeq_filt_seqcontrols,physeq_filt_extcontrols,physeq_filt_study)
count=1

for (a in phylos_obs){
 # Extract abundance matrix from the phyloseq object
 OTU1 <- as(otu_table(a), "matrix") #Need a matrix of OTU information
 OTUdf <- as.data.frame(OTU1) # Coerce to data.frame
 tax_db <- as(tax_table(a), "matrix")#Need OTU to taxonomy database
 
 for (b in rownames(OTUdf)){ #Add genus name (or higher level of taxonomy) for each OTU
  OTUdf[b,"Genus"] <- tax_db[b,"Genus"]
 }
 
 for (c in rownames(OTUdf)){
  genus <- OTUdf[c,"Genus"]
  if(!grepl("g__",genus)){
    OTUdf[c,"Genus"]<-"HigherGenus"
  }
 }
 
 OTUdf<-aggregate(OTUdf[-ncol(OTUdf)],by=list(OTUdf$Genus),FUN="sum") #-64 to remove genus column, since it's not a col to sum
 OTUdf<-t(OTUdf) #transpose for downstream metadata matching
 colnames(OTUdf)<-OTUdf[1,]
 OTUdf<-OTUdf[-1,]
 
 for (d in colnames(OTUdf)){ #standardize genus naming
   if(grepl("g__",d)){
    colname_update <- str_remove(d, "g__")
    colname_update <- gsub("[","",colname_update,fixed=TRUE) #fixed = TRUE disables regex
    colname_update <- gsub("]","",colname_update,fixed=TRUE)
    colnames(OTUdf)[colnames(OTUdf)==d] <- colname_update
   }
 }
 
 colnames(OTUdf)[colnames(OTUdf)==""] <- "Unknown" #blank column (originally was "g__") needs to be renamed

 filename<-paste("OTUs/otu_baseline_",count,".csv",sep="")
 write.csv(OTUdf,filename)
 count=count+1
}

#File ID's
#1: Seq_controls
#2: Ext_controls
#3: Study
```
