---
title: "R Notebook"
output: word_document
editor_options: 
  chunk_output_type: console
---

#To Update
```{r}
project_name=c("Project_NP0084-MB4_5_pt6","Project_NP0084-MB4_5_pt6_2")
manifest_name=c("NP0084-MB4_08_29_19_metadata.txt","NP0084-MB4_09_02_19_metadata.txt")

sample_depth=100000
```

#Load required libraries
############### 
```{r}
library("biom")
library(ape)
library(data.table)
library(ggplot2)
library(tm)
library(vegan)
library("biom")
library(tidyr)
library(ggplot2)
library(cowplot)
library(scales)
library(phyloseq)
library(qiime2R)
library(tibble)
library(gridExtra)
source("sources/miseqR.R")
library(tidyverse)
source("sources/ggrare.R") #github library: https://rdrr.io/github/gauravsk/ranacapa/
library(ggplot2)
```

#Analysis
############### 
#1.Create PhySeq Object
```{r}
count=1
for (a in project_name){
 project_location = paste("T:\\DCEG\\Projects\\Microbiome\\CGR_MB\\MicroBiome\\",a,sep="")

 #Read OTUS
 otus<-read_qza(paste(project_location,"\\Output\\qza_results\\table_dada2_qza_merged_parts_final\\table_dada2_merged_final_filt.qza",sep=""))
 
 #Read rooted tree
 tree<-read_qza(paste(project_location,"\\Output\\qza_results\\phylogeny_qza_results\\rooted_tree.qza",sep=""))
 
 #Read Greengenes taxonomy file
 taxonomy<-read_qza(paste(project_location,"\\Output\\qza_results\\taxonomy_qza_results\\taxonomy_greengenes.qza",sep=""))
 
 #Edit dtable
 tax_table<-do.call(rbind, strsplit(as.character(taxonomy$data$Taxon), "; "))
   colnames(tax_table)<-c("Kingdom","Phylum","Class","Order","Family","Genus","Species")
   rownames(tax_table)<-taxonomy$data$Feature.ID
 
 #read metadata
 metadata<-read.table(paste(project_location,manifest_name[count],sep="\\"),sep='\t', header=T, row.names=1, comment="")
 
 #Create phylo object
 if(count==1){
  physeq_complete<-phyloseq(otu_table(otus$data, taxa_are_rows = T), phy_tree(tree$data), tax_table(tax_table), sample_data(metadata))
 
 } else{
  physeq_complete2<-phyloseq(otu_table(otus$data, taxa_are_rows = T), phy_tree(tree$data), tax_table(tax_table), sample_data(metadata))
  }
 count=count+1
}

#Create filtered list of samples
samplelist1_passed_q2<-as.list(sample_names(sample_data(physeq_complete)))
samplelist2_passed_q2<-as.list(sample_names(sample_data(physeq_complete2)))
interset<-intersect(samplelist1_passed_q2,samplelist2_passed_q2)

for (a in interset){ #Remove all duplicate samples
 physeq_complete2<-subset_samples(physeq_complete2,sample_names(physeq_complete2)!=a)
}
 
physeq_complete<-subset_samples(physeq_complete,External.ID!="Quarantined")#Remove quarantined samples

#Create merges of OTU, Tax, Sample for final phylo object
otu<-as.data.frame(otu_table(physeq_complete))
otu2<-as.data.frame(otu_table(physeq_complete2))

for (a in colnames(otu2)){
 for (b in rownames(otu2)){
  otu[b,a]<-otu2[b,a]
 }
}
otu<-otu_table(otu,taxa_are_rows = TRUE)

tax<-as.data.frame(tax_table(physeq_complete))
tax2<-as.data.frame(tax_table(physeq_complete2))

for (a in colnames(tax2)){
 for (b in rownames(tax2)){
  tax[b,a]<-tax2[b,a]
 }
}
tax<-as.matrix(tax) #Convert to matrix - str that phyloseq requires
tax<-tax_table(tax)

sampledata<-as.data.frame(sample_data(physeq_complete))
sampledata2<-as.data.frame(sample_data(physeq_complete2))

for (a in colnames(sampledata2)){
 for (b in rownames(sampledata2)){
  sampledata[b,a]<-sampledata2[b,a]
 }
}

physeq_complete<-phyloseq(otu,tax,sampledata) #merged phylo object
random_tree = rtree(ntaxa(physeq_complete), rooted=TRUE, tip.label=taxa_names(physeq_complete))
physeq_complete<-phyloseq(otu,tax,sampledata,random_tree) #merged phylo object


s<-summary(sample_data(physeq_complete))
capture.output(s, file = paste(project_location,"\\Output\\R\\Data\\Summary\\summary_prefilter.txt",sep=""))
```

#2 Prune for bacteria only, Create prefilter summary files
```{r}
#Only bacteria
physeq_filt<-physeq_complete %>%
  subset_taxa(
    Kingdom == "k__Bacteria" &
    Family  != "k__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Rickettsiales; f__mitochondria" &
    Class   != "k__Bacteria; p__Cyanobacteria; c__Chloroplast"
  )

#Print summaries
s<-summary(sample_data(physeq_filt)$Sample.Type)
capture.output(s, file = paste(project_location,"\\Output\\R\\Data\\Summary\\summary_prefilter_sample.type.txt",sep=""))
s<-summary(sample_data(physeq_filt)$Sample.Cat)
capture.output(s, file = paste(project_location,"\\Output\\R\\Data\\Summary\\summary_prefilter_sample.cat.txt",sep=""))
```

#3. Filter taxa >.001 ; Determine read counts and plot
```{r}
physeq_filt = filter_taxa(physeq_filt, function(x) mean(x) > 1e-2, TRUE)
physeq_filt #note new taxa number and sample number

# Make a data frame with a column for the read counts of each sample
sample_sum_df<-data.frame(sum = sample_sums(physeq_filt),sample_data(physeq_filt))
colnames(sample_sum_df)

# Histogram of sample read counts by sampletype
p1<-ggplot(sample_sum_df, aes(x = sum, fill=Sample.Type)) + 
  geom_histogram(binwidth = 2500) +
  xlab("Sequencing Reads") +
  ylab("Number of Samples") +
  ggtitle("Distribution of Sample Sequencing Depth - Pre-Filtering")

#Save file
file_name =paste(project_location,"\\Output\\R\\Graphs\\Summary\\SeqDepth_prefilter.tiff",sep="")
tiff(file_name, width=800)
print(p1)
dev.off()

#Read counts of blanks only
physeq_filt_blanks <- physeq_filt %>%
  subset_samples(
    Source.Descrip == "Water" 
  )
# Make a data frame with a column for the read counts of each sample
sample_sum_df<-data.frame(sum = sample_sums(physeq_filt_blanks),sample_data(physeq_filt_blanks))
colnames(sample_sum_df)

# Histogram of sample read counts by sampletype
p1<-ggplot(sample_sum_df, aes(x = sum, fill=Ext.Kit)) + 
  geom_histogram(binwidth = 2500) +
  xlab("Sequencing Reads") +
  ylab("Number of Samples") +
  ggtitle("Distribution of Sample Sequencing Depth - Pre-Filtering")

#Save file
file_name =paste(project_location,"\\Output\\R\\Graphs\\Summary\\SeqDepth_prefilter_blanks.tiff",sep="")
tiff(file_name, width=800)
print(p1)
dev.off()
```

#4. Filter samples with less than 10000 reads: Review filtered read counts and plot
```{r}
physeq_filt = prune_samples(sample_sums(physeq_filt) > 10000, physeq_filt)
physeq_filt #note new sample number

# Make a data frame with a column for the read counts of each sample
sample_sum_df<-data.frame(sum = sample_sums(physeq_filt),sample_data(physeq_filt))
colnames(sample_sum_df)

# Histogram of sample read counts by sample type
p1<-ggplot(sample_sum_df, aes(x = sum, fill=Sample.Type)) + 
  geom_histogram(binwidth = 2500) +
  xlab("Sequencing Reads") +
  ylab("Number of Samples") +
  ggtitle("Distribution of Sample Sequencing Depth - Post-Filtering")
  
#Save file
file_name =paste(project_location,"\\Output\\R\\Graphs\\Summary\\SeqDepth_postfilter.tiff",sep="")
tiff(file_name, width=800)
print(p1)
dev.off()

# Histogram of sample read counts by Run id - explain two clusters
p1<-ggplot(sample_sum_df, aes(x = sum, fill=Run.ID)) + 
  geom_histogram(binwidth = 2500) +
  xlab("Sequencing Reads") +
  ylab("Number of Samples") +
  ggtitle("Distribution of Sample Sequencing Depth - Post-Filtering")
#Save file
file_name =paste(project_location,"\\Output\\R\\Graphs\\Summary\\SeqDepth_postfilter_runid.tiff",sep="")
tiff(file_name, width=800)
print(p1)
dev.off()

#Print summaries
s<-summary(sample_data(physeq_filt))
capture.output(s, file = paste(project_location,"\\Output\\R\\Data\\Summary\\summary_postfilter.txt",sep=""))

s<-summary(sample_data(physeq_filt)$Sample.Type)
capture.output(s, file = paste(project_location,"\\Output\\R\\Data\\Summary\\summary_postfilter_sample.type.txt",sep=""))

s<-summary(sample_data(physeq_filt)$Sample.Cat)
capture.output(s, file = paste(project_location,"\\Output\\R\\Data\\Summary\\summary_postfilter_sample.cat.txt",sep=""))
```

#5. Subset filtered data, sample to specified depth
```{r}
#Only samples
physeq_filt_seqcontrols<-physeq_filt %>%
  subset_samples(
    Sample.Type=="Seq.Control" 
  )
physeq_filt_seqcontrols #verify sample number

#Only controls
physeq_filt_extcontrols<-physeq_filt %>%
  subset_samples(
    Sample.Type == "Ext.Blank" | Sample.Type=="Ext.Control"
  )
physeq_filt_extcontrols #verify sample number

#Only replicates
physeq_filt_study<-physeq_filt %>%
  subset_samples(
    Sample.Type == "Study" 
  )
physeq_filt_study #verify sample number

# Scale reads to even depth 
p1 <- ggrare(physeq_filt, step = 1000, color = "Sample.Type", se = FALSE)
p1 <- p1 + facet_wrap(~Sample.Type)
file_name =paste(project_location,"\\Output\\R\\Graphs\\Summary\\Rarecurve_PreFiltering.tiff",sep="")
tiff(file_name, width=800)
print(p1)
dev.off()
 
physeq_scale<-physeq_filt %>% scale_reads(n=sample_depth) 
physeq_filt_seqcontrols<-physeq_filt_seqcontrols %>% scale_reads(n=sample_depth) 
physeq_filt_extcontrols<-physeq_filt_extcontrols %>% scale_reads(n=sample_depth) 
physeq_filt_study<-physeq_filt_study %>% scale_reads(n=sample_depth) 

p1 <- ggrare(physeq_scale, step = 1000, color = "Sample.Type", se = FALSE)
p1 <- p1 + facet_wrap(~Sample.Type)
file_name =paste(project_location,"\\Output\\R\\Graphs\\Summary\\Rarecurve_PostFiltering.tiff",sep="")
tiff(file_name, width=800)
print(p1)
dev.off()
```

#6 Alpha Diversity - All samples, Run ID
```{r}
# Scale reads to even depth 
physeq_scale<-physeq_filt %>%
  scale_reads(n=1000) 

#Sample Type
p<-plot_richness(physeq_scale, x="Sample.Type") + geom_boxplot()
file_name =paste(project_location,"\\Output\\R\\Graphs\\Summary\\alphadiv_sampletype.tiff",sep="")
tiff(file_name, width=800)
print(p)
dev.off()

#Sequencing Batch
p<-plot_richness(physeq_scale, x="Run.ID") + geom_boxplot()
file_name =paste(project_location,"\\Output\\R\\Graphs\\Summary\\alphadiv_sequencingbatch.tiff",sep="")
tiff(file_name, width=800)
print(p)
dev.off()

#Extraction Kit
p<-plot_richness(physeq_scale, x="Ext.Kit") + geom_boxplot()
file_name =paste(project_location,"\\Output\\R\\Graphs\\Summary\\alphadiv_extractionkit.tiff",sep="")
tiff(file_name, width=800)
print(p)
dev.off()
```

#7 Save data tables for downstream use
```{r}
phylos_obs<-list(physeq_filt_seqcontrols,physeq_filt_extcontrols,physeq_filt_study)
count=1

for (a in phylos_obs){
 # Create a factor corresponding to the Genera
 genfac = factor(tax_table(a)[, "Genus"])
 # Tabulate the counts for each genera in each sample
 gentab = apply(otu_table(a), MARGIN = 2, function(x) {
    tapply(x, INDEX = genfac, FUN = sum, na.rm = TRUE, simplify = TRUE)
 })
 gentab<-as.data.frame(gentab)
 
 for (c in rownames(gentab)){
  genus<-c
  if(!grepl("g__",genus)){
   gentab[c,"Genus"]<-"HigherGenus"
  } else{
   colname_update<-str_remove(c, "g__")
   colname_update<-gsub("[","",colname_update,fixed=TRUE) #fixed = TRUE disables regex
   colname_update<-gsub("]","",colname_update,fixed=TRUE)
   gentab[c,"Genus"]<-colname_update
  }
 }
 
 gentab<-aggregate(gentab[-ncol(gentab)],by=list(gentab$Genus),FUN="sum") #- to remove genus column, since it's not a col to sum
 gentab[1,1]<-"Unknown" #blank column (originally was "g__") needs to be renamed
 gentab<-t(gentab) #transpose for downstream metadata matching
 colnames(gentab)<-gentab[1,]
 gentab<-gentab[-1,]
 
 file_name =paste(project_location,"\\Output\\R\\OTUs\\otu_Summary_",count,".csv",sep="")
 write.csv(gentab,file_name)
 
 metatab <- as.data.frame(sample_data(a))
 file_name =paste(project_location,"\\Output\\R\\OTUs\\metadata_Summary_",count,".csv",sep="")
 write.csv(metatab,file_name)
 count=count+1
}

#File ID's
#1: Seq_controls
#2: Ext_controls
#3: Study
```
