---
title: "R Notebook"
output: word_document
editor_options: 
  chunk_output_type: console
---
#1. Load required libraries
```{r}
library("biom")
library(data.table)
library(ggplot2)
```

#2. Taxonomy Reference - Expected frequencies
```{r}
#Read in controls taxonomy file, grouped by Genus
control_expect <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Controls\\Taxonomy\\Taxonomy_Genus_2019.txt",sep="\t",header=TRUE)
rownames(control_expect) <- unique(control_expect[,"Genus"])
```

#3. Filtering of samples to include project set only
```{r}
#Read in database with variable information
metadata_baseline_complete <- read.table("T:\\DCEG\\Projects\\Microbiome\\CGR_MB\\MicroBiome\\Project_NP0084-MB4_5_pt6\\NP0084-MB4_08_29_19_metadata.txt",sep="\t",header = TRUE)

#Read in database of failed samples
samplelist_baseline_failed <- read.table("T:\\DCEG\\Projects\\Microbiome\\CGR_MB\\MicroBiome\\Project_NP0084-MB4_5_pt6\\samplelist_baseline_failed.txt")

#Remove failed samples from info database
metadata_baseline_filtered <-subset(metadata_baseline_complete, !(SampleID %in% samplelist_baseline_failed$V1))
nrow(metadata_baseline_complete) - nrow(samplelist_baseline_failed) ==nrow(metadata_baseline_filtered) #should be true

#Only include Seq Controls
control_types <- c("SeqControl","SeqBlank")
metadata_baseline_filtered <- metadata_baseline_filtered[metadata_baseline_filtered$Sample.Type %in% control_types,]
row.names(metadata_baseline_filtered) <- metadata_baseline_filtered$SampleID

samplelist_baseline_filtered <- unique(metadata_baseline_filtered$SampleID)

nrow(metadata_baseline_filtered)
```

#4. OTU Counts by sample
Read in OTU information, and determine total OTUs
```{r}
#GreenGenes
otu_baseline_complete <- read.csv("T:\\DCEG\\Projects\\Microbiome\\CGR_MB\\MicroBiome\\Project_NP0084-MB4_5_pt6\\Output\\otu_relative_abundance_results\\qiime2_taxa_counts_genus_green.csv", row.names = 1)
otu_baseline_complete <- read.csv("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Analysis\\ExpA_Run1&2\\TaxaCounts\\qiime2_taxa_counts_genus_green_abrf.csv", row.names = 1)

otu_baseline_complete <- t(otu_baseline_complete) #table has samples as columns, not rows
otu_baseline_filtered <- subset(otu_baseline_complete, !(SampleID %in% samplelist_baseline_filtered$V1))
genus_found <- colnames(otu_baseline_filtered)

#For each sample, determine the total read counts
for (a in samplelist_baseline_filtered){
 #Set sum to 0
 sum=0
 
 #For each genus
 for (b in genus_found){
  
  #Add the number of counts together
  sum <- otu_baseline_filtered[a,b] + sum
 }
 
 #When complete, push the sum to a new "Total_OTUs" row
 otu_baseline_filtered[a,"Total_OTUs"] <- sum
}
```

#5 Relative Abundance - Blanks
```{r}
samplelist_baseline_water <- subset(metadata_baseline_filtered, (metadata_baseline_filtered$Source.Descrip == "Water"))
samplelist_baseline_water <- (samplelist_baseline_water$SampleID)

for (a in samplelist_baseline_water){
 
 for (b in ncol(otu_baseline_filtered)-1){
   relab <- otu_baseline_filtered[a,b]
   
   if(!(is.na(relab) | is.null(relab))){
     relab_baseline_blanks[a,b] <- otu_baseline_filtered[a,b]/otu_baseline_filtered[a,"Total_OTUs"]
   }

 }
 
}

write.csv(relab_baseline_blanks,"/datatables/relabun_baseline_blanks.csv")

```

#6. Genus Filter List - From Seq Blanks
```{r}
genus_sequence_remove <- c("")

```

#7. Observed to Expected Counts - Seq Controls
```{r}
samplelist_baseline_seqcontrol <- subset(metadata_baseline_filtered, (metadata_baseline_filtered$Sample.Type == "SeqControl"))

subjectlist_baseline_seqcontrol <- subset(metadata_baseline_filtered, (metadata_baseline_filtered$Sample.Type == "SeqControl"))
subjectlist_baseline_seqcontrol <- unique(subjectlist_baseline_seqcontrol$SubjectID)

observed_count =0

#Individual Samples
for (a in subjectlist_baseline_seqcontrol){
 genus_temp <- subset(control_expect, !(control_expect[,a] == "A"))
 genus_temp <- genus_temp$Genus
  
 for (b in samplelist_baseline_seqcontrol){
  subject.id <- metadata_baseline_filtered[b,"SubjectID"]
  
  if(a==subject.id){
   
   for (c in genus_temp){
    otu_count <- otu_baseline_filtered[b,c]
    
    if(!(is.na(observed_count) | is.null(otu_count) | otu_count ==0)){
     observed_count = observed_count+1
    }
    
   }
   
  }
  
  observed_baseline_seqcontrols[b,"Observed"] <- observed_count
  observed_baseline_seqcontrols[b,"Expected"] <- length(genus_temp)
  observed_baseline_seqcontrol[b,"SubjectID"] <- subject.id
  observed_count =0
 
 }
 
}

write.csv(observed_baseline_seqcontrols,"/datatables/observed_baseline_seqcontrols.csv")

#Average by SubjectID
count =0
for (a in subjectlist_baseline_seqcontrol){
 for (b in samplelist_baseline_seqcontrol){
  subject.id <- observed_baseline_seqcontrols[b,"SubjectID"]
  
  if(a=subject.id){
   observed_count<-observed_baseline_seqcontrols[b,"Observed"]+observed_count
   expected_count<-observed_baseline_seqcontrols[b,"Expected"]
   count = count+1
  }
 }
 observed_baseline_seqcontrols[a,"Observed"]<-observed_count/count
 observed_baseline_seqcontrols[a,"Expected"]<-expected_count
 observed_baseline_seqcontrol[a,"SubjectID"]<-a
 count =0
}

write.csv(observed_baseline_seqcontrols,"/datatables/observed_baseline_seqcontrols.csv")

```

#8. Relative Abundance  - Seq Controls
```{r}
#Complete Relative Abundance
for (a in subjectlist_baseline_seqcontrol){
 
 for (b in ncol(otu_baseline_filtered)-1){
   relab <- otu_baseline_filtered[a,b]
   
   if(!(is.na(relab) | is.null(relab))){
     relab_baseline_seqcontrols[a,b] <- otu_baseline_filtered[a,b]/otu_baseline_filtered[a,"Total_OTUs"]
   }

 }
 
}
write.csv(relab_baseline_seqcontrols,"/datatables/relabun_baseline_seqcontrols.csv")

#Expected relative abundance
for (a in subjectlist_baseline_seqcontrol){
 genus_temp <- subset(control_expect, !(control_expect[,a] == "A"))
 genus_temp <- genus_temp$Genus
  
 for (b in samplelist_baseline_seqcontrol){
  subject.id <- metadata_baseline_filtered[b,"SubjectID"]
  
  if(a==subject.id){
   
   for (c in genus_temp){
    rel_abund_observed <- relab_baseline_seqcontrols[b,c]
    rel_abund_expected <- control_expect[c,a]
    gram_stain <- control_expect[c,"Gram"]
    
    relab_baseline_seqcontrols_expected[b,paste(c,".Obs",sep="")]<-rel_abund_observed
    relab_baseline_seqcontrols_expected[b,paste(c,".Exp",sep="")]<-rel_abund_expected
    relab_baseline_seqcontrols_expected[b,"GramStain"]<-gram_stain
    relab_baseline_seqcontrols_expected[b,"SubjectID"]<-a
   }
   
  }

 }
 
}

write.csv(relab_baseline_seqcontrols_expected,"/datatables/relab_baseline_seqcontrols_expected.csv")

#Average by SubjectID

count=0
observed_count=0

for (a in subjectlist_baseline_seqcontrol){
 genus_temp <- subset(control_expect, !(control_expect[,a] == "A"))
 genus_temp <- genus_temp$Genus
 
 for (b in genus_temp){
  
  for (c in samplelist_baseline_seqcontrol){
   
   subject.id <- relab_baseline_seqcontrols_expected[b,"SubjectID"]
   
   if(a=subject.id){
    observed_count<-relab_baseline_seqcontrols_expected[c,paste(b,".Obs",sep="")]+observed_count
    expected_count<-relab_baseline_seqcontrols_expected[c,paste(b,".Exp",sep="")]
    gram_stain <- relab_baseline_seqcontrols_expected[c,"GramStain"]
    count = count+1
    
   }
   
  }
  
  relab_baseline_seqcontrols_expected[a,"Observed"]<-observed_count/count
  relab_baseline_seqcontrols_expected[a,"Expected"]<-expected_count
  relab_baseline_seqcontrols_expected[a,"GramStain"]<-gram_stain
  relab_baseline_seqcontrols_expected[a,"SubjectID"]<-a
  count=0
  observed_count=0
 }
}
  
write.csv(relab_baseline_seqcontrols_expected,"/datatables/relab_baseline_seqcontrols_expected.csv")

```

#9 Relative Abundance % Difference
```{r}
for (a in subjectlist_baseline_seqcontrol){
 
 genus_temp <- subset(control_expect, !(control_expect[,a] == "A"))
 genus_temp <- genus_temp$Genus
  
 for (b in genus_temp){
  
  
 }
 observed_count<-relab_baseline_seqcontrols_expected[a,"Observed"]
 expected_count<-relab_baseline_seqcontrols_expected[a,"Expected"]
 
 if(!(is.na(observed_count)|is.null(observed_count)|is.na(observed_count)|is.null(observed_count)|observed_count=0|expected_count=0))

}

```

#5. OTU Counts by Control Type
Determine genus count for each sample, average across control type
```{r}
#Create a missing list for non-annotated Genus
missing_anno <- list()

##For each control type
for (a in control_types){
 
 ##For each sample in filtered list
 for (b in samplelist_baseline_filtered){
  
  #Set count to 0
  count_present = 0
  count_absent = 0
  
  #Create control type groups
  valid <- metadata_baseline_filtered[b,"Type"]

  if (valid==a){
   #Create genus list of specfic control
   genus_temp <- control_expect[!is.na(control_expect[,a]),]
   genus_temp <- (genus_temp[,"Genus"])
   
   #For each genus is generated list, determine number of reads
   for (c in genus_temp){
    value <- otu_baseline_complete[c,b]
  
    #If value is NA (meaning that the genus was not found in sample), add to the absent count
    #If reads are >0 then count_presence increases by 1
    #If reads are =0 then count_absence increases by 1 
  
    if(is.na(value) | is.null(value)) {
     missing_anno[[c]] <- "No Annotations"
     count_absent = count_absent +1
     next
    } else if (value>0) {
     count_present = count_present +1
    } else if (value==0){
     count_absent = count_absent +1
    } else{
     next
     }
   }
   
   #Add final count and length of the list to metadata_baseline_filtered dataframe
   metadata_baseline_filtered[b,"GenusObs"] <- count_present
   metadata_baseline_filtered[b,"GenusAbs"] <- count_absent
   metadata_baseline_filtered[b,"GenusExp"] <- length(genus_temp)
   
  } else{ 
   genus_temp <- data.frame()
  }
 }
}

#List of Genus that no samples contained
missing_anno

###Testing
write.csv(metadata_baseline_filtered,"temp1.csv")

#New Database to store final Genus information
samples_obs_count_type <- data.frame()

#For each of the control types
for (a in control_types){
 
 #Reset counts
 genus=0
 count=0
 
 #For each of the samples in the sample list
 for(b in samplelist_baseline_filtered){
  
  #Create control type groups
  valid <- metadata_baseline_filtered[b,"Type"]
  
  #If the current samples control type matches the control type of the iteration
  if(valid==a){
   
   #Count the number of genus in the sample
   genus <- genus + metadata_baseline_filtered[b,"GenusObs"]
   count <- count + 1
  } else{
    next
  }
  #Find number of genus expected
  genus_expect_count <- metadata_baseline_filtered[b,"GenusExp"]
 }
 
 #Find the average of the genus and totals
 genus_average <- genus/count

 #Push the totals to the new database 
 samples_obs_count_type[1,a] <- genus_average
 samples_obs_count_type[2,a] <- genus_expect_count

 #Add Row names
 row.names(samples_obs_count_type) <- c("Genus Count", "Expected")
}

###Testing
write.csv(samples_obs_count_type,"temp2.csv")
```

#6. OTU Percent - Analysis Genus OTU Abundance
```{r}
#Create new table
samples_obs_abund = data.frame()

#Determine which genus_expect to use
##Run through each sample
for (a in samplelist_baseline_filtered){
 
 #Check for each of the expected genus
 for (b in genus_expect){
  
  #Determine the reads for the genus, and the total reads for the sample
  reads <- otu_baseline_complete[b,a]
  total <- otu_baseline_complete["Total_OTUs",a]

  #If value is NA (meaning that the genus was not found in sample), skip
  #If reads are >0 then add the read count to reads_total

  if(is.na(reads) | is.null(reads)) {
   next
  } else {
   percent_exp <- (reads/total)
   
   #Add percent to the samples_obs_abund database
   samples_obs_abund[b,a] <- percent_exp
  }
 }
}

#Output the table of all samples
write.csv(samples_obs_abund,"Data/seq_obs_abund.csv")
```

#7. OTU Percent - Analysis Genus OTU Abundance by Control Type
```{r}
#New Database to store final Genus information
samples_obs_abund_vari <- data.frame()

#For each of the control types
for (a in control_types){
 
  #For each Genus in list
  for (b in genus_expect){
   
   #Reset Counter
   reads <- 0
   total <- 0
   
   #Create a list of samples that are of the specified control type
   sample_list_temp<- subset(metadata_baseline_filtered, Type==a)
   sample_list_temp <- unique(sample_list_temp$SampleID)
   
   #For each of the samples in the sample list
   for(c in sample_list_temp){
    
    #Add up the total percentages from samples_obs_abund [genus,sampleID]
    reads <- otu_baseline_complete [b,c] + reads
    total <- otu_baseline_complete["Total_OTUs",c] + total
   }
   
   #average of genus level by taking total number of reads of that genus divded by total number of reads for samples included
   #Push value to new dataframe
   percent_obs <- reads / total
   samples_obs_abund_vari[b,a] <- percent_obs
  }
 }

#Output to text file
write.csv(samples_obs_abund_vari,"Data/seq_obs_abund_vari.csv")
```
**********************
#9. OTU Percent - Analysis Exp OTU Rel Abun % Diff by Control
```{r}
#Add expected frequencies and gram positive status to data table
for (a in control_types){
 
 #For each of the genus
 for (b in genus_expect){
  
  #Add gram pos/neg information
  samples_obs_abund_vari[b,"Gram"] <- control_expect[b,10]
  
  #Find value and pass to database
  value <- control_expect[b,a]
  
  if(is.na(value) | is.null(value)){
  } else{
   #Create new column
   col_name <- paste(a,"_Exp",sep="")
   samples_obs_abund_vari[b,col_name] <- value
  }
  }
}

#Output to text file
write.csv(samples_obs_abund_vari,"Data/seq_obs_abund_vari.csv")

#Start counter
count=1

#For each of the Seq Controls
for (i in 1:6){
 
 #For each genus
 for (a in genus_expect){
 
  #Find the value for the observe rel abund and the expected rel abund
  value1 <- samples_obs_abund_vari[a,i]
  value2 <- samples_obs_abund_vari[a,i+7]
  
  #Calculate percent difference
  if(is.na(value1) | is.null(value1)| is.na(value2) | is.null(value2)){
   value <- "NA"
  } else{
   value <- ((value1 - value2)/ value2)
  } 
  
  #Create new column name
  col_name <- paste((colnames(samples_obs_abund_vari[i])),"-Df",sep="")
  
  #Pass values
  samples_obs_abund_vari[a,col_name]<- value
  
 }
 
 #Move the counter to the next control type
 if(count==length(control_types)){
   count=1
 } else{count=count+1}
}

#Output to text file
write.csv(samples_obs_abund_vari,"Data/seq_obs_abund_vari.csv")
```

#10. OTU Percent - Analysis Unexp OTU Rel Abun by Control
```{r}
#Create full list of genus present
genus_present <- rownames(otu_baseline_complete)
genus_expect<-as.character(genus_expect)

#Create list of unexpected
genus_unexpect <- genus_present[which(!genus_present %in% genus_expect)]

#Remove Total_OTUs from list
genus_unexpect <- genus_unexpect[genus_unexpect !="Total_OTUs"]

#Create new dataframe
samples_obs_abund_vari_unexp <- data.frame()

#For each of the control types
for (a in control_types){
 
  #For each Genus in list
  for (b in genus_unexpect){
   
   #Reset Counter
   reads <- 0
   total <- 0
   
   #Create a list of samples that are of the specified control type
   sample_list_temp<- subset(metadata_baseline_filtered, Type==a)
   sample_list_temp <- unique(sample_list_temp$SampleID)
   
   #For each of the samples in the sample list
   for(c in sample_list_temp){
    
    #Add total reads from otu_baseline_complete [genus,sampleID]
    reads <- otu_baseline_complete [b,c] + reads
    total <- otu_baseline_complete["Total_OTUs",c] + total
   }
   
   #Calc Rel abund (total number of reads / total number of reads)
   #Push value to new dataframe
   percent_obs <- reads / total
   samples_obs_abund_vari_unexp[b,a] <- percent_obs
  }
}

#Remove repeated Clostriudium row
samples_obs_abund_vari_unexp<- samples_obs_abund_vari_unexp[!rownames(samples_obs_abund_vari_unexp) %in% "Clostridium sensu stricto 1", ]

#Output to text file
write.csv(samples_obs_abund_vari_unexp,"Data/seq_obs_abund_vari_unexp.csv")
```

#11 Plot graph - Actual to Expected - IntraControl Variability
For each kit, show variability, and map
```{r}
#Create list for plots, start counter
plot_list <- list()
i=1

for (a in control_types){
   #Create genus list of specfic control
   genus_temp <- control_expect[!is.na(control_expect[,a]),]
   genus_temp <- (genus_temp[,"Genus"])
   
   #Create a list of samples that are of the specified control type
   sample_list<- subset(metadata_baseline_filtered, Type==a)
   sample_list <- unique(sample_list$SampleID)

   #Create dataframe
   temp<-data.frame()
   count=1
   
   #For each Genus in list, determine the samples value
   for (b in genus_temp){
    
     #For each of the samples in the sample list
     for(c in sample_list){
      temp[count,"Genus"]<-b
      temp[count,"Obs_Per"]<- samples_obs_abund[b,c]
      temp[count,"Sample"]<-c
      temp[count,"Exp_Per"]<-control_expect[b,a]
      count=count+1
     }
   }
  
    graph_title <- paste (a,": Intracontrol Variablity",sep = "")

    #Plot Data
    graph_new <- ggplot(data=temp,mapping=aes(x=Obs_Per,y=Exp_Per,color=Sample))+
     geom_point()+ #Can add points to graph
     geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
     labs(title=graph_title, x="Observed Rel. Abundance", y= "Expected Rel. Abundance")+
     geom_text(mapping=aes(x=Obs_Per,y=Exp_Per,label=Genus),check_overlap = TRUE)
     
    
    plot_list[[i]] = graph_new
    i=i+1
}

#Save all plots
for (i in 1:length(control_types)){
 
 #ID which control is being run
 a <- control_types[i]
 
 #Save the file
 file_name = paste("Data/Graphs/Intracontrol Variability_",a, ".tiff", sep="")
 tiff(file_name)
 print(plot_list[[i]])
 dev.off()
}
```

#12 Plot graph - Actual to Expected - One graph
For each kit, show variability, and map
```{r}
#Create list for plots, start counter
temp<-data.frame()
count<-1
for (a in control_types){
   #Create genus list of specfic control
   genus_temp <- control_expect[!is.na(control_expect[,a]),]
   genus_temp <- (genus_temp[,"Genus"])
   
   #Create a list of samples that are of the specified control type
   sample_list<- subset(metadata_baseline_filtered, Type==a)
   sample_list <- unique(sample_list$SampleID)

   #For each Genus in list, determine the samples value
   for (b in genus_temp){
    
     #For each of the samples in the sample list
     for(c in sample_list){
      temp[count,"Genus"]<-b
      temp[count,"Obs_Per"]<- samples_obs_abund[b,c]
      temp[count,"Sample"]<-c
      temp[count,"Exp_Per"]<-control_expect[b,a]
      temp[count,"Type"] <- a
      count=count+1
     }
   }
}

#Plot graph
graph_new<- ggplot(data=temp,mapping=aes(x=Obs_Per,y=Exp_Per,color=Type))+
 #geom_point()+ #Can add points to graph
 geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
 labs(title="Sequencing Control Comparison", x="Observed Rel. Abundance", y= "Expected Rel. Abundance")
graph_new
write.csv(temp,"temp2.csv")

file_name =("Data/Graphs/Kit Differences_ZE.tiff")
tiff(file_name)
print(graph_new)
dev.off()


```