---
title: "R Notebook"
output: word_document
editor_options: 
  chunk_output_type: console
---

#To Update
```{r}
output_location<-c("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Analysis\\NP0084-MB4,5,6\\R_SeqVariation\\")

taxa_levels<-c("Genus")
reference_db<-c("greengenes")

manifest_name<-paste("metadata_",taxa_levels,"_",reference_db,"_seqcontrol.csv",sep="")
taxa_file<-paste("taxa_",taxa_levels,"_",reference_db,"_seqcontrol.csv",sep="")

control_location<-c("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Controls\\Taxonomy\\")
```

#Load required libraries
############### 
```{r}
library("ape")
library("biom")
library(cowplot)
library(data.table)
library(ggplot2)
library(grid)
library(gridExtra)
library("igraph")
library(qiime2R)
library(phyloseq)
library(scales)
library(structSSI)
library(tibble)
library(tidyverse)
library(tidyr)
library(tm)
library(vegan)
library(Metrics)
```

#Analysis
###############
#Taxonomy Reference - Expected frequencies
```{r}
#Read in controls taxonomy file, grouped by taxa
for (a in taxa_levels){
 control_expect <- read.table(paste(control_location,"Taxonomy_",taxa_levels,"_2019.txt",sep=""),sep="\t",header=TRUE)
 rownames(control_expect) <- unique(control_expect[,taxa_levels])
 
 control_db_name<-paste("expected_taxa_",a,sep="")
 assign(control_db_name,control_expect)
 remove(control_expect)
}

```

#Taxa and metadata tables
```{r}
for(a in taxa_levels){
 for (b in reference_db){
  metadata_db_name<-paste("metadata_",a,"_",b,sep="")
  
  temp <- read.table(paste(output_location,"Taxa\\",metadata_db_name,"_seqcontrol.csv",sep=""),sep=",",header= TRUE)
  rownames(temp)<-temp$X
  temp<-temp[-1]
  
  assign(metadata_db_name,temp)
  remove(temp)
  
  taxa_db_name<-paste("taxa_",a,"_",b,sep="")
  
  temp<-read.table(paste(output_location,"Taxa\\",taxa_db_name,"_seqcontrol.csv",sep=""),sep=",",header= TRUE)
  rownames(temp)<-temp$X
  temp<-temp[-1]
  
  assign(taxa_db_name,temp)
  remove(temp)
  
  temp <- read.table(paste(output_location,"OTUs\\otu_table.csv",sep=""),sep=",",header= TRUE)
  rownames(temp)<-temp$X
  temp<-temp[-1]
  
  assign("otus",temp)
  remove(temp)
 }
}
```

#Load phylo object
```{r}
#NOTE: must run 1. Initial processing
#physeq_scale include all samples

subset_list<-c("Seq.Control")
physeq_scale_sub<-subset_samples(physeq_scale, Sample.Type %in% subset_list)

```

#OTU Counts by sample
Determine total OTUs
```{r}
taxa_Genus_greengenes$Total_OTUs <- rowSums(taxa_Genus_greengenes)

```

#Observed to Expected Counts
```{r}
tp_fp_fn_samplelevel<- function(output,db,tl,ct){
 sample_list<-intersect(rownames(get(paste("taxa",tl,db,sep="_"))),rownames(subset(get(paste("metadata",tl,db,sep="_")),Source.Descrip==ct)))
 exp_taxa_list<-rownames(get(paste("expected_taxa",tl,sep="_"))[ get(paste("expected_taxa",tl,sep="_"))[[ct]] !='A' , ])
 
 unexp_taxa_list<-setdiff(colnames(get(paste("taxa",tl,db,sep="_"))),exp_taxa_list)
 remove_list<-c("Total_OTUs","Unknown","HigherGenus","HigherFamily") #Remove Total_OTUs column, Higher Genera col, Unknown
 unexp_taxa_list <- unexp_taxa_list[! unexp_taxa_list %in% remove_list ]
 
 for (a in sample_list){
  fp=0
  fn=0
  tp=0
  
  taxa_db<-get(paste("taxa",tl,db,sep="_"))
  
  #Run through expected taxa list
  for (b in exp_taxa_list){
   otu_count<-taxa_db[a,b]

   
   if(is.null(otu_count) || otu_count==0 || is.na(otu_count)){
    fn=fn+1

   } else if (otu_count>0){
    tp=tp+1
   }
  }
  
  #Run through all taxa not expected
  for (b in unexp_taxa_list){
   otu_count<-taxa_db[a,b]
   
   if(otu_count==0){
    next;
   } else if (otu_count>0){
    fp=fp+1
   }
  }
  
  output[a,"ControlType"]<-ct
  output[a,"Expected"]<-length(exp_taxa_list)
  output[a,paste("TP",db,tl,sep="_")]<-tp
  output[a,paste("FP",db,tl,sep="_")]<-fp
  output[a,paste("FN",db,tl,sep="_")]<-fn
  output[a,paste("TAR",db,tl,sep="_")]<-tp/(tp+fp)
  output[a,paste("TDR",db,tl,sep="_")]<-tp/(tp+fn)
  output[a,paste("F1",db,tl,sep="_")]<-2*((tp/(tp+fp))*(tp/(tp+fn)))/((tp/(tp+fp))+(tp/(tp+fn)))
 }
 
 return(output)
}

#Determine which controls
control_type<-c("MSA1000","MSA1001","MSA1002","MSA1003","D6305","D6306","D6311")

tar_tdr<-data.frame()
for(a in reference_db){
 for (b in taxa_levels){
  for (c in control_type){
   tar_tdr<-tp_fp_fn_samplelevel(tar_tdr,a,b,c)
  }
 }
}

write.csv(tar_tdr,paste(output_location,"\\Taxa\\Taxonomy_tar_trd_seq.csv",sep=""))

```

#Average by Control
```{r}
tar_tdr_av<-aggregate(tar_tdr[, 2], list(tar_tdr$ControlType), mean) #Expected tax
row.names(tar_tdr_av)<-tar_tdr_av$Group.1
tar_tdr_av<-tar_tdr_av[,-1]

tar_tdr_av<-cbind(tar_tdr_av,aggregate(tar_tdr[, 6], list(tar_tdr$ControlType), mean)) #TAR Average
row.names(tar_tdr_av)<-tar_tdr_av$Group.1
tar_tdr_av<-tar_tdr_av[,-3]

tar_tdr_av<-cbind(tar_tdr_av,aggregate(tar_tdr[, 7], list(tar_tdr$ControlType), mean)) #TDR Average
row.names(tar_tdr_av)<-tar_tdr_av$Group.1
tar_tdr_av<-tar_tdr_av[,-3]

tar_tdr_av<-cbind(tar_tdr_av,aggregate(tar_tdr[, 8], list(tar_tdr$ControlType), mean)) #F1 Average
tar_tdr_av<-tar_tdr_av[,-4]

colnames(tar_tdr_av)<-c("Expected","TAR","TDR","F1Score")
tar_tdr_av

write.csv(tar_tdr_av,paste(output_location,"\\Taxa\\Taxonomy_tar_trd_seq_av.csv",sep=""))
```

#Relative Abundance  - Seq Controls
```{r}
relabun_seqcont<-data.frame()

#Complete Relative Abundance
sample_list_seqcont<-rownames(taxa_Genus_greengenes)
taxa_list <- colnames(taxa_Genus_greengenes)
taxa_list<-taxa_list[-(length(taxa_list))] #Remove total OTU's from list
taxa_list<-unique(append(taxa_list,rownames(expected_taxa_Genus)))

for (a in sample_list_seqcont){
 relabun_seqcont[a,"Source.Descrip"]<-as.character(metadata_Genus_greengenes[a,"Source.Descrip"])
 
 for (b in taxa_list){
  relabun <- taxa_Genus_greengenes[a,b]
  
  if(!(is.na(relabun) || is.null(relabun) || relabun==0)){
   relabun_seqcont[a,b] <- taxa_Genus_greengenes[a,b]/taxa_Genus_greengenes[a,"Total_OTUs"]
  } else{
   relabun_seqcont[a,b] <- 0
  }
 }
}

write.csv(relabun_seqcont,paste(output_location,"Data\\relabun_seqcont.csv",sep=""))

#Calculate averages
relabun_seqcont_av<-sort(unique(relabun_seqcont$Source.Descrip))

for (a in taxa_list){
 
 temp<-aggregate(relabun_seqcont[,a], list(relabun_seqcont$Source.Descrip), FUN = function(x) mean(as.numeric(as.character(x))))
 relabun_seqcont_av<-cbind(relabun_seqcont_av,temp[,2])
}

#Convert to df, add taxa as colnames
relabun_seqcont_av<-as.data.frame(relabun_seqcont_av)
row.names(relabun_seqcont_av)<-relabun_seqcont_av[,1] #col one is Source.Descrip
relabun_seqcont_av<-relabun_seqcont_av[,-1]
colnames(relabun_seqcont_av)<-taxa_list
relabun_seqcont_av[] <- lapply(relabun_seqcont_av, function(x) {
    as.numeric(as.character(x))})

#add expected rel abundance
temp<-data.frame()
for (a in rownames(relabun_seqcont_av)){
 for (b in taxa_list){
  temp[paste(a,"_exp",sep=""),b]<-as.numeric(as.character(expected_taxa_Genus[b,a]))
 }
}

temp[temp=="A" ]=0 #replace A's with 0
temp[is.na(temp)]=0 #replace A's with 0
temp<-as.data.frame(temp)
relabun_seqcont_av<-rbind(relabun_seqcont_av,temp)

write.csv(relabun_seqcont_av,paste(output_location,"Data\\relabun_seqcont_av.csv",sep=""))
```

#Relative Abundance  - Seq Controls
```{r}
relabun_seqcont<-data.frame()

#Complete Relative Abundance
sample_list_seqcont<-rownames(taxa_Genus_greengenes)
taxa_list <- colnames(taxa_Genus_greengenes)
taxa_list<-taxa_list[-(length(taxa_list))] #Remove total OTU's from list
taxa_list<-append(taxa_list,rownames(expected_taxa_Genus))

for (a in sample_list_seqcont){
 for (b in taxa_list){
  
  relabun <- taxa_Genus_greengenes[a,b]
  
  if(!(is.na(relabun) || is.null(relabun) || relabun==0)){
   relabun_seqcont[a,b] <- taxa_Genus_greengenes[a,b]/taxa_Genus_greengenes[a,"Total_OTUs"]
   relabun_seqcont[a,"Source.Descrip"]<-as.character(metadata_Genus_greengenes[a,"Source.Descrip"])
  }
  control_type<-as.character(metadata_Genus_greengenes[a,"Source.Descrip"])
  relabun_seqcont[control_type,b]<-as.character(expected_taxa_Genus[b,control_type])
  relabun_seqcont[control_type,"Source.Descrip"]<-control_type
 }
}

relabun_seqcont[is.na(relabun_seqcont)]=0 #replace na's with 0
relabun_seqcont[relabun_seqcont=="A"]=0 #replace A's with 0
write.csv(relabun_seqcont,paste(output_location,"Data\\relabun_seqcontrols.csv",sep=""))

```

#Plot graph - IntraControl Variability
For each kit, show variability, and map
```{r}
#Plot type can be: c("DCA", "CCA", "RDA", "CAP", "DPCoA", "NMDS", "MDS", "PCoA")
#dist_meas can be: "unifrac", "wunifrac", "dpcoa", "jsd", "bray"

plot_controls<-function(type,dist_meas){
 ordu = ordinate(physeq_scale_sub, type, dist_meas, weighted=TRUE)
 p1<-plot_ordination(physeq_scale_sub, ordu, color="Source.Descrip",title=paste(type," of Seq Controls",sep=""))
 
 allGroupsColors<- c("grey0", "grey50", "deepskyblue","red", "darkred", "orange", "purple")
 p1<-p1 + geom_point(size = 4) + scale_color_manual(values = allGroupsColors)


 tiff(paste(output_location,"\\Graphs\\",type,"_seqcontrols.tiff",sep=""),compression="lzw")
 grid.draw(p1)
 dev.off()
 
}

plot_controls("NMDS","bray")
plot_controls("PCoA","unifrac")


```

#Stats on rel abund
```{r}
#From PAPER example
#https://www.rdocumentation.org/packages/structSSI/versions/1.1.1/topics/Adaptive.GBH
relabun_sig<-function(phyobj,taxa_level,environments,ctname){
 temp_phyobj<- prune_taxa(taxa_sums(phyobj)>0,phyobj) #remove taxa that have no samples with reads - meant for subsetted samples
 
 taxaPValues <- numeric(length = ntaxa(temp_phyobj))
 names(taxaPValues) <- taxa_names(temp_phyobj)
 abundances <- otu_table(temp_phyobj)
 
 for (taxaIndex in 1:ntaxa(temp_phyobj)) {

  abundModel <- summary(lm(as.vector(abundances[taxaIndex, ]) ~environments))
  taxaPValues[taxaIndex] <- with(abundModel, pf(fstatistic[1],fstatistic[2], fstatistic[3], lower.tail = FALSE))
 }
 
 chlamydiae.families <- na.omit(tax_table(temp_phyobj)[, taxa_level])
 taxaPValues <- taxaPValues[taxa_names(chlamydiae.families)]
 family.AGBH <- Adaptive.GBH(unadj.p = taxaPValues, group.index = matrix(chlamydiae.families), method = "lsl")
 write.csv(family.AGBH@p.vals,paste(output_location,"\\Data\\","ABGH_",ctname,".csv",sep=""))
 
 #Plotting
 sub.AGBH<-family.AGBH@p.vals
 c<-sub.AGBH[grepl("g__", sub.AGBH$group),]#remove all not at taxa level
 plot_AGBH(sub.AGBH,"all",ctname)
 
 #Sig Plotting
 sub.AGBH<-subset(sub.AGBH,sub.AGBH$adjp<.05)
 #If there are sig results, plot separately
 if(nrow(sub.AGBH)>0){
  plot_AGBH(sub.AGBH,"sig",ctname)
 }
}

plot_AGBH<-function(datasub,name,ct){
  sub.AGBH<-datasub[-4] #Remove sig col
  sub.AGBH<-sub.AGBH[-1] #remove unadjp
  melt.AGBH <- melt(sub.AGBH, id.vars="group")
  
  p1<-ggplot(melt.AGBH, aes(1:nrow(melt.AGBH), value, col=group,shape=variable)) + 
   geom_point(size=2) +
   ylab("Adjusted p-values") +
   scale_x_discrete(name ="Hypothesis sorted by adjusted p-values", 
                     limits=c(1:nrow(melt.AGBH))) + 
   theme(axis.text.x = element_text(angle = 90, hjust = 1))
  tiff(paste(output_location,"\\Graphs\\","ABGH_",name,"_",ct,".tiff",sep=""),compression="lzw")
  grid.draw(p1)
  dev.off()
 }

# Testing
# taxa_level<-c("Genus")
# a<-c("MSA1000","MSA1001")
# subset_list<-unique(sample_data(physeq_scale_sub)$Source.Descrip) #List of all controls
# phyobj<-subset_samples(physeq_scale, Source.Descrip %in% subset_list)
# environments <- sample_data(physeq_scale_sub)$Run.ID
# remove(taxa_level,a, subset_list, phyobj, environments, temp_phyobj,taxaPValues,environments,abundances,chlamydiae.families,taxaPValues,family.AGBH)

subset_list<-unique(sample_data(physeq_scale_sub)$Source.Descrip) #List of all controls

#Compare taxa between runs
for (a in subset_list){
 exp_taxa_list<-rownames(get(paste("expected_taxa",taxa_levels,sep="_"))[ get(paste("expected_taxa",taxa_levels,sep="_"))[[a]] !='A' , ])
 physeq_temp<-subset_samples(physeq_scale_sub, Source.Descrip %in% a)
 
 i=1
 for (b in exp_taxa_list){
  newname<-paste(" g__",b,sep="")
  exp_taxa_list[i]<-newname
  i=i+1
 }
 
 physeq_temp<-subset_taxa(physeq_temp, Genus %in% exp_taxa_list)
 environments <- sample_data(physeq_temp)$Run.ID
 relabun_sig(physeq_temp,"Genus",environments,a)
}

```

#Error rates
```{r}
#Determine the Error (r.m.s. for each control)
error_relabun<-function(df,outputname){
 tmp <- data.frame()
 
 #For each expected control
 for(a in row.names(df)){
  if((grepl("_exp",a))){ #only pull the expected sample names
   
   #For each genera
   for (b in colnames(df)){
    exp_val<-df[a,b]
    obs_sample <- gsub("_exp","",a)
    obs_val<-df[obs_sample,b]
    
     #only include expected control values
     if(exp_val!=0){
      tmp[obs_sample,b]<-rmse(obs_val,exp_val)
     }
   }
  }
  
  
 }
 tmp[is.na(tmp)]=0 #replace na's with 0
 tmp<-cbind(tmp,rowMeans(tmp))
 
 write.csv(tmp,paste(output_location,"\\Data\\",outputname,"_seqcontrols.csv",sep=""))
}

error_relabun(relabun_seqcont_av,"errorrate")
```

#Alpha Diversity - Run ID
```{r}
plot_diversity<-function(phyobject,filter,subfeat){
 p<-plot_richness(phyobject, x=filter) + geom_boxplot()
 file_name =paste(output_location,"Graphs\\alphadiv_",filter,"_",subfeat,".tiff",sep="")
 tiff(file_name, width=800)
 print(p)
 dev.off()
} 

alpha_plots<-c("Source.Descrip","Run.ID")

for (a in alpha_plots){
 plot_diversity(physeq_scale_sub,a,"seq")
}

```


#Beta diversity
```{r}
#PCOA
plot_beta<-function(phyobj,type,dist_meas,colby){
 ordu = ordinate(phyobj, type, dist_meas, weighted=TRUE)
 p1<-plot_ordination(phyobj, ordu, color=colby,title=paste("Beta Div (", type,") by ", colby,sep=""))
 
 allGroupsColors<- c("grey0")
 allGroupsColors<-append(allGroupsColors,brewer.pal(n = 9, name = "Set1"))
 allGroupsColors<-append(allGroupsColors,brewer.pal(n = 12, name = "Paired"))

 p1<-p1 + geom_point(size = 4) + scale_color_manual(values = allGroupsColors)
 p1
 
 tiff(paste(output_location,"\\Graphs\\betadiv_",type,"_",colby,".tiff",sep=""))
 grid.draw(p1)
 dev.off()
 
}

plot_beta(physeq_scale_sub,"PCoA","unifrac","Run.ID")
plot_beta(physeq_scale_sub,"PCoA","unifrac","Source.Descrip")

#Jaccard
#plot_net(physeq_scale_sub, maxdist = 0.4,color="Source.Descrip",shape="Run.ID")
jg = make_network(physeq_scale_sub, max.dist=0.3)
p1=plot_network(jg, physeq_scale_sub, color="Source.Descrip",shape="Run.ID",label="Source.Descrip")
df
tiff(paste(output_location,"\\Graphs\jaccard_seq.tiff",sep=""))
grid.draw(p1)
dev.off()```

