---
title: "R Notebook"
output: word_document
editor_options: 
  chunk_output_type: console
---

#To Update
```{r}
project_name="Project_NP0084-MB4_5_pt6_2"
manifest_name="metadata_Summary_1.csv"
otu_name="otu_Summary_1.csv"
sample_depth= 20000
```

#Load required libraries
############### 
```{r}
library("biom")
library(data.table)
library(ggplot2)
library(tm)
library(vegan)
library("biom")
library(tidyr)
library(ggplot2)
library(cowplot)
library(scales)
library(phyloseq)
library(qiime2R)
library(tibble)
library(gridExtra)
library(tidyverse)
```

#Analysis
###############
#1. Taxonomy Reference - Expected frequencies
```{r}
#Read in controls taxonomy file, grouped by Genus
control_expect <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Controls\\Taxonomy\\Taxonomy_Genus_2019.txt",sep="\t",header=TRUE)
rownames(control_expect) <- unique(control_expect[,"Genus"])
```

#2. Load OTU and metadata tables
```{r}
#Create location link
project_location= paste("T:\\DCEG\\Projects\\Microbiome\\CGR_MB\\MicroBiome\\",project_name,sep="")

#Read in database with variable information
metadata_baseline <- read.table(paste(project_location,"\\Output\\R\\OTUs\\",manifest_name,sep=""),sep=",",header= TRUE)
rownames(metadata_baseline)<-metadata_baseline$X
metadata_baseline<-metadata_baseline[-1]

otu_baseline_complete <- read.table(paste(project_location,"\\Output\\R\\OTUs\\",otu_name,sep=""),sep=",",header= TRUE)
rownames(otu_baseline_complete)<-otu_baseline_complete$X
otu_baseline_complete<-otu_baseline_complete[-1]

rownames(metadata_baseline)==rownames(otu_baseline_complete) #Should be true
```

#3. OTU Counts by sample
Read in OTU information, and determine total OTUs
```{r}
genus_found<-colnames(otu_baseline_complete)
samplelist_seqcontrols<-rownames(otu_baseline_complete)

#For each sample, determine the total read counts
for (a in samplelist_seqcontrols){
 #Set sum to 0
 sum=0
 
 #For each genus
 for (b in genus_found){
  
  #Add the number of counts together
  sum <- otu_baseline_complete[a,b] + sum
 }
 
 #When complete, push the sum to a new "Total_OTUs" row
 otu_baseline_complete[a,"Total_OTUs"] <- sum
}
```

#4. Observed to Expected Counts
```{r}
samplelist_seqcontrols<-rownames(otu_baseline_complete)
observed_count=0
observed_baseline_seqcontrols<- data.frame()

#Individual Samples
for (a in samplelist_seqcontrols){
 control_type <- as.String(metadata_baseline[a,"Source.Descrip"])
 genus_temp <- control_expect[,c("Gram",control_type)]
 genus_temp <- row.names(subset(genus_temp,!(genus_temp[,control_type]=="A")))
 
 for (b in genus_temp){
  otu_count <- otu_baseline_complete[a,b]
  
  if(!(is.na(observed_count) || is.null(otu_count) || otu_count==0)){
    observed_count= observed_count+1
  }
    
 }
 
 observed_baseline_seqcontrols[a,"Observed"] <- observed_count
 observed_baseline_seqcontrols[a,"Expected"] <- length(genus_temp)
 observed_baseline_seqcontrols[a,"Source.Descrip"] <- metadata_baseline[a,"Source.Descrip"]
 observed_count=0
}

write.csv(observed_baseline_seqcontrols,paste(project_location,"\\Output\\R\\Data\\Baseline\\observed_baseline_seqcontrols.csv",sep=""))

#Average by Source.Descrip
observed_count=0
count=0
source_list<-unique(observed_baseline_seqcontrols$Source.Descrip)

for (a in source_list){
 for (b in samplelist_seqcontrols){
  control_type <- as.String(metadata_baseline[b,"Source.Descrip"])
  if(a==control_type){
   observed_count<-observed_baseline_seqcontrols[b,"Observed"]+observed_count
   expected_count<-observed_baseline_seqcontrols[b,"Expected"]
   count=count+1
  }
 }
 observed_baseline_seqcontrols[a,"Observed"]<-observed_count/count
 observed_baseline_seqcontrols[a,"Expected"]<-expected_count
 observed_baseline_seqcontrols[a,"Source.Descrip"]<-a
 observed_count=0
 count=0
}

write.csv(observed_baseline_seqcontrols,paste(project_location,"\\Output\\R\\Data\\Baseline\\observed_baseline_seqcontrols.csv",sep=""))
```

#5. Relative Abundance  - Seq Controls
```{r}
relabun_baseline_seqcontrols<-data.frame()

#Complete Relative Abundance
samplelist_seqcontrols<-rownames(otu_baseline_complete)
genus_list <- colnames(otu_baseline_complete)
genus_list<-genus_list[-(length(genus_list))]
genus_list<-append(genus_list,rownames(control_expect))

for (a in samplelist_seqcontrols){
 for (b in genus_list){
  
  relabun <- otu_baseline_complete[a,b]
  
  if(!(is.na(relabun) || is.null(relabun) || relabun==0)){
   relabun_baseline_seqcontrols[a,b] <- otu_baseline_complete[a,b]/otu_baseline_complete[a,"Total_OTUs"]
   relabun_baseline_seqcontrols[a,"Source.Descrip"]<-as.character(metadata_baseline[a,"Source.Descrip"])
  }
  control_type<-as.character(metadata_baseline[a,"Source.Descrip"])
  relabun_baseline_seqcontrols[control_type,b]<-as.character(control_expect[b,control_type])
  relabun_baseline_seqcontrols[control_type,"Source.Descrip"]<-control_type
 }
}

relabun_baseline_seqcontrols[is.na(relabun_baseline_seqcontrols)]=0 #replace na's with 0
relabun_baseline_seqcontrols[relabun_baseline_seqcontrols=="A"]=0 #replace A's with 0
write.csv(relabun_baseline_seqcontrols,paste(project_location,"\\Output\\R\\Data\\Baseline\\relabun_baseline_seqcontrols.csv",sep=""))
```

#6 Plot graph - Actual to Expected - IntraControl Variability
For each kit, show variability, and map
```{r}
#Create list for plots, start counter
plot_list <- list()
i=1

for (a in control_types){
   #Create genus list of specfic control
   genus_temp <- control_expect[!is.na(control_expect[,a]),]
   genus_temp <- (genus_temp[,"Genus"])
   
   #Create a list of samples that are of the specified control type
   sample_list<- subset(metadata_baseline_filtered, Type==a)
   sample_list <- unique(sample_list$SampleID)

   #Create dataframe
   temp<-data.frame()
   count=1
   
   #For each Genus in list, determine the samples value
   for (b in genus_temp){
    
     #For each of the samples in the sample list
     for(c in sample_list){
      temp[count,"Genus"]<-b
      temp[count,"Obs_Per"]<- samples_obs_abund[b,c]
      temp[count,"Sample"]<-c
      temp[count,"Exp_Per"]<-control_expect[b,a]
      count=count+1
     }
   }
  
    graph_title <- paste (a,": Intracontrol Variablity",sep= "")

    #Plot Data
    graph_new <- ggplot(data=temp,mapping=aes(x=Obs_Per,y=Exp_Per,color=Sample))+
     geom_point()+ #Can add points to graph
     geom_smooth(method= "lm",se=FALSE)+ #Add a smooth line
     labs(title=graph_title, x="Observed Rel. Abundance", y= "Expected Rel. Abundance")+
     geom_text(mapping=aes(x=Obs_Per,y=Exp_Per,label=Genus),check_overlap= TRUE)
     
    
    plot_list[[i]]= graph_new
    i=i+1
}

#Save all plots
for (i in 1:length(control_types)){
 
 #ID which control is being run
 a <- control_types[i]
 
 #Save the file
 file_name= paste("Data/Graphs/Intracontrol Variability_",a, ".tiff", sep="")
 tiff(file_name)
 print(plot_list[[i]])
 dev.off()
}
```

#12 Plot graph - Actual to Expected - One graph
For each kit, show variability, and map
```{r}
#Create list for plots, start counter
temp<-data.frame()
count<-1
for (a in control_types){
   #Create genus list of specfic control
   genus_temp <- control_expect[!is.na(control_expect[,a]),]
   genus_temp <- (genus_temp[,"Genus"])
   
   #Create a list of samples that are of the specified control type
   sample_list<- subset(metadata_baseline_filtered, Type==a)
   sample_list <- unique(sample_list$SampleID)

   #For each Genus in list, determine the samples value
   for (b in genus_temp){
    
     #For each of the samples in the sample list
     for(c in sample_list){
      temp[count,"Genus"]<-b
      temp[count,"Obs_Per"]<- samples_obs_abund[b,c]
      temp[count,"Sample"]<-c
      temp[count,"Exp_Per"]<-control_expect[b,a]
      temp[count,"Type"] <- a
      count=count+1
     }
   }
}

#Plot graph
graph_new<- ggplot(data=temp,mapping=aes(x=Obs_Per,y=Exp_Per,color=Type))+
 #geom_point()+ #Can add points to graph
 geom_smooth(method= "lm",se=FALSE)+ #Add a smooth line
 labs(title="Sequencing Control Comparison", x="Observed Rel. Abundance", y= "Expected Rel. Abundance")
graph_new
write.csv(temp,"temp2.csv")

file_name=("Data/Graphs/Kit Differences_ZE.tiff")
tiff(file_name)
print(graph_new)
dev.off()


```