---
title: "R Notebook"
output: word_document
editor_options: 
  chunk_output_type: console
---

#To Update
```{r}
output_location<-c("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Analysis\\NP0084-MB4,5,6\\R_Complete\\")

taxa_levels<-c("Genus")
reference_db<-"greengenes"

file_ext<-"_extcontrol_hmo.csv"

ext_choice<-"HOMO"

control_location<-c("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Controls\\Taxonomy\\")
```

#Load required libraries
############### 
```{r}
library("ape")
library("biom")
library(cowplot)
library(data.table)
library(ggplot2)
library(grid)
library(gridExtra)
library("igraph")
library(qiime2R)
library(phyloseq)
library(scales)
library(structSSI)
library(tibble)
library(tidyverse)
library(tidyr)
library(tm)
library(vegan)
```


#Analysis
###############
#Taxonomy Reference - Expected frequencies
```{r}
#Read in controls taxonomy file, grouped by taxa
for (a in taxa_levels){
 control_expect <- read.table(paste(control_location,"Taxonomy_",taxa_levels,"_2019.txt",sep=""),sep="\t",header=TRUE)
 rownames(control_expect) <- unique(control_expect[,taxa_levels])
 
 control_db_name<-paste("expected_taxa_",a,sep="")
 assign(control_db_name,control_expect)
 remove(control_expect)
}

```

#Load OTU and metadata tables
```{r}
for(a in taxa_levels){
 for (b in reference_db){
  metadata_db_name<-paste("metadata_",a,"_",b,sep="")
  
  temp <- read.table(paste(output_location,"Taxa\\",metadata_db_name,file_ext,sep=""),sep=",",header= TRUE)
  rownames(temp)<-temp$X
  temp<-temp[-1]
  
  assign(metadata_db_name,temp)
  remove(temp)
  
  taxa_db_name<-paste("taxa_",a,"_",b,sep="")
  
  temp<-read.table(paste(output_location,"Taxa\\",taxa_db_name,file_ext,sep=""),sep=",",header= TRUE)
  rownames(temp)<-temp$X
  temp<-temp[-1]
  
  assign(taxa_db_name,temp)
  remove(temp)
 }
}
```

#Load phylo object
```{r}
#NOTE: must run 1. Initial processing
#physeq_scale include all samples

#Subset to create standard ext only
subset_list<-c("Ext.Control") 
physeq_scale_sub<-subset_samples(physeq_scale, Sample.Type %in% subset_list)
 
if(ext_choice=="STD"){

 subset_list<-unique((sample_data(physeq_scale_sub)$Reciept))
 subset_list<-subset_list[1:22]
 physeq_scale_sub<-subset_samples(physeq_scale_sub, Reciept %in% subset_list)
} else{
 #Subset to create homo ext only
 subset_list<-unique((sample_data(physeq_scale_sub)$Reciept))
 subset_list<-subset_list[23:34]
 physeq_scale_sub<-subset_samples(physeq_scale_sub, Reciept %in% subset_list)
}
```

#OTU Counts by sample
Determine total OTUs
```{r}
taxa_Genus_greengenes$Total_OTUs <- rowSums(taxa_Genus_greengenes)

```

#Observed to Expected Counts
```{r}
#Testing
# output<-tar_tdr
# db<-"greengenes"
# tl<-"Genus"
# ct<-"DZ35316"
# otherfilt<-"MagMax Microbiome Ultra Kit"
# remove(output,db,tl,ct,otherfilt)
# remove(sample_list,exp_taxa_list,unexp_taxa_list)

tp_fp_fn_samplelevel<- function(output,db,tl,ct,otherfilt){
 sample_list<-intersect(rownames(get(paste("taxa",tl,db,sep="_"))),rownames(subset(get(paste("metadata",tl,db,sep="_")),Source.Descrip==ct)))
 sample_list<-intersect(sample_list,rownames(subset(get(paste("metadata",tl,db,sep="_")),Ext.Kit==otherfilt)))#add another layer of filtering based on extraction kit
 
 exp_taxa_list<-rownames(get(paste("expected_taxa",tl,sep="_"))[ get(paste("expected_taxa",tl,sep="_"))[[ct]] !='A' , ])
 unexp_taxa_list<-setdiff(colnames(get(paste("taxa",tl,db,sep="_"))),exp_taxa_list)
 unexp_taxa_list<-unexp_taxa_list[-length(unexp_taxa_list)] #Remove Total_OTUs column
 unexp_taxa_list<-unexp_taxa_list[-45]#Remove Higher Genus
 
 for (a in sample_list){
  fp=0
  fn=0
  tp=0
  
  taxa_db<-get(paste("taxa",tl,db,sep="_"))
  
  #Run through expected taxa list
  for (b in exp_taxa_list){
   otu_count<-taxa_db[a,b]

   
   if(is.null(otu_count) || otu_count==0 || is.na(otu_count)){
    fn=fn+1

   } else if (otu_count>0){
    tp=tp+1
   }
  }
  
  #Run through all taxa not expected
  for (b in unexp_taxa_list){
   otu_count<-taxa_db[a,b]
   
   if(otu_count==0){
    next;
   } else if (otu_count>0){
    fp=fp+1
   }
  }
  
  output[a,"ControlType"]<-ct
  output[a,"Kit"]<-get(paste("metadata",tl,db,sep="_"))[a,"Ext.Kit"]
  output[a,"Expected"]<-length(exp_taxa_list)
  output[a,paste("TP",db,tl,sep="_")]<-tp
  output[a,paste("FP",db,tl,sep="_")]<-fp
  output[a,paste("FN",db,tl,sep="_")]<-fn
  output[a,paste("TAR",db,tl,sep="_")]<-tp/(tp+fp)
  output[a,paste("TDR",db,tl,sep="_")]<-tp/(tp+fn)
  output[a,paste("F1",db,tl,sep="_")]<-2*((tp/(tp+fp))*(tp/(tp+fn)))/((tp/(tp+fp))+(tp/(tp+fn)))
  output[a,"Homo.Method"]<-get(paste("metadata",tl,db,sep="_"))[a,"Homo.Method"]
 }
 
 return(output)
}

#Determine which controls
control_type<-unique(sample_data(physeq_scale_sub)$Source.Descrip)
control_type<-control_type[-c(3,4)] #Remove Water and Robogut
control_type

other_filt<-unique(sample_data(physeq_scale_sub)$Ext.Kit)


tar_tdr<-data.frame()
for(a in reference_db){
 for (b in taxa_levels){
  for (c in control_type){
   for (d in other_filt){
    tar_tdr<-tp_fp_fn_samplelevel(tar_tdr,a,b,c,d) #D: change subsetting of sample_list as needed within funciton IE Source.Descrip or Ext.Kit
   }
  }
 }
}

write.csv(tar_tdr,paste(output_location,"\\Taxa\\Taxonomy_tar_trd_ext_",ext_choice,".csv",sep=""))

```

#Average by Control
```{r}
#Testing
db<-subset(tar_tdr,Kit=="DSP Virus")
opts<-"DSP Virus"
db<-tar_tdr
opts<-"HOMO"
remove(db,opts,temp,count)

#update cols for kit inclusion
tar_tdr_av_kit<-function(db,opts,opts2){
 
 temp<-aggregate(db[, 3], list(db$ControlType), mean) #Expected tax
 row.names(temp)<-temp$Group.1
 temp<-temp[,-1]
 
 temp<-cbind(temp,aggregate(db[, 7], list(db$ControlType), mean)) #TAR Average
 row.names(temp)<-temp$Group.1
 temp<-temp[,-2]
 
 temp<-cbind(temp,aggregate(db[, 8], list(db$ControlType), mean)) #TDR Average
 row.names(temp)<-temp$Group.1
 temp<-temp[,-3]
 
 temp<-cbind(temp,aggregate(db[, 9], list(db$ControlType), mean)) #F1 Average
 temp<-temp[,-4]
 
 colnames(temp)<-c("Expected","TAR","TDR","F1Score")
 
 if(opts=="STD"){
  
 } else{
  count=1
  
  for(a in rownames(temp)){
   temp[a,"Kit"]<-opts
   temp[a,"Control"]<-a
   temp[a,"Homo.Method"]<-opts2
   rownames(temp)[count]<-paste(a,opts,opts2,sep="_")
   count=count+1
  }
 }

 return(temp)
}

tar_tdr_av<-data.frame()
if(ext_choice=="STD"){
 #Run for averages
 tar_tdr_av<-tar_tdr_av_kit(tar_tdr,ext_choice) 
 write.csv(tar_tdr_av,paste(output_location,"\\Taxa\\Taxonomy_tar_trd_ext_av.csv",sep=""))
 
 #Run by Kit
 tar_tdr_av<-data.frame("Expected"=1,"TAR"=1,"TDR"=1,"F1Score"=1,"Kit"=1,"Control"=1,"Homo.Method"=1)
 filt_list<-unique(tar_tdr$Kit)
 
 for(a in filt_list){
  temp<-subset(tar_tdr,Kit==a)
 
  tar_tdr_av<-rbind(tar_tdr_av,tar_tdr_av_kit(temp,a,""))
  remove(temp)
 }
 
 tar_tdr_av<-tar_tdr_av[-1,]#Remove starter row
 write.csv(tar_tdr_av,paste(output_location,"\\Taxa\\Taxonomy_tar_trd_ext_av_kit.csv",sep=""))
 
 } else{

 #Run by Homog
 tar_tdr_av<-data.frame("Expected"=1,"TAR"=1,"TDR"=1,"F1Score"=1,"Kit"=1,"Control"=1, "Homo.Method"=1)
 filt_list<-unique(tar_tdr$Kit)
 filt_list1<-unique(tar_tdr$Homo.Method)

 for (a in filt_list1){
  temp<-subset(tar_tdr,Homo.Method==a)
 
  tar_tdr_av<-rbind(tar_tdr_av,tar_tdr_av_kit(temp,"",a)) #db, kit, homo
  #remove(temp)
 }
 
 tar_tdr_av<-tar_tdr_av[-1,]#Remove starter row
 write.csv(tar_tdr_av,paste(output_location,"\\Taxa\\Taxonomy_tar_trd_ext_av_homo.csv",sep=""))
 
 #Run by homog, by kit
 tar_tdr_av<-data.frame("Expected"=1,"TAR"=1,"TDR"=1,"F1Score"=1,"Kit"=1,"Control"=1, "Homo.Method"=1)
 for(a in filt_list){
  temp<-subset(tar_tdr,Kit==a)
  for (b in filt_list1){
   temp1<-subset(temp,Homo.Method==b)
   print(a)
   print(b)
   if(nrow(temp)>0){
    print ("OK")
    tar_tdr_av<-rbind(tar_tdr_av,tar_tdr_av_kit(temp,a,b))
   }
  }
 }
 
 tar_tdr_av<-tar_tdr_av[-1,]#Remove starter row
 write.csv(tar_tdr_av,paste(output_location,"\\Taxa\\Taxonomy_tar_trd_ext_av_homo_kit.csv",sep=""))
 #remove(temp)
}


```

#Relative Abundance  - Ext Controls
```{r}
relabun_extcont<-data.frame()

#Complete Relative Abundance
sample_list_extcont<-rownames(taxa_Genus_greengenes)
taxa_list <- colnames(taxa_Genus_greengenes)
taxa_list<-taxa_list[-(length(taxa_list))] #Remove total OTU's from list
taxa_list<-unique(append(taxa_list,rownames(expected_taxa_Genus)))

for (a in sample_list_extcont){
 relabun_extcont[a,"Source.Descrip"]<-as.character(metadata_Genus_greengenes[a,"Source.Descrip"])
 relabun_extcont[a,"Ext.Kit"]<-as.character(metadata_Genus_greengenes[a,"Ext.Kit"])
 relabun_extcont[a,"Homo.Method"]<-as.character(metadata_Genus_greengenes[a,"Homo.Method"])

 
 for (b in taxa_list){
  relabun <- taxa_Genus_greengenes[a,b]
  
  if(!(is.na(relabun) || is.null(relabun) || relabun==0)){
   relabun_extcont[a,b] <- taxa_Genus_greengenes[a,b]/taxa_Genus_greengenes[a,"Total_OTUs"]
  } else{
   relabun_extcont[a,b] <- 0
  }
 }
}

write.csv(relabun_extcont,paste(output_location,"Data\\relabun_extcont_",ext_choice,".csv",sep=""))

#Calculate averages - Across all kits/homog
relabun_extcont_av<-sort(unique(relabun_extcont$Source.Descrip))

for (a in taxa_list){
 
 temp<-aggregate(relabun_extcont[,a], list(relabun_extcont$Source.Descrip), FUN = function(x) mean(as.numeric(as.character(x))))
 relabun_extcont_av<-cbind(relabun_extcont_av,temp[,2])
}

#Convert to df, add taxa as colnames
relabun_extcont_av<-as.data.frame(relabun_extcont_av)
row.names(relabun_extcont_av)<-relabun_extcont_av[,1] #col one is Source.Descrip
relabun_extcont_av<-relabun_extcont_av[,-1]
colnames(relabun_extcont_av)<-taxa_list
relabun_extcont_av[] <- lapply(relabun_extcont_av, function(x) {
    as.numeric(as.character(x))})

#add expected rel abundance
temp<-data.frame()
for (a in control_type){
 for (b in taxa_list){
  temp[paste(a,"_exp",sep=""),b]<-as.numeric(as.character(expected_taxa_Genus[b,a]))
 }
}

temp[temp=="A" ]=0 #replace A's with 0
temp[is.na(temp)]=0 #replace A's with 0
temp<-as.data.frame(temp)
relabun_extcont_av<-rbind(relabun_extcont_av,temp)

write.csv(relabun_extcont_av,paste(output_location,"Data\\relabun_extcont_av.csv",sep=""))



relabun_subset<-function(){
 
 #Calculate averages - Across all kits/homog
 relabun_extcont_av<-sort(unique(relabun_extcont$Source.Descrip))
 
 for (a in taxa_list){
  
  temp<-aggregate(relabun_extcont[,a], list(relabun_extcont$Source.Descrip), FUN = function(x) mean(as.numeric(as.character(x))))
  relabun_extcont_av<-cbind(relabun_extcont_av,temp[,2])
 }
 
 #Convert to df, add taxa as colnames
 relabun_extcont_av<-as.data.frame(relabun_extcont_av)
 row.names(relabun_extcont_av)<-relabun_extcont_av[,1] #col one is Source.Descrip
 relabun_extcont_av<-relabun_extcont_av[,-1]
 colnames(relabun_extcont_av)<-taxa_list
 relabun_extcont_av[] <- lapply(relabun_extcont_av, function(x) {
     as.numeric(as.character(x))})
 
 #add expected rel abundance
 temp<-data.frame()
 for (a in control_type){
  for (b in taxa_list){
   temp[paste(a,"_exp",sep=""),b]<-as.numeric(as.character(expected_taxa_Genus[b,a]))
  }
 }
 
 temp[temp=="A" ]=0 #replace A's with 0
 temp[is.na(temp)]=0 #replace A's with 0
 temp<-as.data.frame(temp)
 relabun_extcont_av<-rbind(relabun_extcont_av,temp)
 
 write.csv(relabun_extcont_av,paste(output_location,"Data\\relabun_extcont_av.csv",sep=""))
}

```




#5. Relative Abundance  - Ext Controls
```{r}
relabun_baseline_extcontrols<-data.frame()

#Complete Relative Abundance
samplelist_extcontrols<-rownames(otu_baseline_complete)
genus_list <- colnames(otu_baseline_complete)
genus_list<-genus_list[-(length(genus_list))]
genus_list<-append(genus_list,rownames(control_expect))
kit_list <- unique(observed_baseline_extcontrols$Ext.Kit)

for (a in kit_list){
 for (b in samplelist_extcontrols){
  for (c in genus_list){
   
   relabun <- otu_baseline_complete[b,c]
   
   if(!(is.na(relabun) || is.null(relabun) || relabun==0)){
    relabun_baseline_extcontrols[b,"Source.Descrip"]<-as.character(metadata_baseline[b,"Source.Descrip"])
    relabun_baseline_extcontrols[b,"Ext.Kit"]<-as.character(metadata_baseline[b,"Ext.Kit"])
    relabun_baseline_extcontrols[b,c] <- otu_baseline_complete[b,c]/otu_baseline_complete[b,"Total_OTUs"]
    relabun_baseline_extcontrols[b,"Homo.Method"] <- metadata_baseline[b,"Homo.Method"]
    
   }
   control_type<-as.character(metadata_baseline[b,"Source.Descrip"])
   if(control_type!="Water"){
    sample_name<-paste(control_type,"_Expected",sep="") #Add control information to compare expected frequencies
    kit_name <- a
    relabun_baseline_extcontrols[sample_name,c]<-as.character(control_expect[c,control_type])
    relabun_baseline_extcontrols[sample_name,"Source.Descrip"]<-control_type
    relabun_baseline_extcontrols[sample_name,"Ext.Kit"]<-kit_name
   }
  }
 }
}

relabun_baseline_extcontrols[is.na(relabun_baseline_extcontrols)]=0 #replace na's with 0
relabun_baseline_extcontrols[relabun_baseline_extcontrols=="A"]=0 #replace A's with 0
write.csv(relabun_baseline_extcontrols,paste(project_location,"\\Output\\R\\Data\\Baseline\\relabun_baseline_extcontrols2.csv",sep=""))

#Create rel abun df to graph observed to expected frequencies
samplelist_extcontrols_nowater <- subset(metadata_baseline,Source.Descrip!="Water") #only include controls
samplelist_extcontrols_nowater <- rownames(samplelist_extcontrols_nowater)
relabun_baseline_extcontrols_otoe<-data.frame()
count=1

for (a in samplelist_extcontrols_nowater){
 for (b in genus_list){
  control_type<-as.character(metadata_baseline[a,"Source.Descrip"])
  
  relabun_ob <- relabun_baseline_extcontrols[a,b]
  relabun_ex <- as.character(control_expect[b,control_type])
  sample_name  <- a
  genus_temp <- b

  if(!is.null(relabun_ex) && relabun_ex!="A" && !is.na(relabun_ex)){ #Only need genus that are expected for comparisons
   relabun_baseline_extcontrols_otoe[count,"SampleID"]<-sample_name
   relabun_baseline_extcontrols_otoe[count,"RelAbun_Obs"]<-relabun_ob
   relabun_baseline_extcontrols_otoe[count,"RelAbun_Exp"]<-relabun_ex
   relabun_baseline_extcontrols_otoe[count,"Genus"]<-genus_temp
   relabun_baseline_extcontrols_otoe[count,"Ext.Kit"]<-relabun_baseline_extcontrols[a,"Ext.Kit"]
   relabun_baseline_extcontrols_otoe[count,"Source.Descrip"]<-relabun_baseline_extcontrols[a,"Source.Descrip"]
   count=count+1
  }
  
 }
}

write.csv(relabun_baseline_extcontrols_otoe,paste(project_location,"\\Output\\R\\Data\\Baseline\\relabun_baseline_extcontrols_otoe.csv",sep=""))

```

#6 Graph Rel Abun E to O by Control
```{r}
plot_list <- list()
control_types <- unique(relabun_baseline_extcontrols_otoe$Source.Descrip)
i=1

for (a in control_types){
 temp<-subset(relabun_baseline_extcontrols_otoe,Source.Descrip==a) #subset to include only one control at a time
 temp$RelAbun_Obs<-as.numeric(as.character(temp$RelAbun_Obs))
 temp$RelAbun_Exp<-as.numeric(as.character(temp$RelAbun_Exp))

 graph_title <- paste ("Rel Abun Ob to Exp Variablity for ",a,sep = "")
 
 #Plot Data
 graph_new <- ggplot(data=temp,mapping=aes(x=RelAbun_Obs,y=RelAbun_Exp,color=Ext.Kit)) + 
  geom_point()+ #Can add points to graph
  geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
  labs(title=graph_title, x="Observed Rel Abun", y= "Expected Rel. Abun")+
 # geom_text(mapping=aes(x=RelAbun_Obs,y=RelAbun_Exp,label=Genus),check_overlap = TRUE)+
  xlim(0,.25) + ylim(0.04,.2)
 
 plot_list[[i]] = graph_new
 i=i+1
}


#Save all plots
count=1

for (i in 1:(length(control_types))){
 
 #ID which control is being run
 a <- control_types[count]
 
 #Save the file
 file_name = paste(project_location,"\\Output\\R\\Graphs\\Baseline\\Variability_",a,".tiff", sep="")
 tiff(file_name)
 print(plot_list[[i]])
 dev.off()
 
 count = count+1
} 


```

#7 Relative Abundance % Difference
```{r}
#convert rows to numeric values
 relabun_baseline_extcontrols_otoe$RelAbun_Obs<-as.numeric(as.character(relabun_baseline_extcontrols_otoe$RelAbun_Obs))
 relabun_baseline_extcontrols_otoe$RelAbun_Exp<-as.numeric(as.character(relabun_baseline_extcontrols_otoe$RelAbun_Exp))

for (a in rownames(relabun_baseline_extcontrols_otoe)){
 relabun_baseline_extcontrols_otoe[a,"PDiff"]<-((relabun_baseline_extcontrols_otoe[a,"RelAbun_Obs"]-relabun_baseline_extcontrols_otoe[a,"RelAbun_Exp"])/relabun_baseline_extcontrols_otoe[a,"RelAbun_Exp"])*100
}

write.csv(relabun_baseline_extcontrols_otoe,paste(project_location,"\\Output\\R\\Data\\Baseline\\relabun_baseline_extcontrols_otoe.csv",sep=""))

```

#8 Alpha Diversity of Homo Methods
```{r}
otu<-otu_table(physeq_filt_extcontrols)
tax<-tax_table(physeq_filt_extcontrols)
random_tree<-random_tree<-rtree(ntaxa(physeq_filt_extcontrols), rooted=TRUE, tip.label=taxa_names(physeq_filt_extcontrols))
sdata<- sample_data(metadata_baseline)
temp<-phyloseq(otu,tax,sdata,random_tree)

# Scale reads to even depth 
physeq_scale_temp<-temp %>%
  scale_reads(n=1000) 

remove(temp)

#Homo Method
p<-plot_richness(physeq_scale_temp, x="Homo.Method") + geom_boxplot()
file_name =paste(project_location,"\\Output\\R\\Graphs\\Summary\\alphadiv_homomethod.tiff",sep="")
tiff(file_name, width=800)
print(p)
dev.off()

remove(physeq_scale_temp)
```
################
#Unused, but useful?

#Reproducibility - Original vs Replicate
```{r}
#https://cran.r-project.org/web/packages/qualityTools/vignettes/qualityTools.pdf

#Create genus list of Zymo Ext Control only
genus_temp <- control_expect[!is.na(control_expect[,"Zymo.Ext"]),]
genus_temp <- (genus_temp[,"Genus"])

#Create dataframe
temp<-data.frame()
count=1

#Match originals back to residuals 
#SC249391.PC04924.D03 & SC249391.PC04925.H01 to SC261522.PC04925.F11 
#SC304927.PC07578.B02 to SC304937.PC07578.D04
# SC304924.PC07578.B.07.PC07578.B07 to SC304934.PC07578.D.02.PC07578.D02 & SC304934.PC07578.C.05.PC07578.C05
sample_list <- c("SC249391.PC04925.H01", "SC261522.PC04925.F11", "SC304927.PC07578.B02", "SC304937.PC07578.D04", "SC304924.PC07578.B.07.PC07578.B07", "SC304934.PC07578.D.02.PC07578.D02" )

#For all genus
for (a in genus_temp){
  
 #Move through list
 for (b in sample_list){
   temp[count,"Genus"]<-a
   temp[count,"Obs_Per"]<- samples_obs_abund[a,b]
   temp[count,"Sample"]<-b
   count=count+1
 }
}

#Remove rows with NA
temp<- temp %>% drop_na()

###Testing
write.csv(temp,"repo.csv")    
temp <- read.csv("repo.csv",header = TRUE)

#Design Repro features
repro_resvsori <- gageRRDesign(Operators = 2, 
             Parts = length(unique(temp$Genus)), 
             Measurements = length(unique(temp$Sample))/2,
             randomize = FALSE)

#Set the response - must be numeric value
response(repro_resvsori)<- as.numeric(as.list(temp[,"Obs_Per"]))

#Run summary
repro_complete <- gageRR(repro_resvsori, method = "crossed", sigma =6, alpha=.25)
summary(repro_complete)

#Plot results
#Operator A = Original, B = Residual
#Parts = Genus (7 total, 8th was removed due to NA)
#Measurements - Samples (4 total per operator)
plot(repro_complete) 
```

#Reproducibilty - Extraction Kits
```{r}
#Create dataframe
temp<-data.frame()
count=1
for (a in control_types){
 #Create genus list of specfic control
 genus_temp <- control_expect[!is.na(control_expect[,a]),]
 genus_temp <- (genus_temp[,"Genus"])
 
 for (b in kit_types){
  
  ###Create database
  temp2 <- subset(samples_variables_f,Kit==b & Type==a & O.R == "Original")
  sample_list <- unique(temp2$NephID)

  #For each of the samples in the sample list
  for(c in sample_list){
   
   #For each Genus in list, determine the samples value
   for (d in genus_temp){
    temp[count,"Genus"]<-d
    temp[count,"Obs_Per"]<- samples_obs_abund[d,c]
    temp[count,"Sample"]<-c
    temp[count,"Kit"]<-b
    temp[count,"Type"]<-a
    count=count+1
   }
  }
 }
}

###Testing
write.csv(temp,"repo2.csv")    

#Load in final file (manipulated to choose six samples)
temp<- read.csv("Data/repo_ze_final1.csv",header = TRUE)

#Design Repro features
repro_resvsori <- gageRRDesign(Operators = 5, 
             Parts = length(unique(temp$Genus)), 
             Measurements = length(unique(temp$Sample))/5,
             randomize = FALSE)

#Set the response - must be numeric value
response(repro_resvsori)<- as.numeric(as.list(temp[,"Obs_Per"]))

#Run summary
repro_complete <- gageRR(repro_resvsori, method = "crossed", sigma =6, alpha=.25)
summary(repro_complete)

#Plot results
#Operator A = Qiagen DSP Virus, B = Zymo MagBead DNone Kit, C=MagAttract PowerMag Soil, D = MagAttract PowerMag Microbiome, E = Qiagen QIAamp  Modified
#Parts = Genus (8 total)
#Measurements - Samples (6 total per operator)
plot(repro_complete) 

```