---
title: "R Notebook"
output: word_document
editor_options: 
  chunk_output_type: console
---
#To Update
```{r}
project_name="Project_NP0084-MB4_5_pt6_2"
manifest_name="metadata_Summary_3.csv"
otu_name="otu_Summary_3.csv"
sample_depth= 20000
```

#Load required libraries
############### 
```{r}
library("biom")
library(data.table)
library(tm)
library(vegan)
library("biom")
library(tidyr)
library(ggplot2)
library(cowplot)
library(scales)
library(phyloseq)
library(qiime2R)
library(tibble)
library(gridExtra)
library(tidyverse)
```

#Analysis
###############
#1. Taxonomy Reference - Expected frequencies
```{r}
#Read in controls taxonomy file, grouped by Genus
control_expect <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Controls\\Taxonomy\\Taxonomy_Genus_2019.txt",sep="\t",header=TRUE)
rownames(control_expect) <- unique(control_expect[,"Genus"])
```

#2. Load OTU and metadata tables
```{r}
#Create location link
project_location= paste("T:\\DCEG\\Projects\\Microbiome\\CGR_MB\\MicroBiome\\",project_name,sep="")

#Read in database with variable information
metadata_baseline <- read.table(paste(project_location,"\\Output\\R\\OTUs\\",manifest_name,sep=""),sep=",",header= TRUE)
rownames(metadata_baseline)<-metadata_baseline$X
metadata_baseline<-metadata_baseline[-1]

otu_baseline_complete <- read.table(paste(project_location,"\\Output\\R\\OTUs\\",otu_name,sep=""),sep=",",header= TRUE)
rownames(otu_baseline_complete)<-otu_baseline_complete$X
otu_baseline_complete<-otu_baseline_complete[-1]

rownames(metadata_baseline)==rownames(otu_baseline_complete) #Should be true
```

#3. OTU Counts by sample
Read in OTU information, and determine total OTUs
```{r}
genus_found<-colnames(otu_baseline_complete)
samplelist_study<-rownames(otu_baseline_complete)

#For each sample, determine the total read counts
for (a in samplelist_study){
 #Set sum to 0
 sum=0
 
 #For each genus
 for (b in genus_found){
  
  #Add the number of counts together
  sum <- otu_baseline_complete[a,b] + sum
 }
 
 #When complete, push the sum to a new "Total_OTUs" row
 otu_baseline_complete[a,"Total_OTUs"] <- sum
}

write.csv(otu_baseline_complete,"test.csv")
```

#4. Relative Abundance  - Ext Controls
```{r}
relabun_baseline_study<-data.frame()

#Complete Relative Abundance
samplelist_study<-rownames(otu_baseline_complete)
genus_list <- colnames(otu_baseline_complete)
genus_list<-genus_list[-(length(genus_list))]
genus_list<-append(genus_list,rownames(control_expect))
kit_list <- unique(metadata_baseline$Ext.Kit)

for (a in kit_list){
 for (b in samplelist_study){
  for (c in genus_list){
   
   relabun <- otu_baseline_complete[b,c]
   
   if(!(is.na(relabun) || is.null(relabun) || relabun==0)){
    relabun_baseline_study[b,"Source.Descrip"]<-as.character(metadata_baseline[b,"Source.Descrip"])
    relabun_baseline_study[b,"Ext.Kit"]<-as.character(metadata_baseline[b,"Ext.Kit"])
    relabun_baseline_study[b,c] <- otu_baseline_complete[b,c]/otu_baseline_complete[b,"Total_OTUs"]
    relabun_baseline_study[b,"Homo.Method"] <- metadata_baseline[b,"Homo.Method"]
   }
  }
 }
}

relabun_baseline_study[is.na(relabun_baseline_study)]=0 #replace na's with 0
relabun_baseline_study[relabun_baseline_study=="A"]=0 #replace A's with 0
write.csv(relabun_baseline_study,paste(project_location,"\\Output\\R\\Data\\Baseline\\relabun_baseline_study.csv",sep=""))
```


################

#1. Load required libraries
```{r}
library("biom")
library(data.table)
library(ggplot2)
library(scales)
```

#2. Filtering of samples, control variable information
```{r}
#Read in database with variable information
samples_variables_complete <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Analysis\\ExpA_Run1&2\\Manifests\\sampledata_abrf.txt",sep="\t",header=TRUE)

#Read in database of failed samples
samples_failed <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Analysis\\ExpA_Run1&2\\TaxaCounts\\samplesfailed_q2.txt",sep="\t")

#Remove failed samples from info database
samples_filtered <-subset(samples_variables_complete, !(NephID %in% samples_failed$V1))

#Verify filtering - numbers should match
nrow(samples_variables_complete) - nrow(samples_failed)
nrow(samples_filtered)

#Only include Ext Controls
control_types <- c("RG","Study")
samples_variables_f <- samples_filtered[samples_filtered$Type %in% control_types,]

#Add sample ID's as row names
row.names(samples_variables_f) <- samples_variables_f$NephID
```

#3. OTU Counts by sample
Read in OTU information, and determine total counts for all samples
```{r}
#Read in OTU database at Genus level
samples_otu_complete <- read.csv("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fecal\\Fresh Fecal Optimization_2017.08\\Phase I\\Analysis\\ExpA_Run1&2\\TaxaCounts\\qiime2_taxa_counts_genus_silva_abrf.csv", row.names = 1)

#Create a list of all samples and all genus present in at least one sample
sample_list_complete <- colnames(samples_otu_complete)
genus_found <- rownames(samples_otu_complete)

#For each sample, determine the total read counts
for (a in sample_list_complete){
 #Set sum to 0
 sum=0
 
 #For each genus
 for (b in genus_found){
  #Add the number of counts together
  sum <- samples_otu_complete[b,a] + sum
 }
 
 #When complete, push the sum to a new "Total_Reads" row
 samples_otu_complete["Total_Reads",a] <- sum
}
write.csv(samples_otu_complete,"temp.csv")
```

#4. OTU Counts by Control Type
Determine genus count for each sample, average across control type
```{r}
#Create control sample list, control type list
sample_list_f <- unique(samples_variables_f$NephID)
control_types

#Genus count for each sample 
for (a in sample_list_f){
  count_present=0
  
  for (b in genus_found){
   value <- samples_otu_complete[b,a]
   
   if(value==0) {
    next
   } else{
     count_present = count_present +1
    }
  }
  samples_variables_f[a,"GenusObs"] <- count_present

 }
 
###Testing
write.csv(samples_variables_f,"temp1.csv")


#New Database to store final Genus information
samples_obs_count_type <- data.frame()
kit_types<-unique(samples_variables_f$Kit)

#Start counter
count_present=0

#For each control type
for (a in control_types){
 
 #For each kit
 for (b in kit_types){
  #Create temp database for kit and control type
  temp <- subset(samples_variables_f,Type==a)
  temp <- subset(temp,Kit==b)
  
  #Create sample list
  sample_list<- unique(temp$NephID)
  
  #For each of these samples
  for (c in sample_list){
   
   #Find observed value and add
   value <- temp[c,"GenusObs"]
   count_present=count_present+value
  }
  
  #Create new column and pass average
  col_name <- paste(a,"_",b,sep="")
  samples_obs_count_type["GenObs_Av",col_name]<-count_present/length(sample_list)
  
  #Reset counter
  count_present=0
 }
}
write.csv(samples_obs_count_type,"Data/study_avgenus.csv")
```

#5. OTU Percent - Analysis Genus OTU Abundance
```{r}
#Create new table
samples_obs_abund = data.frame()

##Run through each sample
for (a in sample_list_f){
 
 #Check for each of the expected genus
 for (b in genus_found){
  
  #Determine the reads for the genus, and the total reads for the sample
  reads <- samples_otu_complete[b,a]
  total <- samples_otu_complete["Total_Reads",a]

  #If value is NA (meaning that the genus was not found in sample), skip
  #If reads are >0 then add the read count to reads_total

  if(is.na(reads) | is.null(reads)) {
   next
  } else {
   percent_exp <- (reads/total)
   
   #Add percent to the samples_obs_abund database
   samples_obs_abund[b,a] <- percent_exp
  }
 }
}

#Output the table of all samples
write.csv(samples_obs_abund,"Data/study_obs_abund.csv")
```

#6. OTU Percent - Analysis Genus OTU Abundance by Control Type
```{r}
#New Database to store final Genus information
samples_obs_abund_vari <- data.frame()

#For each of the control types
for (a in control_types){
 
  #For each Genus in list
  for (b in genus_found){
   
   #Reset Counter
   reads <- 0
   total <- 0
   
   #Create a list of samples that are of the specified control type
   sample_list_temp<- subset(samples_variables_f, Type==a)
   sample_list_temp <- unique(sample_list_temp$NephID)
   
   #For each of the samples in the sample list
   for(c in sample_list_temp){
    
    #Add up the total percentages from samples_obs_abund [genus,sampleID]
    reads <- samples_otu_complete [b,c] + reads
    total <- samples_otu_complete["Total_Reads",c] + total
   }
   
   #average of genus level by taking total number of reads of that genus divded by total number of reads for samples included
   #Push value to new dataframe
   percent_obs <- reads / total
   samples_obs_abund_vari[b,a] <- percent_obs
  }
 }

#Output to text file
write.csv(samples_obs_abund_vari,"Data/study_obs_abund_vari.csv")
```

#7. OTU Percent - Analysis Exp OTU Rel Abun by Control by Kit
```{r}
#For each of the kits types
for (a in kit_types){
 
  #For each Genus in list
  for (b in genus_found){
   
   #Reset Counter
   reads <- 0
   total <- 0
   
   #Create a list of samples that are of the specified control type
   sample_list_temp<- subset(samples_variables_f, Kit==a)
   sample_list_temp <- unique(sample_list_temp$NephID)
   
   #For each of the samples in the sample list
   for(c in sample_list_temp){
    
    #Add total reads from samples_otu_complete [genus,sampleID]
    reads <- samples_otu_complete [b,c] + reads
    total <- samples_otu_complete["Total_Reads",c] + total
   }
   
   #Calc Rel abund (total number of reads / total number of reads)
   #Push value to new dataframe
   percent_obs <- reads / total
   samples_obs_abund_vari[b,a] <- percent_obs
  }
}
#Output to text file (adds new columns to prevoius version)
write.csv(samples_obs_abund_vari,"Data/study_obs_abund_vari.csv")

#For each of the kits types
for (a in kit_types){
 
 #For each of the control types
 for(b in control_types){
  #Create a list of samples that are of the specified control type
  sample_list_temp<- subset(samples_variables_f, Kit==a & Type==b)
  sample_list_temp <- unique(sample_list_temp$NephID)
  
  #For each of the found genus
  for(c in genus_found){
   
   #Reset Counter
   reads <- 0
   total <- 0

   #For each of the samples in the sample list
   for(d in sample_list_temp){
    
    #Add reads from samples_otu_complete [genus,sampleID]
    reads <- samples_otu_complete [c,d] + reads
    total <- samples_otu_complete["Total_Reads",d] + total
   }
   
   #Calc Rel abund (total number of reads / total number of reads)
   percent_obs <- (reads/total)
   
   #Create new header
   #Push value to new dataframe
   col_name <- paste(a,"-",b)
   samples_obs_abund_vari[c,col_name] <- percent_obs
  }
 }
}
#Output to text file (adds columns to file)
write.csv(samples_obs_abund_vari,"Data/study_obs_abund_vari.csv")
```

#8 - Plot Interkit Sample Variability
```{r}
#Sort genus list alphabetically
genus_found<-sort(genus_found)

#Create list for plots, start counter
plot_list <- list()
i=1

#For each of the control types
for (a in control_types){
 
 #For each kit
 for (b in kit_types){
  
  #Create new dataframe and start counter
  temp<-data.frame()
  count=1
    
  ###Create list of sample ids with specific variables
  temp2 <- subset(samples_variables_f,Kit==b)
  temp2 <- subset(temp2,Type==a)
  sample_list <- unique(temp2$NephID)
  
  #For each sample in list
  for (c in sample_list){
   
   #For each genus
   for (d in genus_found){
    
    #Determine the abundance value
    value <- samples_obs_abund[d,c]
    aliquot<-samples_variables_f[c,"AliquotNumber"]
   
    #If the value is null or smaller than 1 X10-5, skip
    if(is.na(value) | is.null(value) | value<0.0000000001){
     next}
    
    #Otherwise add to datatable
    else{
     temp[count,"Abundance"]<-value
     temp[count,"Genus"]<-d
     
     if(aliquot=="None"){
      temp[count,"SampleID"]<-c
     }else{
      temp[count,"SampleID"]<-aliquot
     }
     count=count+1
     }
   }
  }
  
  #Create file name
  file_name <- paste ("Data/Study/SampleVari",b,a,sep = "_")
  file_name <- paste(file_name,".csv",sep="")
  
  #Sort and save the datatable
  temp<-temp[order(temp$SampleID), ]
  write.csv(temp,file_name)
  
  #Create graph title
  graph_title<-paste(a,b,sep="_")
    
  #Create Graph for the datatable
  graph_new<- ggplot(temp,aes(x =SampleID , y = Abundance, fill = Genus)) + 
   geom_bar(position = position_fill(), stat = "identity") +
   scale_y_continuous(labels = percent_format())+
   labs(title=graph_title, x="Sample ID", y= "Relative Abundance")+
   theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
   #+ theme(legend.position = "bottom",legend.box = "vertical")
   theme(legend.position = "none") ##Remove Legend
  
  graph_new  
  plot_list[[i]] = graph_new
  i=i+1
    
  #Reset data table
    temp<-data.frame()
  }
}

#Print the plots
i=1
for (a in control_types){
 
 for (b in kit_types){
  
  #Create file name
  file_name<-paste("Interkit Variability",a,b,sep="_")
  file_name<-paste("Data/Graphs/",file_name,".tiff",sep="")
  tiff(file_name)
  print(plot_list[[i]])
  dev.off()
  i=i+1
 }
}
```

#9 - Plot All Sample Variability in individual kit graphs
```{r}
#Create list for plots, start counter
plot_list <- list()
i=1

#For each of the control types
for (a in control_types){
 
 #Create new dataframe and start counter
 temp<-data.frame()
 count=1
    
 ###Create list of sample ids with specific variables
 temp2 <- subset(samples_variables_f,Type==a)
 sample_list <- unique(temp2$NephID)
 
 #For each sample in list
 for (b in sample_list){
  
  #For each genus
  for (c in genus_found){
   
   #Determine the abundance value
   value <- samples_obs_abund[c,b]
   aliquot<-samples_variables_f[b,"AliquotNumber"]
   
   #If the value is null or smaller than 1 X10-7, skip
   if(is.na(value) | is.null(value) | value<0.000000000001){
    next}
    
   #Otherwise add to datatable
   else{
    temp[count,"Abundance"]<-value
    temp[count,"Genus"]<-c
    
    if(aliquot=="None"){
     
     temp[count,"SampleID"]<-b
     }else{
      
      #Add kit info
      aliquot <- paste(aliquot,samples_variables_f[b,"Kit"],sep="_")
      temp[count,"SampleID"]<-aliquot
     }
    count=count+1
   }
  }
 }
 
 #Create file name
 file_name <- paste ("Data/Study/StudyVari",a,sep = "_")
 file_name <- paste(file_name,".csv",sep="")
 
 #Sort and save the datatable
 temp<-temp[order(temp$SampleID), ]
 write.csv(temp,file_name)
  
 #Create graph title
 graph_title<-paste(a,sep="_")
 
 #Create Graph for the datatable
 graph_new<- ggplot(temp,aes(x =SampleID , y = Abundance, fill = Genus)) + 
  geom_bar(position = position_fill(), stat = "identity") +
  scale_y_continuous(labels = percent_format())+
  labs(title=graph_title, x="Sample ID", y= "Relative Abundance")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  #+ theme(legend.position = "bottom",legend.box = "vertical")
  theme(legend.position = "none") ##Remove Legend
  
 plot_list[[i]] = graph_new
 i=i+1
    
 #Reset data table
   temp<-data.frame()
}

#Print the plots
i=1
for (a in control_types){
 
 #Create file name
 file_name<-paste("Interstudy Variability",a,sep="_")
 file_name<-paste("Data/Graphs/",file_name,".tiff",sep="")
 tiff(file_name)
 print(plot_list[[i]])
 dev.off()
 i=i+1
}
```

#9 - Plot All Sample Variability in individual kit graphs
```{r}
#Create list for plots, start counter
plot_list <- list()
i=1

#Create new dataframe and start counter
temp<-data.frame()
count=1
    
for (a in kit_types){
 ###Create list of sample ids with specific variables
 temp2 <- subset(samples_variables_f,Type=="Study" & Kit==a)
 sample_list <- unique(temp2$NephID)
 
 for (b in sample_list){
  
  for (c in genus_found){
   temp[count,"Value"]<-samples_obs_abund[c,b]
   temp[count,"Kit"]<-a
   temp[count,"Aliquot"]<-samples_variables_f[b,"AliquotNumber"]
   temp[count,"SampleID"]<- paste(a,samples_variables_f[b,"AliquotNumber"],sep="_")
   temp[count,"ID"]<-b
   temp[count,"Genus"]<-c
   count=count+1
  }
 }
}
write.csv(temp,"temp4.csv")
#Remove DSP kit samples


temp<-read.csv("temp4.csv",header = TRUE)

#Create Graph for the datatable
graph_title <- "Fecal Samples by Kit by Aliquot ID"
graph_new<- ggplot(temp,aes(x =SampleID , y = Value, fill = Genus)) + 
  geom_bar(position = position_fill(), stat = "identity") +
  scale_y_continuous(labels = percent_format())+
  labs(title=graph_title, x="Sample ID", y= "Relative Abundance")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  #+ theme(legend.position = "bottom",legend.box = "vertical")
  theme(legend.position = "none") ##Remove Legend
  
plot_list[[i]] = graph_new
 
#Print the plot
file_name <- "StudyVariability"
file_name<-paste("Data/Graphs/",file_name,".tiff",sep="")
tiff(file_name)
print(plot_list[[i]])
dev.off()

```



#10 - Plot all sample variability in one graph
```{r}
#Edit the data to be able to numerically order the aliquots
temp<-read.csv("Data/Study/StudyVari_Study_Edited(Aliquot).csv",header = TRUE)

#Edit the data to be able to numerically order the aliquots by kit
temp<-read.csv("Data/Study/StudyVari_Study_Edited(Kit).csv",header = TRUE)
temp$X<- factor(temp$X, levels=temp$X)

#Create a plot for 
ggplot(temp,aes(x =SampleID , y = Abundance, fill = Genus)) + 
 geom_bar(position = position_fill(), stat = "identity") +
 scale_y_continuous(labels = percent_format())+
 labs(title="Fresh-Frozen Fecal Study Samples", x="Extraction Kit_Aliquot ID", y= "Relative Abundance")+
 theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
 #+ theme(legend.position = "bottom",legend.box = "vertical")
 theme(legend.position = "none") ##Remove Legend

```